{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This juypter notebook is tartget towards defining alleles "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The idea is that alleles are on corresponding haplotig and a good blast hit.\n",
    "\n",
    "map all h contigs on their corresponding p alleles. Use the Assembletic '*oriented_coords.csv' file for this purpose.\n",
    "This needs to be converted to a bed file or a gff file in order to pull out the overlap between gene models and the h_contig using either bedtools or pandas. Check what overlaps can spit out on the other side. \n",
    ">Considerations here are what happens if nothing overlaps or what if muliptle overlaps. \n",
    "If nothing overlaps it would be good to look if this area is covered by 2x coverage when doing ph mapping.\n",
    "\n",
    "Do gene blast of p on h with certain cut off. Pull out the best hit that is located on the corresponding haplotig. If that is not there look for best hit on other haplotigs or primary contigs that is not itself.\n",
    "\n",
    ">Considerations are that p-h parings that are best hits and corresponding haplotig overule all others. There can only be a single hit. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from pybedtools import BedTool\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "import subprocess\n",
    "import shutil\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Define the PATH\n",
    "BASE_AA_PATH = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12'\n",
    "BASE_A_PATH = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly'\n",
    "ASSEMBLETICS_INPATH = os.path.join(BASE_AA_PATH, 'Assembletics')\n",
    "BLAST_DB = os.path.join(BASE_AA_PATH, 'blast_DB')\n",
    "OUT_PATH = os.path.join(BASE_AA_PATH, 'allele_analysis')\n",
    "OUT_tmp = os.path.join(OUT_PATH, 'tmp')\n",
    "ALLELE_PATH = os.path.join(OUT_PATH, 'alleles')\n",
    "if not os.path.isdir(OUT_PATH):\n",
    "    os.mkdir(OUT_PATH)\n",
    "if not os.path.isdir(OUT_tmp):\n",
    "    os.mkdir(OUT_tmp)\n",
    "if not os.path.exists(os.path.join(OUT_PATH, 'alleles')):\n",
    "    os.mkdir(os.path.join(OUT_PATH, 'alleles'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Define your p and h genome and move it into the allele analysis folder\n",
    "p_genome = 'Pst_104E_v12_p_ctg'\n",
    "h_genome = 'Pst_104E_v12_h_ctg'\n",
    "genome_file_suffix = '.genome_file'\n",
    "for x in (x + '.fa' for x in [p_genome, h_genome]):\n",
    "    shutil.copy2(BASE_A_PATH+'/'+x, OUT_tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Define ENV parameters for blast hits and threads used in blast analysis\n",
    "n_threads = 16\n",
    "e_value = 1e-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def att_column_p(x, y, z, a, b):\n",
    "    '''Generate attribute column of h on p p_gff dataframe.\n",
    "    '''\n",
    "    string = 'h_contig=%s;query_start=%s;query_stop=%s;query_length=%s;query_aln_ln=%s' % (x, str(y), str(z), str(a), str(b))\n",
    "    return string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def att_column_h(x, y, z, a, b):\n",
    "    '''Generate attribute column of h on p h_gff dataframe.\n",
    "    '''\n",
    "    string = 'p_contig=%s;ref_start=%s;ref_stop=%s;ref_length=%s;ref_aln_ln=%s' % (x, str(y), str(z), str(a), str(b))\n",
    "    return string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def gene_contig_subset(df, contig, feature):\n",
    "    '''Input the gff and subset the feature columne by gene.\n",
    "        Return data frame subset by contig and gene'''\n",
    "    tmp_df = df[(df[0].str.contains(contig)) & (df[2]==feature)]\n",
    "    tmp_df.sort_values(3,inplace=True)\n",
    "    tmp_df.reset_index(drop=True, inplace=True)\n",
    "    return tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def same_contig_blast(x,y):\n",
    "    '''Function that checks if the blast hit in columne y is on the same contig as the the query sequence in\n",
    "    column y.\n",
    "    '''\n",
    "    q_contig = re.search(r'[p|h]([a-z]*_[0-9]*)_?', x).group(1)\n",
    "    if y != 'NaN':\n",
    "        hit_contig = re.search(r'[p|h]([a-z]*_[0-9]*)_?', y).group(1)\n",
    "    else:\n",
    "        hit_contig = 'NaN'\n",
    "    if q_contig == hit_contig:\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def target_on_maped_haplotig(t_contig, h_contig_overlap):\n",
    "    '''Simple function that checks if the target contig is in the list of h overlaping contigs'''\n",
    "    if t_contig == False or h_contig_overlap == False:\n",
    "        return False  \n",
    "    else:\n",
    "        if t_contig in h_contig_overlap:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Considerations for blast analysis\n",
    "Do gene blast and protein blast. Initially go off protein blast results. They should be more imformative as it also includes the coding region and the frame.\n",
    "> Pull this in as a dataframe and filter it down as well to each contig. Do blasts both ways as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All databases generated and ready to go!\n"
     ]
    }
   ],
   "source": [
    "#generate the blast databases if not already present\n",
    "os.chdir(BLAST_DB)\n",
    "blast_dir_content = os.listdir(BLAST_DB)\n",
    "for x in blast_dir_content:\n",
    "    if x.endswith('.fa') and ({os.path.isfile(x + e) for e in ['.psq', '.phr', '.pin'] } != {True}\\\n",
    "           and {os.path.isfile(x + e) for e in ['.nin', '.nhr', '.nsq'] } != {True} ):\n",
    "\n",
    "        make_DB_options = ['-in']\n",
    "        make_DB_options.append(x)\n",
    "        make_DB_options.append('-dbtype')\n",
    "        if 'protein' in x:\n",
    "            make_DB_options.append('prot')\n",
    "        else:\n",
    "            make_DB_options.append('nucl')\n",
    "        make_DB_command = 'makeblastdb %s' % ' '.join(make_DB_options)\n",
    "        make_DB_stderr = subprocess.check_output(make_DB_command, shell=True, stderr=subprocess.STDOUT)\n",
    "        print('%s is done!' % make_DB_command)\n",
    "print(\"All databases generated and ready to go!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastp -query Pst_104E_v12_p_ctg.anno.protein.fa -db Pst_104E_v12_h_ctg.anno.protein.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_p_ctg.Pst_104E_v12_h_ctg.0.001.blastp.outfmt6\n",
      "Previously done already!\n",
      "blastp -query Pst_104E_v12_h_ctg.anno.protein.fa -db Pst_104E_v12_p_ctg.anno.protein.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_h_ctg.Pst_104E_v12_p_ctg.0.001.blastp.outfmt6\n",
      "Previously done already!\n",
      "blastn -query Pst_104E_v12_p_ctg.anno.gene.fa -db Pst_104E_v12_h_ctg.anno.gene.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_p_ctg.Pst_104E_v12_h_ctg.0.001.blastn.outfmt6\n",
      "Previously done already!\n",
      "blastn -query Pst_104E_v12_h_ctg.anno.gene.fa -db Pst_104E_v12_p_ctg.anno.gene.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_h_ctg.Pst_104E_v12_p_ctg.0.001.blastn.outfmt6\n",
      "Previously done already!\n"
     ]
    }
   ],
   "source": [
    "blast_dict = {}\n",
    "blast_dict['gene_fa'] = [x for x  in os.listdir(BLAST_DB) if x.endswith('gene.fa') and 'ph_ctg' not in x]\n",
    "blast_dict['protein_fa'] = [x for x  in os.listdir(BLAST_DB) if x.endswith('protein.fa') and 'ph_ctg' not in x]\n",
    "blast_stderr_dict = {}\n",
    "for key in blast_dict.keys():\n",
    "    for n, query in enumerate(blast_dict[key]):\n",
    "        tmp_list = blast_dict[key][:]\n",
    "        tmp_stderr_list = []\n",
    "        #this loops through all remaining files and does blast against all those\n",
    "        del tmp_list[n]\n",
    "        for db in tmp_list:\n",
    "            blast_options = ['-query']\n",
    "            blast_options.append(query)\n",
    "            blast_options.append('-db')\n",
    "\n",
    "\n",
    "            blast_options.append(db)\n",
    "            blast_options.append('-outfmt 6')\n",
    "            blast_options.append('-evalue')\n",
    "            blast_options.append(str(e_value))\n",
    "            blast_options.append('-num_threads')\n",
    "            blast_options.append(str(n_threads))\n",
    "            blast_options.append('>')\n",
    "            if 'gene' in query and 'gene' in db:\n",
    "                out_name_list = [ query.split('.')[0], db.split('.')[0], str(e_value), 'blastn.outfmt6']\n",
    "                out_name = os.path.join(OUT_PATH,'.'.join(out_name_list))\n",
    "                blast_options.append(out_name)\n",
    "                blast_command = 'blastn %s' % ' '.join(blast_options)\n",
    "            elif 'protein' in query and 'protein' in db:\n",
    "                out_name_list = [ query.split('.')[0], db.split('.')[0], str(e_value), 'blastp.outfmt6']\n",
    "                out_name = os.path.join(OUT_PATH,'.'.join(out_name_list))\n",
    "                blast_options.append(out_name)\n",
    "                blast_command = 'blastp %s' % ' '.join(blast_options)\n",
    "            print(blast_command)\n",
    "            if not os.path.exists(out_name):\n",
    "                blast_stderr_dict[blast_command] = subprocess.check_output(blast_command, shell=True, stderr=subprocess.STDOUT)\n",
    "                print(\"New blast run and done!\")\n",
    "            else:\n",
    "                blast_stderr_dict[blast_command] = 'Previously done already!'\n",
    "                print('Previously done already!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#make a sequnece length dict for all genes and proteins for which blast was performed\n",
    "seq_list = []\n",
    "length_list =[]\n",
    "for key in blast_dict.keys():\n",
    "    for file in blast_dict[key]:\n",
    "        for seq in SeqIO.parse(open(file), 'fasta'):\n",
    "            seq_list.append(seq.id)\n",
    "            length_list.append(len(seq.seq))\n",
    "length_dict = dict(zip(seq_list, length_list))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60498"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(length_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#get assembletics folders\n",
    "ass_folders = [os.path.join(ASSEMBLETICS_INPATH, x) for x in os.listdir(ASSEMBLETICS_INPATH) if x.endswith('_php_8kbp')]\n",
    "ass_folders.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#read in the gff files\n",
    "p_gff = pd.read_csv(BASE_A_PATH+'/'+p_genome+'.anno.gff3', sep='\\t', header=None)\n",
    "h_gff = pd.read_csv(BASE_A_PATH+'/'+h_genome+'.anno.gff3', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:33: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:34: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    }
   ],
   "source": [
    "#read in the blast df and add QLgth and QCov columns\n",
    "blast_out_dict = {}\n",
    "blast_header = ['Query', 'Target', 'PctID', 'AlnLgth', 'NumMis', 'NumGap', 'StartQuery', 'StopQuery', 'StartTarget',\\\n",
    "              'StopTarget', 'e-value','BitScore']\n",
    "for key in blast_stderr_dict.keys():\n",
    "    file = key.split('>')[-1][1:]\n",
    "    tmp_key = file.split('/')[-1]\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header=None, names=blast_header)\n",
    "    tmp_df[\"QLgth\"] = tmp_df[\"Query\"].apply(lambda x: length_dict[x])\n",
    "    tmp_df[\"QCov\"] = tmp_df['AlnLgth']/tmp_df['QLgth']*100\n",
    "    tmp_df.sort_values(by=['Query', 'e-value','BitScore', ],ascending=[True, True, False], inplace=True)\n",
    "    #now make sure to add proteins/genes without blast hit to the dataframes\n",
    "    #check if gene blast or protein blast and pull out respective query file\n",
    "    if file.split('.')[-2] == 'blastp':\n",
    "        tmp_blast_seq = [x for x in blast_dict['protein_fa'] if x.startswith(tmp_key.split('.')[0])][0]\n",
    "    elif file.split('.')[-2] == 'blastn':\n",
    "        tmp_blast_seq = [x for x in blast_dict['gene_fa'] if x.startswith(tmp_key.split('.')[0])][0]\n",
    "    tmp_all_blast_seq = []\n",
    "    #make list off all query sequences\n",
    "    for seq in SeqIO.parse(os.path.join(BLAST_DB, tmp_blast_seq), 'fasta'):\n",
    "        tmp_all_blast_seq.append(str(seq.id))\n",
    "    tmp_all_queries_w_hit = tmp_df[\"Query\"].unique()\n",
    "    tmp_queries_no_hit = set(tmp_all_blast_seq) - set(tmp_all_queries_w_hit)\n",
    "    no_hit_list = []\n",
    "    #loop over the quieres with no hit and make list of list out of them the first element being the query id\n",
    "    for x in tmp_queries_no_hit:\n",
    "        NA_list = ['NaN'] * len(tmp_df.columns)\n",
    "        NA_list[0] = x\n",
    "        no_hit_list.append(NA_list)\n",
    "    tmp_no_hit_df = pd.DataFrame(no_hit_list)\n",
    "    tmp_no_hit_df.columns = tmp_df.columns\n",
    "    tmp_no_hit_df['QLgth'] = tmp_no_hit_df.Query.apply(lambda x: length_dict[x])\n",
    "    tmp_df = tmp_df.append(tmp_no_hit_df)\n",
    "    tmp_df['q_contig'] = tmp_df['Query'].str.extract(r'([p|h][a-z]*_[^.]*).?')\n",
    "    tmp_df['t_contig'] = tmp_df['Target'].str.extract(r'([p|h][a-z]*_[^.]*).?')\n",
    "    #fix that if you don't extract anything return False and not 'nan'\n",
    "    tmp_df['t_contig'].fillna(False, inplace=True)\n",
    "    tmp_df['q_contig == t_contig'] = tmp_df.apply(lambda row: same_contig_blast(row['Query'], row['Target']), axis =1)\n",
    "    tmp_df.reset_index(inplace=True, drop=True)\n",
    "    blast_out_dict[tmp_key] = tmp_df.iloc[:,:]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Check on /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/Assembletics/Pst_104E_v12_pcontig_118_php_8kbp assembletics\n",
      "Check on /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/Assembletics/Pst_104E_v12_pcontig_193_php_8kbp assembletics\n"
     ]
    }
   ],
   "source": [
    "for folder in ass_folders:\n",
    "    '''\n",
    "    From oriented_coords.csv files generate gff files with the following set up.\n",
    "    h_contig overlap\n",
    "    'query', 'ref', 'h_feature', 'query_start', 'query_end', 'query_aln_len', 'strand', 'frame', 'attribute_h'\n",
    "    p_contig overlap\n",
    "    'ref', 'query', 'p_feature','ref_start','ref_end', 'alignment_length', 'strand', 'frame', 'attribute_p'.\n",
    "    Save those file in a new folder for further downstream analysis. The same folder should include the contig and gene filtered\n",
    "    gffs. \n",
    "    '''\n",
    "    orient_coords_file = [os.path.join(folder, x) for x in os.listdir(folder) if x.endswith('oriented_coords.csv')][0]\n",
    "    #load in df and generate several additions columns\n",
    "    tmp_df = pd.read_csv(orient_coords_file, sep=',')\n",
    "    #check if there is any resonable alignment if not go to the next one\n",
    "    if len(tmp_df) == 0:\n",
    "        print('Check on %s assembletics' % (folder))\n",
    "        continue\n",
    "    tmp_df['p_feature'] = \"haplotig\"\n",
    "    tmp_df['h_feature'] ='primary_contig'\n",
    "    tmp_df['strand'] = \"+\"\n",
    "    tmp_df['frame'] = 0\n",
    "    tmp_df['query_aln_len'] = abs(tmp_df['query_end']-tmp_df['query_start'])\n",
    "    tmp_df['alignment_length'] = abs(tmp_df['ref_end'] - tmp_df['ref_start'])\n",
    "    tmp_df.reset_index(drop=True, inplace=True)\n",
    "    tmp_df.sort_values('query', inplace=True)\n",
    "    tmp_df['attribute_p'] = tmp_df.apply(lambda row: att_column_p(row['query'],row['query_start'], row['query_end'],\\\n",
    "                                                         row['query_length'], row['query_aln_len']), axis=1)\n",
    "    \n",
    "    tmp_df['attribute_h'] = tmp_df.apply(lambda row: att_column_h(row['ref'], row['ref_start'], row['ref_end'],\\\n",
    "                                                                  row['ref_length'], row['alignment_length']), axis=1)\n",
    "    \n",
    "    #generate tmp gff dataframe\n",
    "    tmp_p_gff_df = tmp_df.loc[:, ['ref', 'query', 'p_feature','ref_start','ref_end', 'alignment_length', 'strand', 'frame', 'attribute_p']]\n",
    "    tmp_p_gff_df.sort_values('ref_start',inplace=True)\n",
    "    tmp_h_gff_df = tmp_df.loc[:, ['query', 'ref', 'h_feature', 'query_start', 'query_end', 'query_aln_len', 'strand', 'frame', 'attribute_h']]\n",
    "    tmp_h_gff_df.sort_values('query_start', inplace=True)\n",
    "    #make the outfolder and save gff files of overlap and filtered gene files\n",
    "    folder_suffix = folder.split('/')[-1]\n",
    "    #get the contig\n",
    "    contig = re.search(r'p([a-z]*_[0-9]*)_', folder_suffix).group(1)\n",
    "    out_folder = os.path.join(OUT_PATH, folder_suffix)\n",
    "    if not os.path.exists(out_folder):\n",
    "        os.mkdir(out_folder)\n",
    "    out_name_p = os.path.join(out_folder, folder_suffix )\n",
    "    tmp_p_gff_df.to_csv(out_name_p+'.p_by_h_cov.gff', sep='\\t' ,header=None, index =None)\n",
    "    out_name_h =  os.path.join(out_folder, folder_suffix)\n",
    "    tmp_h_gff_df.to_csv(out_name_h+'.h_by_p_cov.gff' , sep='\\t' ,header=None, index =None)\n",
    "    #write those out to new folder for php together with the p and h gff of genes\n",
    "    p_gene_gff = gene_contig_subset(p_gff, contig, 'gene')\n",
    "    p_gene_gff.to_csv(out_name_p+'.p_gene.gff', sep='\\t' ,header=None, index =None)\n",
    "    h_gene_gff = gene_contig_subset(h_gff, contig, 'gene')\n",
    "    h_gene_gff.to_csv(out_name_h+'.h_gene.gff', sep='\\t' ,header=None, index =None)\n",
    "    \n",
    "    #next would be to do overlap between the gene gff and the corresponding alingment gff. Write this out. Dict for each gene and its corresponding h_contig/p_contig\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#now loop over the outfolders\n",
    "out_folders = [os.path.join(OUT_PATH, x) for x in os.listdir(OUT_PATH) if x.endswith('_php_8kbp')]\n",
    "out_folders.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 588\n",
      "This is the lenght of the intersect 548\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 496\n",
      "This is the lenght of the intersect 417\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 319\n",
      "This is the lenght of the intersect 352\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 388\n",
      "This is the lenght of the intersect 308\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 266\n",
      "This is the lenght of the intersect 278\n",
      "This is the lenght of the intersect 255\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 335\n",
      "This is the lenght of the intersect 299\n",
      "This is the lenght of the intersect 282\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 213\n",
      "This is the lenght of the intersect 226\n",
      "This is the lenght of the intersect 276\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 262\n",
      "This is the lenght of the intersect 294\n",
      "This is the lenght of the intersect 289\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 253\n",
      "This is the lenght of the intersect 206\n",
      "This is the lenght of the intersect 240\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 116\n",
      "This is the lenght of the intersect 240\n",
      "This is the lenght of the intersect 174\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 196\n",
      "This is the lenght of the intersect 227\n",
      "This is the lenght of the intersect 181\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 164\n",
      "This is the lenght of the intersect 7\n",
      "This is the lenght of the intersect 155\n",
      "This is the lenght of the intersect 186\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 133\n",
      "This is the lenght of the intersect 191\n",
      "This is the lenght of the intersect 153\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 152\n",
      "This is the lenght of the intersect 136\n",
      "This is the lenght of the intersect 135\n",
      "This is the lenght of the intersect 165\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 111\n",
      "This is the lenght of the intersect 156\n",
      "This is the lenght of the intersect 85\n",
      "This is the lenght of the intersect 95\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 140\n",
      "This is the lenght of the intersect 126\n",
      "This is the lenght of the intersect 69\n",
      "This is the lenght of the intersect 112\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 157\n",
      "This is the lenght of the intersect 108\n",
      "This is the lenght of the intersect 131\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 95\n",
      "This is the lenght of the intersect 136\n",
      "This is the lenght of the intersect 62\n",
      "This is the lenght of the intersect 104\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 55\n",
      "This is the lenght of the intersect 87\n",
      "This is the lenght of the intersect 92\n",
      "This is the lenght of the intersect 94\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 33\n",
      "This is the lenght of the intersect 61\n",
      "This is the lenght of the intersect 4\n",
      "This is the lenght of the intersect 91\n",
      "This is the lenght of the intersect 97\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 31\n",
      "This is the lenght of the intersect 13\n",
      "This is the lenght of the intersect 10\n",
      "This is the lenght of the intersect 9\n",
      "This is the lenght of the intersect 33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 18\n",
      "This is the lenght of the intersect 2\n",
      "This is the lenght of the intersect 16\n",
      "This is the lenght of the intersect 64\n",
      "This is the lenght of the intersect 36\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 50\n",
      "This is the lenght of the intersect 37\n",
      "This is the lenght of the intersect 37\n",
      "This is the lenght of the intersect 31\n",
      "This is the lenght of the intersect 11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 2\n",
      "This is the lenght of the intersect 20\n",
      "This is the lenght of the intersect 21\n",
      "This is the lenght of the intersect 20\n",
      "This is the lenght of the intersect 0\n",
      "This is the lenght of the intersect 6\n",
      "This is the lenght of the intersect 7\n",
      "This is the lenght of the intersect 17\n",
      "This is the lenght of the intersect 8\n",
      "This is the lenght of the intersect 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 12\n",
      "This is the lenght of the intersect 7\n",
      "This is the lenght of the intersect 4\n",
      "This is the lenght of the intersect 0\n",
      "This is the lenght of the intersect 5\n",
      "This is the lenght of the intersect 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the lenght of the intersect 6\n",
      "This is the lenght of the intersect 9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n",
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:16: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n"
     ]
    }
   ],
   "source": [
    "p_gene_h_contig_overlap_dict = {}\n",
    "for out_folder in out_folders:\n",
    "    tmp_folder_content = os.listdir(out_folder)\n",
    "    tmp_p_gene_gff = [os.path.join(out_folder, x) for x in tmp_folder_content if x.endswith('p_gene.gff') ][0]\n",
    "    tmp_p_gene_bed = pybedtools.BedTool(tmp_p_gene_gff).remove_invalid().saveas()\n",
    "    tmp_p_by_h_gff = [os.path.join(out_folder, x) for x in tmp_folder_content if x.endswith('p_by_h_cov.gff') ][0]\n",
    "    tmp_p_by_h_bed = pybedtools.BedTool(tmp_p_by_h_gff).remove_invalid().saveas()\n",
    "    #generate a bed intersect\n",
    "    p_gene_and_p_by_h = tmp_p_gene_bed.intersect(tmp_p_by_h_bed, wb=True)\n",
    "    #check the length of the obtained intersect is \n",
    "    print(\"This is the lenght of the intersect %i\" %len(p_gene_and_p_by_h ))\n",
    "    if len(p_gene_and_p_by_h) == 0:\n",
    "        continue\n",
    "    #turn this into a dataframe \n",
    "    p_gene_and_p_by_h_df = p_gene_and_p_by_h.to_dataframe()\n",
    "    p_gene_and_p_by_h_df['p_gene'] = p_gene_and_p_by_h_df[8].str.extract(r'ID=([^;]*);')\n",
    "    p_gene_and_p_by_h_df['p_protein'] = p_gene_and_p_by_h_df['p_gene'].str.replace('TU', 'model')\n",
    "    #safe this in the same folder as dataframe\n",
    "    p_gene_and_p_by_h_df.to_csv(tmp_p_gene_gff.replace('p_gene.gff', 'gene_haplotig_intersect.df'), sep='\\t', index=None)\n",
    "    p_gene_and_p_by_h_df_2 = p_gene_and_p_by_h_df.loc[:, [10, 'p_protein']]\n",
    "    #make a dict for the overlap of each gene with it's correspoding haplotig\n",
    "    tmp_p_gene_h_overlap_df = p_gene_and_p_by_h_df_2.groupby('p_protein')[10].apply(list)\n",
    "    tmp_p_gene_h_overlap_dict = dict(zip(tmp_p_gene_h_overlap_df.index, tmp_p_gene_h_overlap_df.values))\n",
    "    p_gene_h_contig_overlap_dict.update(tmp_p_gene_h_overlap_dict)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#get out the p_protein on h_protein blast out dataframe and appand the h_contig overlap for each protein\n",
    "_key = [x for x in blast_out_dict.keys() if x.startswith(p_genome) and x.split('.')[-2] == 'blastp' and x.endswith('outfmt6')][0]\n",
    "p_gene_h_contig_overlap_df = pd.DataFrame([p_gene_h_contig_overlap_dict.keys(), p_gene_h_contig_overlap_dict.values()],  index = ['p_protein', 'h_contig_overlap']).T\n",
    "_tmp_df = blast_out_dict[_key]\n",
    "_tmp_df = _tmp_df.merge(p_gene_h_contig_overlap_df, how='outer' ,left_on='Query', right_on='p_protein')\n",
    "_tmp_df['h_contig_overlap'].fillna(False, inplace=True)\n",
    "#check if the protein hit resites on the overlapping haplotig\n",
    "_tmp_df['t_contig == h_contig_overlap'] = _tmp_df.apply(lambda row: target_on_maped_haplotig(row['t_contig'], row['h_contig_overlap']), axis =1 )\n",
    "#blast_out_dict[_key] = _tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#write the dataframe again\n",
    "_tmp_df.to_csv(os.path.join(OUT_PATH, _key+'.allele_analysis'), index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now mangale the data a bit\n",
    "#write out t_contig == h_contig_overlap best hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_tmp_grouped = _tmp_df[_tmp_df['t_contig == h_contig_overlap'] == True].groupby('Query')\n",
    "\n",
    "\n",
    "best_hits_t_contig_and_h_contig = _tmp_grouped.apply(lambda g: g[(g['e-value'] == g['e-value'].min())])\n",
    "\n",
    "out_name = p_genome + '.h_contig_overlap.alleles'\n",
    "\n",
    "best_hits_t_contig_and_h_contig.loc[:, ['Query', 'Target']].to_csv(os.path.join(ALLELE_PATH, out_name), sep='\\t', index=None, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now drop everything that was already captured\n",
    "query_ids_to_drop = best_hits_t_contig_and_h_contig['Query'].unique().tolist()\n",
    "target_ids_to_drop = best_hits_t_contig_and_h_contig['Target'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#get all queries that are on the respective haplotig but have not been caught before\n",
    "_tmp_groupbed = _tmp_df[(~_tmp_df['Query'].isin(query_ids_to_drop))& (_tmp_df['q_contig == t_contig'] == True)].groupby('Query')\n",
    "best_hits_same_contig_no_overlap = _tmp_groupbed.apply(lambda g: g[(g['e-value'] == g['e-value'].min())]) \n",
    "out_name = p_genome + '.no_specific_h_contig_overlap.alleles'\n",
    "best_hits_same_contig_no_overlap.loc[:, ['Query', 'Target']].to_csv(os.path.join(ALLELE_PATH, out_name), sep='\\t', index=None, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now drop everything that was already captured\n",
    "query_ids_to_drop = query_ids_to_drop + best_hits_same_contig_no_overlap['Query'].unique().tolist()\n",
    "target_ids_to_drop = target_ids_to_drop + best_hits_same_contig_no_overlap['Target'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#get all else. Meaning not on the corresponding haplotig but other haplotig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_tmp_groupbed = _tmp_df[(~_tmp_df['Query'].isin(query_ids_to_drop))& (_tmp_df['q_contig == t_contig'] == False) & (_tmp_df['Target'] != 'NaN' )].groupby('Query')\n",
    "best_hits_different_contig = _tmp_groupbed.apply(lambda g: g[(g['e-value'] == g['e-value'].min()) ])\n",
    "out_name = p_genome + '.no_respective_h_contig_overlap.alleles'\n",
    "best_hits_different_contig.loc[:, ['Query', 'Target']].to_csv(os.path.join(ALLELE_PATH, out_name), sep='\\t', index=None, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now drop everything that was already captured\n",
    "query_ids_to_drop = query_ids_to_drop + best_hits_different_contig['Query'].unique().tolist()\n",
    "target_ids_to_drop = target_ids_to_drop + best_hits_different_contig['Target'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#no print out the p_proteins without alleles\n",
    "out_name = p_genome + '.no.alleles'\n",
    "_tmp_df[~_tmp_df['Query'].isin(query_ids_to_drop)].drop_duplicates(subset='Query')['Query'].to_csv(os.path.join(ALLELE_PATH, out_name), sep='\\t', index=None, header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Done till here today"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_tmp_grouped = _tmp_df[_tmp_df['t_contig == h_contig_overlap'] == True].groupby('Query')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "filter function returned a Series, but expected a scalar bool",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-131-71e16bfe47e1>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mbest_hits_t_contig_and_h_contig\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_tmp_grouped\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfilter\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mlambda\u001b[0m \u001b[0mg\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mg\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'e-value'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m>=\u001b[0m \u001b[0mg\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'e-value'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/home/benjamin/anaconda3/lib/python3.5/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36mfilter\u001b[1;34m(self, func, dropna, *args, **kwargs)\u001b[0m\n\u001b[0;32m   3577\u001b[0m                         \u001b[1;31m# assign the name to this series\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3578\u001b[0m                         \u001b[1;32mif\u001b[0m \u001b[0msingular_series\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3579\u001b[1;33m                             \u001b[0mvalues\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mname\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mkeys\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3580\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3581\u001b[0m                             \u001b[1;31m# GH2893\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: filter function returned a Series, but expected a scalar bool"
     ]
    }
   ],
   "source": [
    "best_hits_t_contig_and_h_contig = _tmp_grouped.gr .filter(lambda g: g['e-value'] >= g['e-value'].min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>Target</th>\n",
       "      <th>PctID</th>\n",
       "      <th>AlnLgth</th>\n",
       "      <th>NumMis</th>\n",
       "      <th>NumGap</th>\n",
       "      <th>StartQuery</th>\n",
       "      <th>StopQuery</th>\n",
       "      <th>StartTarget</th>\n",
       "      <th>StopTarget</th>\n",
       "      <th>e-value</th>\n",
       "      <th>BitScore</th>\n",
       "      <th>QLgth</th>\n",
       "      <th>QCov</th>\n",
       "      <th>q_contig</th>\n",
       "      <th>t_contig</th>\n",
       "      <th>q_contig == t_contig</th>\n",
       "      <th>p_protein</th>\n",
       "      <th>h_contig_overlap</th>\n",
       "      <th>t_contig == h_contig_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_031.5</td>\n",
       "      <td>70.19</td>\n",
       "      <td>161</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>1</td>\n",
       "      <td>161</td>\n",
       "      <td>4e-69</td>\n",
       "      <td>207</td>\n",
       "      <td>114</td>\n",
       "      <td>141.228</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>[hcontig_000_003, hcontig_000_031]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.107</td>\n",
       "      <td>99.21</td>\n",
       "      <td>1141</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>0</td>\n",
       "      <td>2330</td>\n",
       "      <td>1144</td>\n",
       "      <td>99.7378</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>evm.model.pcontig_000.101</td>\n",
       "      <td>evm.model.hcontig_000_003.108</td>\n",
       "      <td>89.03</td>\n",
       "      <td>237</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>231</td>\n",
       "      <td>1</td>\n",
       "      <td>237</td>\n",
       "      <td>2e-150</td>\n",
       "      <td>421</td>\n",
       "      <td>231</td>\n",
       "      <td>102.597</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.101</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>evm.model.pcontig_000.102</td>\n",
       "      <td>evm.model.hcontig_000_003.109</td>\n",
       "      <td>100</td>\n",
       "      <td>107</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "      <td>7e-73</td>\n",
       "      <td>214</td>\n",
       "      <td>107</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.102</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>evm.model.pcontig_000.103</td>\n",
       "      <td>evm.model.hcontig_000_003.110</td>\n",
       "      <td>97.73</td>\n",
       "      <td>88</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>2e-57</td>\n",
       "      <td>174</td>\n",
       "      <td>95</td>\n",
       "      <td>92.6316</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.103</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>evm.model.pcontig_000.104</td>\n",
       "      <td>evm.model.hcontig_000_003.111</td>\n",
       "      <td>98.53</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>3e-43</td>\n",
       "      <td>136</td>\n",
       "      <td>68</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.104</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>evm.model.pcontig_000.105</td>\n",
       "      <td>evm.model.hcontig_000_003.112</td>\n",
       "      <td>100</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1e-85</td>\n",
       "      <td>248</td>\n",
       "      <td>122</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.105</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>evm.model.pcontig_000.108</td>\n",
       "      <td>evm.model.hcontig_000_003.117</td>\n",
       "      <td>99.27</td>\n",
       "      <td>682</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>0</td>\n",
       "      <td>1392</td>\n",
       "      <td>682</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.108</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>evm.model.pcontig_000.109</td>\n",
       "      <td>evm.model.hcontig_000_003.117</td>\n",
       "      <td>99.27</td>\n",
       "      <td>682</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>0</td>\n",
       "      <td>1392</td>\n",
       "      <td>682</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.109</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>evm.model.pcontig_000.11</td>\n",
       "      <td>evm.model.hcontig_000_003.12</td>\n",
       "      <td>100</td>\n",
       "      <td>543</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>543</td>\n",
       "      <td>1</td>\n",
       "      <td>543</td>\n",
       "      <td>0</td>\n",
       "      <td>1109</td>\n",
       "      <td>543</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.11</td>\n",
       "      <td>[hcontig_000_003, hcontig_000_031]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>evm.model.pcontig_000.11</td>\n",
       "      <td>evm.model.hcontig_000_031.7</td>\n",
       "      <td>100</td>\n",
       "      <td>446</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>543</td>\n",
       "      <td>1</td>\n",
       "      <td>446</td>\n",
       "      <td>0</td>\n",
       "      <td>907</td>\n",
       "      <td>543</td>\n",
       "      <td>82.1363</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.11</td>\n",
       "      <td>[hcontig_000_003, hcontig_000_031]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>evm.model.pcontig_000.110</td>\n",
       "      <td>evm.model.hcontig_000_003.118</td>\n",
       "      <td>99.1</td>\n",
       "      <td>446</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>446</td>\n",
       "      <td>1</td>\n",
       "      <td>446</td>\n",
       "      <td>0</td>\n",
       "      <td>895</td>\n",
       "      <td>446</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.110</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>evm.model.pcontig_000.111</td>\n",
       "      <td>evm.model.hcontig_000_003.119</td>\n",
       "      <td>98.72</td>\n",
       "      <td>467</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>467</td>\n",
       "      <td>1</td>\n",
       "      <td>467</td>\n",
       "      <td>0</td>\n",
       "      <td>949</td>\n",
       "      <td>467</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.111</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>evm.model.pcontig_000.112</td>\n",
       "      <td>evm.model.hcontig_000_003.120</td>\n",
       "      <td>100</td>\n",
       "      <td>232</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>232</td>\n",
       "      <td>1</td>\n",
       "      <td>232</td>\n",
       "      <td>1e-166</td>\n",
       "      <td>462</td>\n",
       "      <td>232</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.112</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>evm.model.pcontig_000.114</td>\n",
       "      <td>evm.model.hcontig_000_003.121</td>\n",
       "      <td>100</td>\n",
       "      <td>66</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>3e-40</td>\n",
       "      <td>138</td>\n",
       "      <td>92</td>\n",
       "      <td>71.7391</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.114</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>evm.model.pcontig_000.115</td>\n",
       "      <td>evm.model.hcontig_000_003.121</td>\n",
       "      <td>99.01</td>\n",
       "      <td>202</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>202</td>\n",
       "      <td>223</td>\n",
       "      <td>424</td>\n",
       "      <td>3e-145</td>\n",
       "      <td>414</td>\n",
       "      <td>202</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.115</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>evm.model.pcontig_000.116</td>\n",
       "      <td>evm.model.hcontig_000_003.122</td>\n",
       "      <td>97.28</td>\n",
       "      <td>404</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>404</td>\n",
       "      <td>1</td>\n",
       "      <td>402</td>\n",
       "      <td>0</td>\n",
       "      <td>789</td>\n",
       "      <td>404</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.116</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>evm.model.pcontig_000.117</td>\n",
       "      <td>evm.model.hcontig_000_003.123</td>\n",
       "      <td>99.72</td>\n",
       "      <td>360</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>0</td>\n",
       "      <td>739</td>\n",
       "      <td>360</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.117</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>evm.model.pcontig_000.12</td>\n",
       "      <td>evm.model.hcontig_000_003.13</td>\n",
       "      <td>97.85</td>\n",
       "      <td>326</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>326</td>\n",
       "      <td>1</td>\n",
       "      <td>326</td>\n",
       "      <td>0</td>\n",
       "      <td>644</td>\n",
       "      <td>326</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.12</td>\n",
       "      <td>[hcontig_000_003, hcontig_000_031]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>evm.model.pcontig_000.13</td>\n",
       "      <td>evm.model.hcontig_000_003.14</td>\n",
       "      <td>99.77</td>\n",
       "      <td>432</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>432</td>\n",
       "      <td>1</td>\n",
       "      <td>432</td>\n",
       "      <td>0</td>\n",
       "      <td>893</td>\n",
       "      <td>432</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.13</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>evm.model.pcontig_000.132</td>\n",
       "      <td>evm.model.hcontig_000_003.146</td>\n",
       "      <td>99.89</td>\n",
       "      <td>947</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>947</td>\n",
       "      <td>1</td>\n",
       "      <td>947</td>\n",
       "      <td>0</td>\n",
       "      <td>1919</td>\n",
       "      <td>947</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.132</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>evm.model.pcontig_000.133</td>\n",
       "      <td>evm.model.hcontig_000_003.147</td>\n",
       "      <td>100</td>\n",
       "      <td>325</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>325</td>\n",
       "      <td>1</td>\n",
       "      <td>325</td>\n",
       "      <td>0</td>\n",
       "      <td>665</td>\n",
       "      <td>325</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.133</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>evm.model.pcontig_000.134</td>\n",
       "      <td>evm.model.hcontig_000_003.148</td>\n",
       "      <td>99.77</td>\n",
       "      <td>1314</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1314</td>\n",
       "      <td>62</td>\n",
       "      <td>1375</td>\n",
       "      <td>0</td>\n",
       "      <td>2697</td>\n",
       "      <td>1314</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.134</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>evm.model.pcontig_000.135</td>\n",
       "      <td>evm.model.hcontig_000_003.149</td>\n",
       "      <td>99.72</td>\n",
       "      <td>356</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>356</td>\n",
       "      <td>1</td>\n",
       "      <td>356</td>\n",
       "      <td>0</td>\n",
       "      <td>720</td>\n",
       "      <td>356</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.135</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>evm.model.pcontig_000.136</td>\n",
       "      <td>evm.model.hcontig_000_003.150</td>\n",
       "      <td>97.24</td>\n",
       "      <td>435</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>436</td>\n",
       "      <td>1</td>\n",
       "      <td>435</td>\n",
       "      <td>0</td>\n",
       "      <td>870</td>\n",
       "      <td>440</td>\n",
       "      <td>98.8636</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.136</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>evm.model.pcontig_000.137</td>\n",
       "      <td>evm.model.hcontig_000_003.151</td>\n",
       "      <td>99.64</td>\n",
       "      <td>276</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>276</td>\n",
       "      <td>1</td>\n",
       "      <td>276</td>\n",
       "      <td>0</td>\n",
       "      <td>570</td>\n",
       "      <td>276</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.137</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>evm.model.pcontig_000.138</td>\n",
       "      <td>evm.model.hcontig_000_003.152</td>\n",
       "      <td>99.76</td>\n",
       "      <td>820</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>820</td>\n",
       "      <td>1</td>\n",
       "      <td>820</td>\n",
       "      <td>0</td>\n",
       "      <td>1701</td>\n",
       "      <td>820</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.138</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>evm.model.pcontig_000.138</td>\n",
       "      <td>evm.model.hcontig_000_003.155</td>\n",
       "      <td>61.7</td>\n",
       "      <td>752</td>\n",
       "      <td>237</td>\n",
       "      <td>7</td>\n",
       "      <td>56</td>\n",
       "      <td>767</td>\n",
       "      <td>2</td>\n",
       "      <td>742</td>\n",
       "      <td>0</td>\n",
       "      <td>905</td>\n",
       "      <td>820</td>\n",
       "      <td>91.7073</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.138</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>evm.model.pcontig_000.139</td>\n",
       "      <td>evm.model.hcontig_000_003.153</td>\n",
       "      <td>98.61</td>\n",
       "      <td>577</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>574</td>\n",
       "      <td>1</td>\n",
       "      <td>577</td>\n",
       "      <td>0</td>\n",
       "      <td>1170</td>\n",
       "      <td>574</td>\n",
       "      <td>100.523</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.139</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>evm.model.pcontig_000.14</td>\n",
       "      <td>evm.model.hcontig_000_003.15</td>\n",
       "      <td>99.75</td>\n",
       "      <td>408</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>408</td>\n",
       "      <td>1</td>\n",
       "      <td>408</td>\n",
       "      <td>0</td>\n",
       "      <td>847</td>\n",
       "      <td>408</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_000.14</td>\n",
       "      <td>[hcontig_000_003]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12112</th>\n",
       "      <td>evm.model.pcontig_147.10</td>\n",
       "      <td>evm.model.hcontig_147_001.3</td>\n",
       "      <td>99.76</td>\n",
       "      <td>425</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>425</td>\n",
       "      <td>1</td>\n",
       "      <td>425</td>\n",
       "      <td>0</td>\n",
       "      <td>888</td>\n",
       "      <td>425</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_147</td>\n",
       "      <td>hcontig_147_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_147.10</td>\n",
       "      <td>[hcontig_147_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12113</th>\n",
       "      <td>evm.model.pcontig_147.11</td>\n",
       "      <td>evm.model.hcontig_147_001.8</td>\n",
       "      <td>90.91</td>\n",
       "      <td>66</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>156</td>\n",
       "      <td>221</td>\n",
       "      <td>6e-37</td>\n",
       "      <td>127</td>\n",
       "      <td>76</td>\n",
       "      <td>86.8421</td>\n",
       "      <td>pcontig_147</td>\n",
       "      <td>hcontig_147_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_147.11</td>\n",
       "      <td>[hcontig_147_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12114</th>\n",
       "      <td>evm.model.pcontig_147.12</td>\n",
       "      <td>evm.model.hcontig_147_001.4</td>\n",
       "      <td>78.75</td>\n",
       "      <td>367</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>329</td>\n",
       "      <td>1</td>\n",
       "      <td>367</td>\n",
       "      <td>7e-177</td>\n",
       "      <td>500</td>\n",
       "      <td>330</td>\n",
       "      <td>111.212</td>\n",
       "      <td>pcontig_147</td>\n",
       "      <td>hcontig_147_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_147.12</td>\n",
       "      <td>[hcontig_147_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12115</th>\n",
       "      <td>evm.model.pcontig_147.17</td>\n",
       "      <td>evm.model.hcontig_147_001.8</td>\n",
       "      <td>89.01</td>\n",
       "      <td>273</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>248</td>\n",
       "      <td>1</td>\n",
       "      <td>273</td>\n",
       "      <td>3e-176</td>\n",
       "      <td>493</td>\n",
       "      <td>293</td>\n",
       "      <td>93.1741</td>\n",
       "      <td>pcontig_147</td>\n",
       "      <td>hcontig_147_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_147.17</td>\n",
       "      <td>[hcontig_147_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12116</th>\n",
       "      <td>evm.model.pcontig_147.18</td>\n",
       "      <td>evm.model.hcontig_147_001.9</td>\n",
       "      <td>100</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>89</td>\n",
       "      <td>6</td>\n",
       "      <td>65</td>\n",
       "      <td>4e-37</td>\n",
       "      <td>121</td>\n",
       "      <td>89</td>\n",
       "      <td>67.4157</td>\n",
       "      <td>pcontig_147</td>\n",
       "      <td>hcontig_147_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_147.18</td>\n",
       "      <td>[hcontig_147_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12117</th>\n",
       "      <td>evm.model.pcontig_147.9</td>\n",
       "      <td>evm.model.hcontig_147_001.2</td>\n",
       "      <td>100</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>2e-43</td>\n",
       "      <td>136</td>\n",
       "      <td>68</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_147</td>\n",
       "      <td>hcontig_147_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_147.9</td>\n",
       "      <td>[hcontig_147_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12118</th>\n",
       "      <td>evm.model.pcontig_148.1</td>\n",
       "      <td>evm.model.hcontig_148_001.2</td>\n",
       "      <td>100</td>\n",
       "      <td>855</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>855</td>\n",
       "      <td>1</td>\n",
       "      <td>855</td>\n",
       "      <td>0</td>\n",
       "      <td>1770</td>\n",
       "      <td>855</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_148</td>\n",
       "      <td>hcontig_148_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_148.1</td>\n",
       "      <td>[hcontig_148_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12119</th>\n",
       "      <td>evm.model.pcontig_148.2</td>\n",
       "      <td>evm.model.hcontig_148_001.3</td>\n",
       "      <td>100</td>\n",
       "      <td>128</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>128</td>\n",
       "      <td>1</td>\n",
       "      <td>128</td>\n",
       "      <td>4e-89</td>\n",
       "      <td>257</td>\n",
       "      <td>128</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_148</td>\n",
       "      <td>hcontig_148_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_148.2</td>\n",
       "      <td>[hcontig_148_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12120</th>\n",
       "      <td>evm.model.pcontig_148.3</td>\n",
       "      <td>evm.model.hcontig_148_001.4</td>\n",
       "      <td>51.88</td>\n",
       "      <td>133</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>133</td>\n",
       "      <td>6e-34</td>\n",
       "      <td>123</td>\n",
       "      <td>337</td>\n",
       "      <td>39.4659</td>\n",
       "      <td>pcontig_148</td>\n",
       "      <td>hcontig_148_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_148.3</td>\n",
       "      <td>[hcontig_148_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12121</th>\n",
       "      <td>evm.model.pcontig_148.4</td>\n",
       "      <td>evm.model.hcontig_148_001.5</td>\n",
       "      <td>97.44</td>\n",
       "      <td>195</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>195</td>\n",
       "      <td>1</td>\n",
       "      <td>195</td>\n",
       "      <td>1e-138</td>\n",
       "      <td>394</td>\n",
       "      <td>203</td>\n",
       "      <td>96.0591</td>\n",
       "      <td>pcontig_148</td>\n",
       "      <td>hcontig_148_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_148.4</td>\n",
       "      <td>[hcontig_148_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12122</th>\n",
       "      <td>evm.model.pcontig_166.10</td>\n",
       "      <td>evm.model.hcontig_166_001.4</td>\n",
       "      <td>100</td>\n",
       "      <td>927</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>927</td>\n",
       "      <td>1</td>\n",
       "      <td>927</td>\n",
       "      <td>0</td>\n",
       "      <td>1922</td>\n",
       "      <td>927</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_166</td>\n",
       "      <td>hcontig_166_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_166.10</td>\n",
       "      <td>[hcontig_166_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12123</th>\n",
       "      <td>evm.model.pcontig_166.11</td>\n",
       "      <td>evm.model.hcontig_166_001.5</td>\n",
       "      <td>99.58</td>\n",
       "      <td>237</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>256</td>\n",
       "      <td>3</td>\n",
       "      <td>239</td>\n",
       "      <td>1e-175</td>\n",
       "      <td>486</td>\n",
       "      <td>256</td>\n",
       "      <td>92.5781</td>\n",
       "      <td>pcontig_166</td>\n",
       "      <td>hcontig_166_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_166.11</td>\n",
       "      <td>[hcontig_166_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12124</th>\n",
       "      <td>evm.model.pcontig_166.6</td>\n",
       "      <td>evm.model.hcontig_166_001.1</td>\n",
       "      <td>100</td>\n",
       "      <td>315</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>315</td>\n",
       "      <td>1</td>\n",
       "      <td>315</td>\n",
       "      <td>0</td>\n",
       "      <td>643</td>\n",
       "      <td>1212</td>\n",
       "      <td>25.9901</td>\n",
       "      <td>pcontig_166</td>\n",
       "      <td>hcontig_166_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_166.6</td>\n",
       "      <td>[hcontig_166_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12125</th>\n",
       "      <td>evm.model.pcontig_166.7</td>\n",
       "      <td>evm.model.hcontig_166_001.2</td>\n",
       "      <td>100</td>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "      <td>307</td>\n",
       "      <td>410</td>\n",
       "      <td>6e-69</td>\n",
       "      <td>214</td>\n",
       "      <td>104</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_166</td>\n",
       "      <td>hcontig_166_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_166.7</td>\n",
       "      <td>[hcontig_166_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12126</th>\n",
       "      <td>evm.model.pcontig_166.8</td>\n",
       "      <td>evm.model.hcontig_166_001.2</td>\n",
       "      <td>96.24</td>\n",
       "      <td>266</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>266</td>\n",
       "      <td>1</td>\n",
       "      <td>264</td>\n",
       "      <td>0</td>\n",
       "      <td>526</td>\n",
       "      <td>299</td>\n",
       "      <td>88.9632</td>\n",
       "      <td>pcontig_166</td>\n",
       "      <td>hcontig_166_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_166.8</td>\n",
       "      <td>[hcontig_166_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12127</th>\n",
       "      <td>evm.model.pcontig_189.1</td>\n",
       "      <td>evm.model.hcontig_189_001.3</td>\n",
       "      <td>96.1</td>\n",
       "      <td>154</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>151</td>\n",
       "      <td>229</td>\n",
       "      <td>382</td>\n",
       "      <td>3e-102</td>\n",
       "      <td>307</td>\n",
       "      <td>170</td>\n",
       "      <td>90.5882</td>\n",
       "      <td>pcontig_189</td>\n",
       "      <td>hcontig_189_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_189.1</td>\n",
       "      <td>[hcontig_189_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12128</th>\n",
       "      <td>evm.model.pcontig_189.2</td>\n",
       "      <td>evm.model.hcontig_189_001.4</td>\n",
       "      <td>100</td>\n",
       "      <td>156</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>156</td>\n",
       "      <td>63</td>\n",
       "      <td>218</td>\n",
       "      <td>2e-111</td>\n",
       "      <td>318</td>\n",
       "      <td>156</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_189</td>\n",
       "      <td>hcontig_189_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_189.2</td>\n",
       "      <td>[hcontig_189_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12129</th>\n",
       "      <td>evm.model.pcontig_189.3</td>\n",
       "      <td>evm.model.hcontig_189_001.5</td>\n",
       "      <td>98.36</td>\n",
       "      <td>122</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>461</td>\n",
       "      <td>582</td>\n",
       "      <td>6e-78</td>\n",
       "      <td>245</td>\n",
       "      <td>158</td>\n",
       "      <td>77.2152</td>\n",
       "      <td>pcontig_189</td>\n",
       "      <td>hcontig_189_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_189.3</td>\n",
       "      <td>[hcontig_189_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12130</th>\n",
       "      <td>evm.model.pcontig_189.4</td>\n",
       "      <td>evm.model.hcontig_189_001.5</td>\n",
       "      <td>99.11</td>\n",
       "      <td>338</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>338</td>\n",
       "      <td>1</td>\n",
       "      <td>338</td>\n",
       "      <td>0</td>\n",
       "      <td>705</td>\n",
       "      <td>401</td>\n",
       "      <td>84.2893</td>\n",
       "      <td>pcontig_189</td>\n",
       "      <td>hcontig_189_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_189.4</td>\n",
       "      <td>[hcontig_189_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12131</th>\n",
       "      <td>evm.model.pcontig_189.5</td>\n",
       "      <td>evm.model.hcontig_189_001.6</td>\n",
       "      <td>69.63</td>\n",
       "      <td>1057</td>\n",
       "      <td>199</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1038</td>\n",
       "      <td>1</td>\n",
       "      <td>954</td>\n",
       "      <td>0</td>\n",
       "      <td>1273</td>\n",
       "      <td>1337</td>\n",
       "      <td>79.0576</td>\n",
       "      <td>pcontig_189</td>\n",
       "      <td>hcontig_189_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_189.5</td>\n",
       "      <td>[hcontig_189_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12132</th>\n",
       "      <td>evm.model.pcontig_189.6</td>\n",
       "      <td>evm.model.hcontig_189_001.7</td>\n",
       "      <td>100</td>\n",
       "      <td>659</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1163</td>\n",
       "      <td>1821</td>\n",
       "      <td>1</td>\n",
       "      <td>659</td>\n",
       "      <td>0</td>\n",
       "      <td>1360</td>\n",
       "      <td>1821</td>\n",
       "      <td>36.1889</td>\n",
       "      <td>pcontig_189</td>\n",
       "      <td>hcontig_189_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_189.6</td>\n",
       "      <td>[hcontig_189_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12133</th>\n",
       "      <td>evm.model.pcontig_225.1</td>\n",
       "      <td>evm.model.hcontig_225_001.3</td>\n",
       "      <td>90.74</td>\n",
       "      <td>54</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>151</td>\n",
       "      <td>204</td>\n",
       "      <td>2e-28</td>\n",
       "      <td>103</td>\n",
       "      <td>58</td>\n",
       "      <td>93.1034</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.1</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12134</th>\n",
       "      <td>evm.model.pcontig_225.2</td>\n",
       "      <td>evm.model.hcontig_225_001.3</td>\n",
       "      <td>98.11</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>53</td>\n",
       "      <td>265</td>\n",
       "      <td>317</td>\n",
       "      <td>1e-29</td>\n",
       "      <td>106</td>\n",
       "      <td>53</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.2</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12135</th>\n",
       "      <td>evm.model.pcontig_225.3</td>\n",
       "      <td>evm.model.hcontig_225_001.4</td>\n",
       "      <td>100</td>\n",
       "      <td>376</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>376</td>\n",
       "      <td>37</td>\n",
       "      <td>412</td>\n",
       "      <td>0</td>\n",
       "      <td>776</td>\n",
       "      <td>376</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.3</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12136</th>\n",
       "      <td>evm.model.pcontig_225.4</td>\n",
       "      <td>evm.model.hcontig_225_001.5</td>\n",
       "      <td>100</td>\n",
       "      <td>566</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>566</td>\n",
       "      <td>1</td>\n",
       "      <td>566</td>\n",
       "      <td>0</td>\n",
       "      <td>1181</td>\n",
       "      <td>566</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.4</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12137</th>\n",
       "      <td>evm.model.pcontig_225.5</td>\n",
       "      <td>evm.model.hcontig_225_001.6</td>\n",
       "      <td>99.39</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>163</td>\n",
       "      <td>6e-120</td>\n",
       "      <td>343</td>\n",
       "      <td>292</td>\n",
       "      <td>55.8219</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.5</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12138</th>\n",
       "      <td>evm.model.pcontig_225.6</td>\n",
       "      <td>evm.model.hcontig_225_001.7</td>\n",
       "      <td>99.87</td>\n",
       "      <td>759</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>759</td>\n",
       "      <td>1</td>\n",
       "      <td>759</td>\n",
       "      <td>0</td>\n",
       "      <td>1561</td>\n",
       "      <td>759</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.6</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12139</th>\n",
       "      <td>evm.model.pcontig_225.7</td>\n",
       "      <td>evm.model.hcontig_225_001.8</td>\n",
       "      <td>100</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>119</td>\n",
       "      <td>1</td>\n",
       "      <td>119</td>\n",
       "      <td>1e-83</td>\n",
       "      <td>243</td>\n",
       "      <td>119</td>\n",
       "      <td>100</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.7</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12140</th>\n",
       "      <td>evm.model.pcontig_225.8</td>\n",
       "      <td>evm.model.hcontig_225_001.9</td>\n",
       "      <td>98.51</td>\n",
       "      <td>134</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>134</td>\n",
       "      <td>1</td>\n",
       "      <td>134</td>\n",
       "      <td>5e-84</td>\n",
       "      <td>257</td>\n",
       "      <td>140</td>\n",
       "      <td>95.7143</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.8</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12141</th>\n",
       "      <td>evm.model.pcontig_225.9</td>\n",
       "      <td>evm.model.hcontig_225_001.9</td>\n",
       "      <td>100</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>206</td>\n",
       "      <td>283</td>\n",
       "      <td>488</td>\n",
       "      <td>9e-148</td>\n",
       "      <td>439</td>\n",
       "      <td>613</td>\n",
       "      <td>33.6052</td>\n",
       "      <td>pcontig_225</td>\n",
       "      <td>hcontig_225_001</td>\n",
       "      <td>True</td>\n",
       "      <td>evm.model.pcontig_225.9</td>\n",
       "      <td>[hcontig_225_001]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12142 rows  20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Query                         Target  PctID  \\\n",
       "0       evm.model.pcontig_000.10    evm.model.hcontig_000_031.5  70.19   \n",
       "1      evm.model.pcontig_000.100  evm.model.hcontig_000_003.107  99.21   \n",
       "2      evm.model.pcontig_000.101  evm.model.hcontig_000_003.108  89.03   \n",
       "3      evm.model.pcontig_000.102  evm.model.hcontig_000_003.109    100   \n",
       "4      evm.model.pcontig_000.103  evm.model.hcontig_000_003.110  97.73   \n",
       "5      evm.model.pcontig_000.104  evm.model.hcontig_000_003.111  98.53   \n",
       "6      evm.model.pcontig_000.105  evm.model.hcontig_000_003.112    100   \n",
       "7      evm.model.pcontig_000.108  evm.model.hcontig_000_003.117  99.27   \n",
       "8      evm.model.pcontig_000.109  evm.model.hcontig_000_003.117  99.27   \n",
       "9       evm.model.pcontig_000.11   evm.model.hcontig_000_003.12    100   \n",
       "10      evm.model.pcontig_000.11    evm.model.hcontig_000_031.7    100   \n",
       "11     evm.model.pcontig_000.110  evm.model.hcontig_000_003.118   99.1   \n",
       "12     evm.model.pcontig_000.111  evm.model.hcontig_000_003.119  98.72   \n",
       "13     evm.model.pcontig_000.112  evm.model.hcontig_000_003.120    100   \n",
       "14     evm.model.pcontig_000.114  evm.model.hcontig_000_003.121    100   \n",
       "15     evm.model.pcontig_000.115  evm.model.hcontig_000_003.121  99.01   \n",
       "16     evm.model.pcontig_000.116  evm.model.hcontig_000_003.122  97.28   \n",
       "17     evm.model.pcontig_000.117  evm.model.hcontig_000_003.123  99.72   \n",
       "18      evm.model.pcontig_000.12   evm.model.hcontig_000_003.13  97.85   \n",
       "19      evm.model.pcontig_000.13   evm.model.hcontig_000_003.14  99.77   \n",
       "20     evm.model.pcontig_000.132  evm.model.hcontig_000_003.146  99.89   \n",
       "21     evm.model.pcontig_000.133  evm.model.hcontig_000_003.147    100   \n",
       "22     evm.model.pcontig_000.134  evm.model.hcontig_000_003.148  99.77   \n",
       "23     evm.model.pcontig_000.135  evm.model.hcontig_000_003.149  99.72   \n",
       "24     evm.model.pcontig_000.136  evm.model.hcontig_000_003.150  97.24   \n",
       "25     evm.model.pcontig_000.137  evm.model.hcontig_000_003.151  99.64   \n",
       "26     evm.model.pcontig_000.138  evm.model.hcontig_000_003.152  99.76   \n",
       "27     evm.model.pcontig_000.138  evm.model.hcontig_000_003.155   61.7   \n",
       "28     evm.model.pcontig_000.139  evm.model.hcontig_000_003.153  98.61   \n",
       "29      evm.model.pcontig_000.14   evm.model.hcontig_000_003.15  99.75   \n",
       "...                          ...                            ...    ...   \n",
       "12112   evm.model.pcontig_147.10    evm.model.hcontig_147_001.3  99.76   \n",
       "12113   evm.model.pcontig_147.11    evm.model.hcontig_147_001.8  90.91   \n",
       "12114   evm.model.pcontig_147.12    evm.model.hcontig_147_001.4  78.75   \n",
       "12115   evm.model.pcontig_147.17    evm.model.hcontig_147_001.8  89.01   \n",
       "12116   evm.model.pcontig_147.18    evm.model.hcontig_147_001.9    100   \n",
       "12117    evm.model.pcontig_147.9    evm.model.hcontig_147_001.2    100   \n",
       "12118    evm.model.pcontig_148.1    evm.model.hcontig_148_001.2    100   \n",
       "12119    evm.model.pcontig_148.2    evm.model.hcontig_148_001.3    100   \n",
       "12120    evm.model.pcontig_148.3    evm.model.hcontig_148_001.4  51.88   \n",
       "12121    evm.model.pcontig_148.4    evm.model.hcontig_148_001.5  97.44   \n",
       "12122   evm.model.pcontig_166.10    evm.model.hcontig_166_001.4    100   \n",
       "12123   evm.model.pcontig_166.11    evm.model.hcontig_166_001.5  99.58   \n",
       "12124    evm.model.pcontig_166.6    evm.model.hcontig_166_001.1    100   \n",
       "12125    evm.model.pcontig_166.7    evm.model.hcontig_166_001.2    100   \n",
       "12126    evm.model.pcontig_166.8    evm.model.hcontig_166_001.2  96.24   \n",
       "12127    evm.model.pcontig_189.1    evm.model.hcontig_189_001.3   96.1   \n",
       "12128    evm.model.pcontig_189.2    evm.model.hcontig_189_001.4    100   \n",
       "12129    evm.model.pcontig_189.3    evm.model.hcontig_189_001.5  98.36   \n",
       "12130    evm.model.pcontig_189.4    evm.model.hcontig_189_001.5  99.11   \n",
       "12131    evm.model.pcontig_189.5    evm.model.hcontig_189_001.6  69.63   \n",
       "12132    evm.model.pcontig_189.6    evm.model.hcontig_189_001.7    100   \n",
       "12133    evm.model.pcontig_225.1    evm.model.hcontig_225_001.3  90.74   \n",
       "12134    evm.model.pcontig_225.2    evm.model.hcontig_225_001.3  98.11   \n",
       "12135    evm.model.pcontig_225.3    evm.model.hcontig_225_001.4    100   \n",
       "12136    evm.model.pcontig_225.4    evm.model.hcontig_225_001.5    100   \n",
       "12137    evm.model.pcontig_225.5    evm.model.hcontig_225_001.6  99.39   \n",
       "12138    evm.model.pcontig_225.6    evm.model.hcontig_225_001.7  99.87   \n",
       "12139    evm.model.pcontig_225.7    evm.model.hcontig_225_001.8    100   \n",
       "12140    evm.model.pcontig_225.8    evm.model.hcontig_225_001.9  98.51   \n",
       "12141    evm.model.pcontig_225.9    evm.model.hcontig_225_001.9    100   \n",
       "\n",
       "      AlnLgth NumMis NumGap StartQuery StopQuery StartTarget StopTarget  \\\n",
       "0         161      1      1          1       114           1        161   \n",
       "1        1141      9      0          1      1141           1       1141   \n",
       "2         237     20      1          1       231           1        237   \n",
       "3         107      0      0          1       107           1        107   \n",
       "4          88      2      0          8        95           1         88   \n",
       "5          68      1      0          1        68           1         68   \n",
       "6         122      0      0          1       122           1        122   \n",
       "7         682      5      0          1       682           1        682   \n",
       "8         682      5      0          1       682           1        682   \n",
       "9         543      0      0          1       543           1        543   \n",
       "10        446      0      0         98       543           1        446   \n",
       "11        446      4      0          1       446           1        446   \n",
       "12        467      6      0          1       467           1        467   \n",
       "13        232      0      0          1       232           1        232   \n",
       "14         66      0      0          1        66           1         66   \n",
       "15        202      2      0          1       202         223        424   \n",
       "16        404      9      1          1       404           1        402   \n",
       "17        360      1      0          1       360           1        360   \n",
       "18        326      7      0          1       326           1        326   \n",
       "19        432      1      0          1       432           1        432   \n",
       "20        947      1      0          1       947           1        947   \n",
       "21        325      0      0          1       325           1        325   \n",
       "22       1314      3      0          1      1314          62       1375   \n",
       "23        356      1      0          1       356           1        356   \n",
       "24        435     12      0          2       436           1        435   \n",
       "25        276      1      0          1       276           1        276   \n",
       "26        820      2      0          1       820           1        820   \n",
       "27        752    237      7         56       767           2        742   \n",
       "28        577      5      1          1       574           1        577   \n",
       "29        408      1      0          1       408           1        408   \n",
       "...       ...    ...    ...        ...       ...         ...        ...   \n",
       "12112     425      1      0          1       425           1        425   \n",
       "12113      66      6      0          1        66         156        221   \n",
       "12114     367     40      3          1       329           1        367   \n",
       "12115     273      5      1          1       248           1        273   \n",
       "12116      60      0      0         30        89           6         65   \n",
       "12117      68      0      0          1        68           1         68   \n",
       "12118     855      0      0          1       855           1        855   \n",
       "12119     128      0      0          1       128           1        128   \n",
       "12120     133     64      0         79       211           1        133   \n",
       "12121     195      5      0          1       195           1        195   \n",
       "12122     927      0      0          1       927           1        927   \n",
       "12123     237      1      0         20       256           3        239   \n",
       "12124     315      0      0          1       315           1        315   \n",
       "12125     104      0      0          1       104         307        410   \n",
       "12126     266      8      1          1       266           1        264   \n",
       "12127     154      3      1          1       151         229        382   \n",
       "12128     156      0      0          1       156          63        218   \n",
       "12129     122      2      0          1       122         461        582   \n",
       "12130     338      3      0          1       338           1        338   \n",
       "12131    1057    199     14          1      1038           1        954   \n",
       "12132     659      0      0       1163      1821           1        659   \n",
       "12133      54      5      0          2        55         151        204   \n",
       "12134      53      1      0          1        53         265        317   \n",
       "12135     376      0      0          1       376          37        412   \n",
       "12136     566      0      0          1       566           1        566   \n",
       "12137     163      1      0          1       163           1        163   \n",
       "12138     759      1      0          1       759           1        759   \n",
       "12139     119      0      0          1       119           1        119   \n",
       "12140     134      2      0          1       134           1        134   \n",
       "12141     206      0      0          1       206         283        488   \n",
       "\n",
       "      e-value BitScore  QLgth     QCov     q_contig         t_contig  \\\n",
       "0       4e-69      207    114  141.228  pcontig_000  hcontig_000_031   \n",
       "1           0     2330   1144  99.7378  pcontig_000  hcontig_000_003   \n",
       "2      2e-150      421    231  102.597  pcontig_000  hcontig_000_003   \n",
       "3       7e-73      214    107      100  pcontig_000  hcontig_000_003   \n",
       "4       2e-57      174     95  92.6316  pcontig_000  hcontig_000_003   \n",
       "5       3e-43      136     68      100  pcontig_000  hcontig_000_003   \n",
       "6       1e-85      248    122      100  pcontig_000  hcontig_000_003   \n",
       "7           0     1392    682      100  pcontig_000  hcontig_000_003   \n",
       "8           0     1392    682      100  pcontig_000  hcontig_000_003   \n",
       "9           0     1109    543      100  pcontig_000  hcontig_000_003   \n",
       "10          0      907    543  82.1363  pcontig_000  hcontig_000_031   \n",
       "11          0      895    446      100  pcontig_000  hcontig_000_003   \n",
       "12          0      949    467      100  pcontig_000  hcontig_000_003   \n",
       "13     1e-166      462    232      100  pcontig_000  hcontig_000_003   \n",
       "14      3e-40      138     92  71.7391  pcontig_000  hcontig_000_003   \n",
       "15     3e-145      414    202      100  pcontig_000  hcontig_000_003   \n",
       "16          0      789    404      100  pcontig_000  hcontig_000_003   \n",
       "17          0      739    360      100  pcontig_000  hcontig_000_003   \n",
       "18          0      644    326      100  pcontig_000  hcontig_000_003   \n",
       "19          0      893    432      100  pcontig_000  hcontig_000_003   \n",
       "20          0     1919    947      100  pcontig_000  hcontig_000_003   \n",
       "21          0      665    325      100  pcontig_000  hcontig_000_003   \n",
       "22          0     2697   1314      100  pcontig_000  hcontig_000_003   \n",
       "23          0      720    356      100  pcontig_000  hcontig_000_003   \n",
       "24          0      870    440  98.8636  pcontig_000  hcontig_000_003   \n",
       "25          0      570    276      100  pcontig_000  hcontig_000_003   \n",
       "26          0     1701    820      100  pcontig_000  hcontig_000_003   \n",
       "27          0      905    820  91.7073  pcontig_000  hcontig_000_003   \n",
       "28          0     1170    574  100.523  pcontig_000  hcontig_000_003   \n",
       "29          0      847    408      100  pcontig_000  hcontig_000_003   \n",
       "...       ...      ...    ...      ...          ...              ...   \n",
       "12112       0      888    425      100  pcontig_147  hcontig_147_001   \n",
       "12113   6e-37      127     76  86.8421  pcontig_147  hcontig_147_001   \n",
       "12114  7e-177      500    330  111.212  pcontig_147  hcontig_147_001   \n",
       "12115  3e-176      493    293  93.1741  pcontig_147  hcontig_147_001   \n",
       "12116   4e-37      121     89  67.4157  pcontig_147  hcontig_147_001   \n",
       "12117   2e-43      136     68      100  pcontig_147  hcontig_147_001   \n",
       "12118       0     1770    855      100  pcontig_148  hcontig_148_001   \n",
       "12119   4e-89      257    128      100  pcontig_148  hcontig_148_001   \n",
       "12120   6e-34      123    337  39.4659  pcontig_148  hcontig_148_001   \n",
       "12121  1e-138      394    203  96.0591  pcontig_148  hcontig_148_001   \n",
       "12122       0     1922    927      100  pcontig_166  hcontig_166_001   \n",
       "12123  1e-175      486    256  92.5781  pcontig_166  hcontig_166_001   \n",
       "12124       0      643   1212  25.9901  pcontig_166  hcontig_166_001   \n",
       "12125   6e-69      214    104      100  pcontig_166  hcontig_166_001   \n",
       "12126       0      526    299  88.9632  pcontig_166  hcontig_166_001   \n",
       "12127  3e-102      307    170  90.5882  pcontig_189  hcontig_189_001   \n",
       "12128  2e-111      318    156      100  pcontig_189  hcontig_189_001   \n",
       "12129   6e-78      245    158  77.2152  pcontig_189  hcontig_189_001   \n",
       "12130       0      705    401  84.2893  pcontig_189  hcontig_189_001   \n",
       "12131       0     1273   1337  79.0576  pcontig_189  hcontig_189_001   \n",
       "12132       0     1360   1821  36.1889  pcontig_189  hcontig_189_001   \n",
       "12133   2e-28      103     58  93.1034  pcontig_225  hcontig_225_001   \n",
       "12134   1e-29      106     53      100  pcontig_225  hcontig_225_001   \n",
       "12135       0      776    376      100  pcontig_225  hcontig_225_001   \n",
       "12136       0     1181    566      100  pcontig_225  hcontig_225_001   \n",
       "12137  6e-120      343    292  55.8219  pcontig_225  hcontig_225_001   \n",
       "12138       0     1561    759      100  pcontig_225  hcontig_225_001   \n",
       "12139   1e-83      243    119      100  pcontig_225  hcontig_225_001   \n",
       "12140   5e-84      257    140  95.7143  pcontig_225  hcontig_225_001   \n",
       "12141  9e-148      439    613  33.6052  pcontig_225  hcontig_225_001   \n",
       "\n",
       "      q_contig == t_contig                  p_protein  \\\n",
       "0                     True   evm.model.pcontig_000.10   \n",
       "1                     True  evm.model.pcontig_000.100   \n",
       "2                     True  evm.model.pcontig_000.101   \n",
       "3                     True  evm.model.pcontig_000.102   \n",
       "4                     True  evm.model.pcontig_000.103   \n",
       "5                     True  evm.model.pcontig_000.104   \n",
       "6                     True  evm.model.pcontig_000.105   \n",
       "7                     True  evm.model.pcontig_000.108   \n",
       "8                     True  evm.model.pcontig_000.109   \n",
       "9                     True   evm.model.pcontig_000.11   \n",
       "10                    True   evm.model.pcontig_000.11   \n",
       "11                    True  evm.model.pcontig_000.110   \n",
       "12                    True  evm.model.pcontig_000.111   \n",
       "13                    True  evm.model.pcontig_000.112   \n",
       "14                    True  evm.model.pcontig_000.114   \n",
       "15                    True  evm.model.pcontig_000.115   \n",
       "16                    True  evm.model.pcontig_000.116   \n",
       "17                    True  evm.model.pcontig_000.117   \n",
       "18                    True   evm.model.pcontig_000.12   \n",
       "19                    True   evm.model.pcontig_000.13   \n",
       "20                    True  evm.model.pcontig_000.132   \n",
       "21                    True  evm.model.pcontig_000.133   \n",
       "22                    True  evm.model.pcontig_000.134   \n",
       "23                    True  evm.model.pcontig_000.135   \n",
       "24                    True  evm.model.pcontig_000.136   \n",
       "25                    True  evm.model.pcontig_000.137   \n",
       "26                    True  evm.model.pcontig_000.138   \n",
       "27                    True  evm.model.pcontig_000.138   \n",
       "28                    True  evm.model.pcontig_000.139   \n",
       "29                    True   evm.model.pcontig_000.14   \n",
       "...                    ...                        ...   \n",
       "12112                 True   evm.model.pcontig_147.10   \n",
       "12113                 True   evm.model.pcontig_147.11   \n",
       "12114                 True   evm.model.pcontig_147.12   \n",
       "12115                 True   evm.model.pcontig_147.17   \n",
       "12116                 True   evm.model.pcontig_147.18   \n",
       "12117                 True    evm.model.pcontig_147.9   \n",
       "12118                 True    evm.model.pcontig_148.1   \n",
       "12119                 True    evm.model.pcontig_148.2   \n",
       "12120                 True    evm.model.pcontig_148.3   \n",
       "12121                 True    evm.model.pcontig_148.4   \n",
       "12122                 True   evm.model.pcontig_166.10   \n",
       "12123                 True   evm.model.pcontig_166.11   \n",
       "12124                 True    evm.model.pcontig_166.6   \n",
       "12125                 True    evm.model.pcontig_166.7   \n",
       "12126                 True    evm.model.pcontig_166.8   \n",
       "12127                 True    evm.model.pcontig_189.1   \n",
       "12128                 True    evm.model.pcontig_189.2   \n",
       "12129                 True    evm.model.pcontig_189.3   \n",
       "12130                 True    evm.model.pcontig_189.4   \n",
       "12131                 True    evm.model.pcontig_189.5   \n",
       "12132                 True    evm.model.pcontig_189.6   \n",
       "12133                 True    evm.model.pcontig_225.1   \n",
       "12134                 True    evm.model.pcontig_225.2   \n",
       "12135                 True    evm.model.pcontig_225.3   \n",
       "12136                 True    evm.model.pcontig_225.4   \n",
       "12137                 True    evm.model.pcontig_225.5   \n",
       "12138                 True    evm.model.pcontig_225.6   \n",
       "12139                 True    evm.model.pcontig_225.7   \n",
       "12140                 True    evm.model.pcontig_225.8   \n",
       "12141                 True    evm.model.pcontig_225.9   \n",
       "\n",
       "                         h_contig_overlap t_contig == h_contig_overlap  \n",
       "0      [hcontig_000_003, hcontig_000_031]                         True  \n",
       "1                       [hcontig_000_003]                         True  \n",
       "2                       [hcontig_000_003]                         True  \n",
       "3                       [hcontig_000_003]                         True  \n",
       "4                       [hcontig_000_003]                         True  \n",
       "5                       [hcontig_000_003]                         True  \n",
       "6                       [hcontig_000_003]                         True  \n",
       "7                       [hcontig_000_003]                         True  \n",
       "8                       [hcontig_000_003]                         True  \n",
       "9      [hcontig_000_003, hcontig_000_031]                         True  \n",
       "10     [hcontig_000_003, hcontig_000_031]                         True  \n",
       "11                      [hcontig_000_003]                         True  \n",
       "12                      [hcontig_000_003]                         True  \n",
       "13                      [hcontig_000_003]                         True  \n",
       "14                      [hcontig_000_003]                         True  \n",
       "15                      [hcontig_000_003]                         True  \n",
       "16                      [hcontig_000_003]                         True  \n",
       "17                      [hcontig_000_003]                         True  \n",
       "18     [hcontig_000_003, hcontig_000_031]                         True  \n",
       "19                      [hcontig_000_003]                         True  \n",
       "20                      [hcontig_000_003]                         True  \n",
       "21                      [hcontig_000_003]                         True  \n",
       "22                      [hcontig_000_003]                         True  \n",
       "23                      [hcontig_000_003]                         True  \n",
       "24                      [hcontig_000_003]                         True  \n",
       "25                      [hcontig_000_003]                         True  \n",
       "26                      [hcontig_000_003]                         True  \n",
       "27                      [hcontig_000_003]                         True  \n",
       "28                      [hcontig_000_003]                         True  \n",
       "29                      [hcontig_000_003]                         True  \n",
       "...                                   ...                          ...  \n",
       "12112                   [hcontig_147_001]                         True  \n",
       "12113                   [hcontig_147_001]                         True  \n",
       "12114                   [hcontig_147_001]                         True  \n",
       "12115                   [hcontig_147_001]                         True  \n",
       "12116                   [hcontig_147_001]                         True  \n",
       "12117                   [hcontig_147_001]                         True  \n",
       "12118                   [hcontig_148_001]                         True  \n",
       "12119                   [hcontig_148_001]                         True  \n",
       "12120                   [hcontig_148_001]                         True  \n",
       "12121                   [hcontig_148_001]                         True  \n",
       "12122                   [hcontig_166_001]                         True  \n",
       "12123                   [hcontig_166_001]                         True  \n",
       "12124                   [hcontig_166_001]                         True  \n",
       "12125                   [hcontig_166_001]                         True  \n",
       "12126                   [hcontig_166_001]                         True  \n",
       "12127                   [hcontig_189_001]                         True  \n",
       "12128                   [hcontig_189_001]                         True  \n",
       "12129                   [hcontig_189_001]                         True  \n",
       "12130                   [hcontig_189_001]                         True  \n",
       "12131                   [hcontig_189_001]                         True  \n",
       "12132                   [hcontig_189_001]                         True  \n",
       "12133                   [hcontig_225_001]                         True  \n",
       "12134                   [hcontig_225_001]                         True  \n",
       "12135                   [hcontig_225_001]                         True  \n",
       "12136                   [hcontig_225_001]                         True  \n",
       "12137                   [hcontig_225_001]                         True  \n",
       "12138                   [hcontig_225_001]                         True  \n",
       "12139                   [hcontig_225_001]                         True  \n",
       "12140                   [hcontig_225_001]                         True  \n",
       "12141                   [hcontig_225_001]                         True  \n",
       "\n",
       "[12142 rows x 20 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_hits_t_contig_and_h_contig.reset_index(drop=True).groupby('Query').apply()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp_blast_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp_p_gene_h_overlap_df = p_gene_and_p_by_h_df_2.groupby('p_protein')[10].apply(list)\n",
    "\n",
    "tmp_p_gene_h_overlap_dict = dict(zip(p_gene_h_overlap_df.index, p_gene_h_overlap_df.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp_p_gene_h_overlap_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "zip(p_gene_and_p_by_h_df_2['p_protein'], p_gene_and_p_by_h_df_2[10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp_blast_out[tmp_blast_out['Query'] == 'evm.model.pcontig_096.9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#write out the blast out dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "blast_out_dict[_key].iloc[14900:14999,[0,1,-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h = tmp_p_gene_bed.intersect(tmp_p_by_h_bed, wb=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df = p_gene_and_p_by_h.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df['p_gene'] = p_gene_and_p_by_h_df[8].str.extract(r'ID=([^;]*);')\n",
    "p_gene_and_p_by_h_df['p_protein'] = p_gene_and_p_by_h_df['p_gene'].str.replace('TU', 'model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df_2 = p_gene_and_p_by_h_df.loc[:, [10, 'p_protein']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df['On p_by_h'] = contig_000_df.apply(lambda row: haplotig_checker(row['Query'], row['Target']), axis =1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def haplotig_checker(Query, Target):\n",
    "    '''Check if protein blast hit lies on the corresponding h_contig'''\n",
    "    _tmp_df = p_gene_and_p_by_h_df_2[p_gene_and_p_by_h_df_2['p_protein'] == Query ]\n",
    "    if Target == 'NaN':\n",
    "        return False\n",
    "    else:\n",
    "        h_contig = re.search(r'(hcontig_[^.]*).', Target).group(1)\n",
    "        if sum(_tmp_df[10].str.contains(h_contig)) > 0:\n",
    "               return True\n",
    "        else:\n",
    "               return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "re.search(r'(hcontig[^.]*).','evm.model.hcontig_000_031.5').group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "grouby = [(p_gene_and_p_by_h_df.groupby('p_gene').count()  > 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "grouby #[grouby[0] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "len(p_gene_and_p_by_h_df['p_gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_gene_to_haplotig_dict = p_gene_and_p_by_h_df['p']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df = blast_out_dict['Pst_104E_v12_p_ctg.Pst_104E_v12_h_ctg.0.001.blastp.outfmt6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df = contig_000_df[contig_000_df['Query'].str.contains('pcontig_000')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df[(contig_000_df['Query'].str.contains('pcontig_000.10')) & (contig_000_df['Target'].str.contains('hcontig_000'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def feature_specific_haplotig(x,y df):\n",
    "    '''Transfers the specific haplotig that overlaps with the gene sequence'''\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this is slow and might need to be checked. Maybe simple add a h_contig column to the blast_out_dict dataframes\n",
    "def haplotig_checker(Query, Target):\n",
    "    '''Check if protein blast hit lies on the corresponding h_contig. Right now assumes p_gene_and_p_by_h_df_2 exists.'''\n",
    "    _tmp_df = p_gene_and_p_by_h_df_2[p_gene_and_p_by_h_df_2['p_protein'] == Query ]\n",
    "    if Target == 'NaN':\n",
    "        return False\n",
    "    else:\n",
    "        h_contig = re.search(r'(hcontig_[^.]*).', Target).group(1)\n",
    "        if sum(_tmp_df[10].str.contains(h_contig)) > 0:\n",
    "               return True\n",
    "        else:\n",
    "               return False"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
