{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/Bio/SearchIO/__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.\n",
      "  BiopythonExperimentalWarning)\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SearchIO\n",
    "from pybedtools import BedTool\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import multiprocessing\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "source_dir = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genome = 'Pst_104E_v12_p_ctg'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "out_dir = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/TE_analysis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#remove all commenting lines from the initial repet file\n",
    "!grep -v \"^#\" {source_dir}/{genome}.REPET.gff > {out_dir}/{genome}.REPET.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff = pd.read_csv(out_dir+'/'+genome+'.REPET.gff', sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "TE_post_analysis_p = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/REPET/Pst79_p/Pst79_p_full_annotate/postanalysis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TE_post_analysis_p_header = 'TE      length  covg    frags   fullLgthFrags   copies  fullLgthCopies  meanId  sdId    minId   q25Id   medId   q75Id   maxId   meanLgth        sdLgth  minLgth q25Lgth medLgth q75Lgth maxLgth meanLgthPerc    sdLgthPerc      minLgthPerc  q25LgthPerc     medLgthPerc     q75LgthPerc     maxLgthPerc'.split(' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "TE_post_analysis_p_header = [x for x in TE_post_analysis_p_header if x != '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#this needs to be fixed up to pick the proper summary table\n",
    "p_repet_summary_df = pd.read_csv(TE_post_analysis_p+'/'+'Pst79p_anno_chr_allTEs_nr_noSSR_join_path.annotStatsPerTE.tab' ,\\\n",
    "                                names = TE_post_analysis_p_header, header=None, sep='\\t', skiprows=1 )\n",
    "\n",
    "#check if I can filter the tab files for removing all TEs that are on the 2000 plus contigs\n",
    "#remove tRNAs TEs with infernal\n",
    "\n",
    "p_repet_summary_df['Code'] = p_repet_summary_df['TE'].apply(lambda x: x.split('_')[0])\n",
    "\n",
    "code_keys = p_repet_summary_df['Code'].unique()\n",
    "\n",
    "code_keys.sort()\n",
    "\n",
    "code_long = ['DNA_transposon Helitron', 'DNA_transposon Helitron', 'DNA_transposon Helitron', 'DNA_transposon Maverick',\\\n",
    "            'DNA_transposon TIR', 'DNA_transposon TIR', 'DNA_transposon TIR', 'DNA_transposon TIR', 'DNA_transposon noCat',\\\n",
    "             'DNA_transposon MITE','DNA_transposon MITE', 'Potential Host Gene', 'Retrotransposon LINE', 'Retrotransposon LINE',\\\n",
    "             'Retrotransposon LINE','Retrotransposon LTR','Retrotransposon LTR', 'Retrotransposon LTR', 'Retrotransposon LTR', 'Retrotransposon PLE', \\\n",
    "             'Retrotransposon SINE',  'Retrotransposon SINE', 'Retrotransposon noCat', 'Retrotransposon LARD',\\\n",
    "             'Retrotransposon LARD', 'Retrotransposon TRIM', 'Retrotransposon TRIM', 'Retrotransposon noCat',  \\\n",
    "             'Retrotransposon DIRS','Retrotransposon DIRS','Retrotransposon DIRS','Retrotransposon DIRS',\\\n",
    "             'noCat', 'noCat']\n",
    "\n",
    "code_dict = dict(zip(code_keys, code_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>copies</th>\n",
       "      <th>covg</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Code long</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA_transposon Helitron</th>\n",
       "      <td>1075</td>\n",
       "      <td>817566</td>\n",
       "      <td>2989.764706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon MITE</th>\n",
       "      <td>3789</td>\n",
       "      <td>886304</td>\n",
       "      <td>490.240741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon Maverick</th>\n",
       "      <td>268</td>\n",
       "      <td>345406</td>\n",
       "      <td>8562.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon TIR</th>\n",
       "      <td>19166</td>\n",
       "      <td>12711595</td>\n",
       "      <td>4020.474359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon noCat</th>\n",
       "      <td>5286</td>\n",
       "      <td>2376456</td>\n",
       "      <td>3034.734513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Potential Host Gene</th>\n",
       "      <td>1372</td>\n",
       "      <td>1375304</td>\n",
       "      <td>6120.490566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon DIRS</th>\n",
       "      <td>1337</td>\n",
       "      <td>1049299</td>\n",
       "      <td>6874.526316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon LARD</th>\n",
       "      <td>10752</td>\n",
       "      <td>4947564</td>\n",
       "      <td>5407.736111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon LINE</th>\n",
       "      <td>323</td>\n",
       "      <td>237992</td>\n",
       "      <td>4446.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon LTR</th>\n",
       "      <td>18893</td>\n",
       "      <td>16276421</td>\n",
       "      <td>6384.627841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon PLE</th>\n",
       "      <td>75</td>\n",
       "      <td>41843</td>\n",
       "      <td>8954.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon SINE</th>\n",
       "      <td>163</td>\n",
       "      <td>31138</td>\n",
       "      <td>317.923077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon TRIM</th>\n",
       "      <td>507</td>\n",
       "      <td>238695</td>\n",
       "      <td>1583.434783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon noCat</th>\n",
       "      <td>112</td>\n",
       "      <td>101098</td>\n",
       "      <td>4365.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>noCat</th>\n",
       "      <td>1218</td>\n",
       "      <td>850828</td>\n",
       "      <td>1679.829268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         copies      covg       length\n",
       "Code long                                             \n",
       "DNA_transposon Helitron    1075    817566  2989.764706\n",
       "DNA_transposon MITE        3789    886304   490.240741\n",
       "DNA_transposon Maverick     268    345406  8562.666667\n",
       "DNA_transposon TIR        19166  12711595  4020.474359\n",
       "DNA_transposon noCat       5286   2376456  3034.734513\n",
       "Potential Host Gene        1372   1375304  6120.490566\n",
       "Retrotransposon DIRS       1337   1049299  6874.526316\n",
       "Retrotransposon LARD      10752   4947564  5407.736111\n",
       "Retrotransposon LINE        323    237992  4446.625000\n",
       "Retrotransposon LTR       18893  16276421  6384.627841\n",
       "Retrotransposon PLE          75     41843  8954.000000\n",
       "Retrotransposon SINE        163     31138   317.923077\n",
       "Retrotransposon TRIM        507    238695  1583.434783\n",
       "Retrotransposon noCat       112    101098  4365.857143\n",
       "noCat                      1218    850828  1679.829268"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_summary_df['Code long'] = p_repet_summary_df['Code'].apply(lambda x: code_dict[x])\n",
    "\n",
    "p_repet_summary_sum_df = pd.pivot_table(p_repet_summary_df, values=['covg', 'copies'], index='Code long', aggfunc=np.sum)\n",
    "\n",
    "p_repet_summary_mean_df = pd.pivot_table(p_repet_summary_df, values='length', index='Code long', aggfunc=np.mean)\n",
    "\n",
    "pd.concat([p_repet_summary_sum_df,p_repet_summary_mean_df], axis=1 )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "p_repet_gff needs to be filtered to remove all the the high coverage contigs and such\n",
    "the filters are located in this folder\n",
    "/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now filter the gff dataframe to delete all the high coverage contigs\n",
    "#This might would have to be fixed as well. If we don't delete it as files should be already filtered\n",
    "contigs_smaller_2000 = pd.read_csv('/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly/pcontig_smaller_2000.txt',\\\n",
    "                                  header=None)[0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered = p_repet_gff[p_repet_gff[0].isin(contigs_smaller_2000)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def ID_filter_gff(_feature, _id):\n",
    "    \"\"\"\n",
    "    This filter parses out the top level id form the 9th gff column form a REPET gff file.\n",
    "    It has a specific search pattern for each feature type in column 2.\n",
    "    _type is defined by the feature '_'.join(feature.split(\"_\")[-2:])\n",
    "    \"\"\"\n",
    "    _type = '_'.join(_feature.split(\"_\")[-2:])\n",
    "    if _type == 'REPET_TEs':\n",
    "\n",
    "        TE_pattern = 'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9,-]*)_'\n",
    "        TE_prog = re.compile(TE_pattern)\n",
    "        TE_match = TE_prog.search(_id)\n",
    "        return TE_match.group(1)\n",
    "    if _type == 'REPET_SSRs':\n",
    "        \n",
    "        SSR_pattern = 'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([A-Z,a-z,0-9,-]*)'\n",
    "        SSR_prog = re.compile(SSR_pattern)\n",
    "        SSR_match = SSR_prog.search(_id)\n",
    "        return SSR_match.group(1)\n",
    "    if _type == 'REPET_tblastx' or _type == 'REPET_blastx':\n",
    "        #if \"#\" in _id:\n",
    "        #     blast_pattern = 'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([A-Z,a-z,0-9,-]*[_]?[A-Z,a-z,0-9,-]*[_|#|0-9]+?:[A-Z,a-z,0-9,-,:]*)'\n",
    "        #else:\n",
    "        #    blast_pattern = r'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([\\w+|:|-]*)'\n",
    "        blast_prog = re.compile(r'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([^;| ]*)')\n",
    "        #blast_prog = re.compile(blast_pattern)\n",
    "        blast_match = blast_prog.search(_id)\n",
    "        return blast_match.group(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered['ID'] = p_repet_gff_filtered.apply(lambda row: ID_filter_gff(row[1], row[8]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#filter out potential host genes\n",
    "p_repet_gff_filtered = p_repet_gff_filtered[~p_repet_gff_filtered[8].str.contains(\"Potential\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "num_unique_REs = len(p_repet_gff_filtered['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the number of unique repetitive elements: 48659\n"
     ]
    }
   ],
   "source": [
    "print('This is the number of unique repetitive elements: %i' % num_unique_REs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "frequency_of_REs  = p_repet_gff_filtered.groupby('ID')[0].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "frequency_of_REs.sort_values(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered.to_csv(out_dir+'/'+genome+'.REPET.filtered.gff', sep='\\t', header = None, index=None,columns=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered[8] = p_repet_gff_filtered['ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered_2 = p_repet_gff_filtered.iloc[:,0:9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered_2.to_csv(out_dir+'/'+genome+'.REPET.ID_column.gff', sep='\\t', header = None, index=None,columns=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#have another dataframe that only contains the REPET denovo annotation and not blast hits\n",
    "repet_gff_filtered_TEs = p_repet_gff_filtered_2[~p_repet_gff_filtered_2[1].str.contains('blast')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#get all the 'Code' phrases blast hits and make a blast_code_dict\n",
    "\n",
    "blast_codes = p_repet_gff_filtered_2[(p_repet_gff_filtered_2[1]=='Pst79p_anno_REPET_tblastx') | (p_repet_gff_filtered_2[1]=='Pst79p_anno_REPET_blastx')][8]\n",
    "\n",
    "blast_codes_list = [ ':'.join(x.split(':')[1:-1]) for x in blast_codes.unique()]\n",
    "\n",
    "blast_codes_list_unique = list(set(blast_codes_list))\n",
    "\n",
    "blast_codes_list_unique.sort()\n",
    "\n",
    "blast_codes_list_unique\n",
    "\n",
    "blast_code_long = ['Retrotransposon noCat', 'Retrotransposon DIRS', 'Retrotransposon LINE', 'Retrotransposon LTR', 'Retrotransposon PLE','DNA_transposon noCat',\\\n",
    "                   'DNA_transposon Crypton','DNA_transposon Helitron','DNA_transposon Maverick','DNA_transposon TIR']\n",
    "\n",
    "blast_code_dict = dict(zip(blast_codes_list_unique, blast_code_long))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ClassI:?': 'Retrotransposon noCat',\n",
       " 'ClassI:DIRS': 'Retrotransposon DIRS',\n",
       " 'ClassI:LINE': 'Retrotransposon LINE',\n",
       " 'ClassI:LTR': 'Retrotransposon LTR',\n",
       " 'ClassI:PLE': 'Retrotransposon PLE',\n",
       " 'ClassII:?': 'DNA_transposon noCat',\n",
       " 'ClassII:Crypton': 'DNA_transposon Crypton',\n",
       " 'ClassII:Helitron': 'DNA_transposon Helitron',\n",
       " 'ClassII:Maverick': 'DNA_transposon Maverick',\n",
       " 'ClassII:TIR': 'DNA_transposon TIR'}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_code_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#write a filer function that adds a 'code long'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def code_long_filter_gff(_feature, _id):\n",
    "    \"\"\"\n",
    "    This filter parses out Order and class of the TE based on Wicker et al. using the previously generated ID column. \n",
    "    It has a specific search pattern for each feature type in column 2.\n",
    "    _type is defined by the feature '_'.join(feature.split(\"_\")[-2:])\n",
    "    \"\"\"\n",
    "    _type = '_'.join(_feature.split(\"_\")[-2:])\n",
    "    if _type == 'REPET_TEs':\n",
    "        #split the \n",
    "        code = _id.split('_')[0]\n",
    "        return code_dict[code]\n",
    "    if _type == 'REPET_SSRs':\n",
    "        return 'SSR'\n",
    "    if _type == 'REPET_tblastx' or _type == 'REPET_blastx':\n",
    "        code = ':'.join(_id.split(':')[1:-1])\n",
    "        return blast_code_dict[code]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered_2['Classification'] = p_repet_gff_filtered_2.apply(lambda row: code_long_filter_gff(row[1], row[8]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DNA_transposon TIR', 'Retrotransposon DIRS', 'Retrotransposon LTR',\n",
       "       'SSR', 'DNA_transposon noCat', 'DNA_transposon MITE',\n",
       "       'Retrotransposon LARD', 'Retrotransposon SINE', 'noCat',\n",
       "       'DNA_transposon Helitron', 'DNA_transposon Maverick',\n",
       "       'Retrotransposon LINE', 'Retrotransposon TRIM',\n",
       "       'Retrotransposon noCat', 'Retrotransposon PLE',\n",
       "       'DNA_transposon Crypton'], dtype=object)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_gff_filtered_2[\"Classification\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#in the .classificaiton gff the feature column 2 is the Wicker classification of the transposon\n",
    "p_repet_gff_filtered_2[8] = p_repet_gff_filtered_2['Classification'] \n",
    "#repet_gff_filtered_TEs[1] = repet_gff_filtered_TEs[\"Classification\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered_2.iloc[:,:-1].to_csv(out_dir+'/'+genome+'.REPET.classification.gff', sep='\\t', header = None, index=None,columns=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered_2[~p_repet_gff_filtered_2[1].str.contains('blast')].iloc[:,:-1].to_csv(out_dir+'/'+genome+'.REPET_noblast.classification.gff', sep='\\t', header = None, index=None,columns=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#write a similar filter as before using bedtools to move through all the classifications and caclulate coverage\n",
    "#summarize this in a table and compare to published stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#generate the directory structure to safe specific coverage files\n",
    "os.chdir(out_dir)\n",
    "TE_types = ['Retrotransposon', 'DNA_transposon', 'noCat', 'SSR']\n",
    "TE_path = [os.path.join(out_dir, x) for x in TE_types]\n",
    "TE_path_dict = dict(zip(TE_types, TE_path))\n",
    "for TE_type in TE_types:\n",
    "    new_path = os.path.join(out_dir, TE_type)\n",
    "    if not os.path.exists(new_path):\n",
    "        os.mkdir(new_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "repet_prefix = genome+'.REPET.classification'\n",
    "p_genome_file = genome+'.genome_file'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Next, we create a function to pass only features for a particular\n",
    "# featuretype.  This is similar to a \"grep\" operation when applied to every\n",
    "# feature in a BedTool\n",
    "def id_filter_classification(feature, _id):\n",
    "    if feature[8] == _id:\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# subset the id and safe in specific folder\n",
    "# return the subsetted file as bedtool\n",
    "def subset_id_classification(_id, bed_object):\n",
    "    #retrotransposon \n",
    "    if 'Retrotransposon' in _id:\n",
    "        out_path = TE_path_dict['Retrotransposon']   \n",
    "    #DNA_transponson\n",
    "    elif 'DNA_transposon' in _id:\n",
    "        out_path = TE_path_dict['DNA_transposon'] \n",
    "    #notCat \n",
    "    elif 'noCat' in _id:\n",
    "        out_path = TE_path_dict['noCat']\n",
    "    #SSR\n",
    "    elif 'SSR' in _id:\n",
    "        out_path = TE_path_dict['SSR']\n",
    "    repet_prefix = genome+ '.'+bed_object.fn.split('.')[-3] + '.classification'\n",
    "    out_fn = out_path+'/'+repet_prefix+'.'+_id+'.gff'\n",
    "    out_fn= out_fn.replace(\" \", '_')\n",
    "    result = bed_object.filter(id_filter_classification, _id).saveas(out_fn)\n",
    "    cov_fn = out_fn.replace('gff','cov')\n",
    "    cov = result.genome_coverage(dz=True,g=p_genome_file)\n",
    "    cov.saveas(cov_fn)\n",
    "    print(\"Done with %s \" % (out_fn))\n",
    "    #return pybedtools.BedTool(result.fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#pull in the classification gff, make classification array, loop over array to save all the cov_dataframes\n",
    "RE_id_gff = pybedtools.BedTool(out_dir+'/'+genome+'.REPET.classification.gff')\n",
    "g = RE_id_gff.remove_invalid().saveas(out_dir+'/'+genome+'.REPET.classification.bedobject')\n",
    "#use the blast filtered dataframe as well\n",
    "RE_id_gff_noblast = pybedtools.BedTool(out_dir+'/'+genome+'.REPET_noblast.classification.gff')\n",
    "g_noblast = RE_id_gff_noblast.remove_invalid().saveas(out_dir+'/'+genome+'.REPET_noblast.classification.bedobject')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#filter through the whole REPET_TE bedobject\n",
    "#maybe make some multiproccesses out of this\n",
    "classifications = p_repet_gff_filtered_2[\"Classification\"].unique()\n",
    "[subset_id_classification(x, g) for x in classifications]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#filter through the whole REPET_TE bedobject having removed the blast hits\n",
    "#maybe make some multiproccesses out of this\n",
    "classifications_noblast = p_repet_gff_filtered_2[~p_repet_gff_filtered_2[1].str.contains('blast')][\"Classification\"].unique()\n",
    "[subset_id_classification(x, g_noblast) for x in classifications_noblast]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cur_dir = os.path.abspath(os.path.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genome_df = pd.read_csv(p_genome_file, sep='\\t', header=None,names=['contig', 'length'])\n",
    "\n",
    "genome_size = genome_df['length'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "noCat\n",
      "SSR\n",
      "DNA_transposon_noCat\n",
      "DNA_transposon_Helitron\n",
      "DNA_transposon_Crypton\n",
      "DNA_transposon_Maverick\n",
      "DNA_transposon_TIR\n",
      "DNA_transposon_MITE\n",
      "Retrotransposon_PLE\n",
      "Retrotransposon_Crypton\n",
      "Retrotransposon_LTR\n",
      "Retrotransposon_DIRS\n",
      "Retrotransposon_TRIM\n",
      "Retrotransposon_LARD\n",
      "Retrotransposon_noCat\n",
      "Retrotransposon_LINE\n",
      "Retrotransposon_SINE\n"
     ]
    }
   ],
   "source": [
    "#this caputures all REPET classifications including blast and internal REPET_TEs\n",
    "#the problem found here is that the blast and the REPET annotation is sometimes contratictory and overlapping\n",
    "class_cov_files = []\n",
    "for dirpath, dirname, filenames in os.walk(cur_dir):\n",
    "    #print(dirpath)\n",
    "    #print(len(filenames))\n",
    "    if len(filenames) == 0:  # empty folder\n",
    "        continue\n",
    "    cov_files = [dirpath +'/'+x for x in filenames if x.endswith('.cov') and 'REPET.classification' in x]\n",
    "    for file in cov_files:\n",
    "        class_cov_files.append(file)\n",
    "\n",
    "#make a large summary dataframe from all the cov files where the last \n",
    "df_list =[]\n",
    "class_cov_files.sort()\n",
    "for file in class_cov_files:\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header = None)\n",
    "    tmp_df[\"Code long\"] = file.split('.')[-2]\n",
    "    tmp_df.drop_duplicates(inplace=True)\n",
    "    df_list.append(tmp_df)\n",
    "    print(file.split('.')[-2])\n",
    "\n",
    "df_REPET_classification = pd.concat(df_list)\n",
    "\n",
    "cov_per_class = df_REPET_classification.pivot_table(values=1, columns= 'Code long', aggfunc='count')\n",
    "cov_per_contig_per_class = df_REPET_classification.groupby([0, 'Code long'])[1].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#this caputures all REPET classifications including internal REPET_TEs classification only. No blast hits included\n",
    "class_cov_files = []\n",
    "for dirpath, dirname, filenames in os.walk(cur_dir):\n",
    "    #print(dirpath)\n",
    "    #print(len(filenames))\n",
    "    if len(filenames) == 0:  # empty folder\n",
    "        continue\n",
    "    cov_files = [dirpath +'/'+x for x in filenames if x.endswith('.cov') and 'REPET_noblast.classification' in x ]\n",
    "    for file in cov_files:\n",
    "        class_cov_files.append(file)\n",
    "\n",
    "#make a large summary dataframe from all the cov files where the last \n",
    "df_list =[]\n",
    "class_cov_files.sort()\n",
    "for file in class_cov_files:\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header = None)\n",
    "    tmp_df[\"Code long\"] = file.split('.')[-2]\n",
    "    tmp_df.drop_duplicates(inplace=True)\n",
    "    df_list.append(tmp_df)\n",
    "    print(file.split('.')[-2])\n",
    "\n",
    "df_REPET_noblast_classification = pd.concat(df_list)\n",
    "\n",
    "cov_per_class_noblast = df_REPET_noblast_classification.pivot_table(values=1, columns= 'Code long', aggfunc='count')\n",
    "cov_per_contig_per_class_noblast = df_REPET_noblast_classification.groupby([0, 'Code long'])[1].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Code long\n",
       "DNA_transposon_Crypton     0.000015\n",
       "DNA_transposon_Helitron    0.014380\n",
       "DNA_transposon_MITE        0.012342\n",
       "DNA_transposon_Maverick    0.004510\n",
       "DNA_transposon_TIR         0.180167\n",
       "DNA_transposon_noCat       0.055862\n",
       "Retrotransposon_Crypton    0.014513\n",
       "Retrotransposon_DIRS       0.003626\n",
       "Retrotransposon_LARD       0.066880\n",
       "Retrotransposon_LINE       0.004594\n",
       "Retrotransposon_LTR        0.258021\n",
       "Retrotransposon_PLE        0.000692\n",
       "Retrotransposon_SINE       0.000381\n",
       "Retrotransposon_TRIM       0.003303\n",
       "Retrotransposon_noCat      0.001614\n",
       "SSR                        0.023948\n",
       "noCat                      0.012289\n",
       "Name: 1, dtype: float64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_per_class/genome_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cov_per_class_noblast/genome_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65.71364309754486"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(cov_per_class/genome_size*100).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "(cov_per_class_noblast/genome_size*100).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54776012"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_REPET_classification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "len(df_REPET_noblast_classification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "len(df_REPET_noblast_classification.drop_duplicates([0,1]))/genome_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test = p_repet_gff_filtered_2[~p_repet_gff_filtered_2[1].str.contains('blast')].iloc[:,:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>485095</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>4244</td>\n",
       "      <td>4312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485096</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>4244</td>\n",
       "      <td>4312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483287</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>10017</td>\n",
       "      <td>10551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483288</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>10017</td>\n",
       "      <td>10551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483305</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>10552</td>\n",
       "      <td>10574</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                      1           2      3      4    5  6  \\\n",
       "485095  pcontig_000  Pst79p_anno_REPET_TEs       match   4244   4312  0.0  +   \n",
       "485096  pcontig_000  Pst79p_anno_REPET_TEs  match_part   4244   4312  0.0  +   \n",
       "483287  pcontig_000  Pst79p_anno_REPET_TEs       match  10017  10551  0.0  -   \n",
       "483288  pcontig_000  Pst79p_anno_REPET_TEs  match_part  10017  10551  0.0  -   \n",
       "483305  pcontig_000  Pst79p_anno_REPET_TEs       match  10552  10574  0.0  -   \n",
       "\n",
       "        7                    8  \n",
       "485095  .  DNA_transposon MITE  \n",
       "485096  .  DNA_transposon MITE  \n",
       "483287  .  Retrotransposon LTR  \n",
       "483288  .  Retrotransposon LTR  \n",
       "483305  .  Retrotransposon LTR  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>485095</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>4244</td>\n",
       "      <td>4312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485096</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>4244</td>\n",
       "      <td>4312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483287</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>10017</td>\n",
       "      <td>10551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483288</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>10017</td>\n",
       "      <td>10551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483305</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>10552</td>\n",
       "      <td>10574</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483306</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>10552</td>\n",
       "      <td>10574</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483951</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14601</td>\n",
       "      <td>14785</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483952</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14601</td>\n",
       "      <td>14785</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485309</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14786</td>\n",
       "      <td>14834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485310</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14786</td>\n",
       "      <td>14834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485268</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14838</td>\n",
       "      <td>14859</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485269</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14838</td>\n",
       "      <td>14859</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482923</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14860</td>\n",
       "      <td>15079</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482924</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14860</td>\n",
       "      <td>15079</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483033</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>15091</td>\n",
       "      <td>16376</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483034</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15091</td>\n",
       "      <td>15140</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483085</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>15141</td>\n",
       "      <td>15855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483086</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15141</td>\n",
       "      <td>15855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483035</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15856</td>\n",
       "      <td>16376</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485048</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485049</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485105</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485106</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485003</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>16792</td>\n",
       "      <td>16813</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485004</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16792</td>\n",
       "      <td>16813</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484074</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>17086</td>\n",
       "      <td>17659</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon DIRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484075</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>17086</td>\n",
       "      <td>17659</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon DIRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485191</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>17687</td>\n",
       "      <td>17851</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485192</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>17687</td>\n",
       "      <td>17851</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483909</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>18338</td>\n",
       "      <td>18359</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483910</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>18338</td>\n",
       "      <td>18359</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482933</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>18360</td>\n",
       "      <td>18560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482934</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>18360</td>\n",
       "      <td>18560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485149</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>18914</td>\n",
       "      <td>19263</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485150</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>18914</td>\n",
       "      <td>19263</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485067</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>19267</td>\n",
       "      <td>19524</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485068</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>19267</td>\n",
       "      <td>19524</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482880</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>19525</td>\n",
       "      <td>20138</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482881</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>19525</td>\n",
       "      <td>20138</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                      1           2      3      4    5  6  \\\n",
       "485095  pcontig_000  Pst79p_anno_REPET_TEs       match   4244   4312  0.0  +   \n",
       "485096  pcontig_000  Pst79p_anno_REPET_TEs  match_part   4244   4312  0.0  +   \n",
       "483287  pcontig_000  Pst79p_anno_REPET_TEs       match  10017  10551  0.0  -   \n",
       "483288  pcontig_000  Pst79p_anno_REPET_TEs  match_part  10017  10551  0.0  -   \n",
       "483305  pcontig_000  Pst79p_anno_REPET_TEs       match  10552  10574  0.0  -   \n",
       "483306  pcontig_000  Pst79p_anno_REPET_TEs  match_part  10552  10574  0.0  -   \n",
       "483951  pcontig_000  Pst79p_anno_REPET_TEs       match  14601  14785  0.0  -   \n",
       "483952  pcontig_000  Pst79p_anno_REPET_TEs  match_part  14601  14785  0.0  -   \n",
       "485309  pcontig_000  Pst79p_anno_REPET_TEs       match  14786  14834  0.0  -   \n",
       "485310  pcontig_000  Pst79p_anno_REPET_TEs  match_part  14786  14834  0.0  -   \n",
       "485268  pcontig_000  Pst79p_anno_REPET_TEs       match  14838  14859  0.0  -   \n",
       "485269  pcontig_000  Pst79p_anno_REPET_TEs  match_part  14838  14859  0.0  -   \n",
       "482923  pcontig_000  Pst79p_anno_REPET_TEs       match  14860  15079  0.0  -   \n",
       "482924  pcontig_000  Pst79p_anno_REPET_TEs  match_part  14860  15079  0.0  -   \n",
       "483033  pcontig_000  Pst79p_anno_REPET_TEs       match  15091  16376  0.0  +   \n",
       "483034  pcontig_000  Pst79p_anno_REPET_TEs  match_part  15091  15140  0.0  +   \n",
       "483085  pcontig_000  Pst79p_anno_REPET_TEs       match  15141  15855  0.0  +   \n",
       "483086  pcontig_000  Pst79p_anno_REPET_TEs  match_part  15141  15855  0.0  +   \n",
       "483035  pcontig_000  Pst79p_anno_REPET_TEs  match_part  15856  16376  0.0  +   \n",
       "485048  pcontig_000  Pst79p_anno_REPET_TEs       match  16484  16791  0.0  -   \n",
       "485049  pcontig_000  Pst79p_anno_REPET_TEs  match_part  16484  16791  0.0  -   \n",
       "485105  pcontig_000  Pst79p_anno_REPET_TEs       match  16484  16791  0.0  -   \n",
       "485106  pcontig_000  Pst79p_anno_REPET_TEs  match_part  16484  16791  0.0  -   \n",
       "485003  pcontig_000  Pst79p_anno_REPET_TEs       match  16792  16813  0.0  -   \n",
       "485004  pcontig_000  Pst79p_anno_REPET_TEs  match_part  16792  16813  0.0  -   \n",
       "484074  pcontig_000  Pst79p_anno_REPET_TEs       match  17086  17659  0.0  -   \n",
       "484075  pcontig_000  Pst79p_anno_REPET_TEs  match_part  17086  17659  0.0  -   \n",
       "485191  pcontig_000  Pst79p_anno_REPET_TEs       match  17687  17851  0.0  -   \n",
       "485192  pcontig_000  Pst79p_anno_REPET_TEs  match_part  17687  17851  0.0  -   \n",
       "483909  pcontig_000  Pst79p_anno_REPET_TEs       match  18338  18359  0.0  -   \n",
       "483910  pcontig_000  Pst79p_anno_REPET_TEs  match_part  18338  18359  0.0  -   \n",
       "482933  pcontig_000  Pst79p_anno_REPET_TEs       match  18360  18560  0.0  +   \n",
       "482934  pcontig_000  Pst79p_anno_REPET_TEs  match_part  18360  18560  0.0  +   \n",
       "485149  pcontig_000  Pst79p_anno_REPET_TEs       match  18914  19263  0.0  +   \n",
       "485150  pcontig_000  Pst79p_anno_REPET_TEs  match_part  18914  19263  0.0  +   \n",
       "485067  pcontig_000  Pst79p_anno_REPET_TEs       match  19267  19524  0.0  -   \n",
       "485068  pcontig_000  Pst79p_anno_REPET_TEs  match_part  19267  19524  0.0  -   \n",
       "482880  pcontig_000  Pst79p_anno_REPET_TEs       match  19525  20138  0.0  -   \n",
       "482881  pcontig_000  Pst79p_anno_REPET_TEs  match_part  19525  20138  0.0  -   \n",
       "\n",
       "        7                     8  \n",
       "485095  .   DNA_transposon MITE  \n",
       "485096  .   DNA_transposon MITE  \n",
       "483287  .   Retrotransposon LTR  \n",
       "483288  .   Retrotransposon LTR  \n",
       "483305  .   Retrotransposon LTR  \n",
       "483306  .   Retrotransposon LTR  \n",
       "483951  .  Retrotransposon LARD  \n",
       "483952  .  Retrotransposon LARD  \n",
       "485309  .  Retrotransposon LARD  \n",
       "485310  .  Retrotransposon LARD  \n",
       "485268  .  Retrotransposon LARD  \n",
       "485269  .  Retrotransposon LARD  \n",
       "482923  .   DNA_transposon MITE  \n",
       "482924  .   DNA_transposon MITE  \n",
       "483033  .  DNA_transposon noCat  \n",
       "483034  .  DNA_transposon noCat  \n",
       "483085  .   Retrotransposon LTR  \n",
       "483086  .   Retrotransposon LTR  \n",
       "483035  .  DNA_transposon noCat  \n",
       "485048  .    DNA_transposon TIR  \n",
       "485049  .    DNA_transposon TIR  \n",
       "485105  .  DNA_transposon noCat  \n",
       "485106  .  DNA_transposon noCat  \n",
       "485003  .    DNA_transposon TIR  \n",
       "485004  .    DNA_transposon TIR  \n",
       "484074  .  Retrotransposon DIRS  \n",
       "484075  .  Retrotransposon DIRS  \n",
       "485191  .   Retrotransposon LTR  \n",
       "485192  .   Retrotransposon LTR  \n",
       "483909  .  Retrotransposon LARD  \n",
       "483910  .  Retrotransposon LARD  \n",
       "482933  .   DNA_transposon MITE  \n",
       "482934  .   DNA_transposon MITE  \n",
       "485149  .   Retrotransposon LTR  \n",
       "485150  .   Retrotransposon LTR  \n",
       "485067  .    DNA_transposon TIR  \n",
       "485068  .    DNA_transposon TIR  \n",
       "482880  .    DNA_transposon TIR  \n",
       "482881  .    DNA_transposon TIR  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test[(test[0]== 'pcontig_000' )&(test[3] < 20000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test.sort_values([0,3],inplace =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_REPET_noblast_classification.sort_values([0,1], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>Code long</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3960575</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14602</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030528</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14602</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960576</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14603</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030529</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14603</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960577</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14604</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030530</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14604</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960578</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14605</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030531</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14605</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960579</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14606</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030532</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14606</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960580</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14607</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030533</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14607</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960581</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14608</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030534</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14608</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960582</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14609</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030535</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14609</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960583</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14610</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030536</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14610</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960584</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14611</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030537</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14611</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960585</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14612</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030538</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14612</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960586</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14613</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030539</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14613</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960587</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14614</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030540</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14614</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960588</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14615</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030541</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14615</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960589</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14616</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA_transposon_noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5030542</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>14616</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428098</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>343</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369523</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>343</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428099</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>344</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369524</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>344</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428100</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>345</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369525</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>345</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428101</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>346</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369526</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>346</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428102</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>347</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369527</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>347</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428103</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>348</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369528</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>348</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428104</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>349</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369529</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>349</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428105</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>350</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369530</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>350</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428106</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>351</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369531</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>351</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428107</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>352</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369532</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>352</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428108</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>353</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369533</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>353</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428109</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369534</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>354</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428110</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>355</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369535</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>355</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428111</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>356</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369536</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>356</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428112</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>357</td>\n",
       "      <td>2</td>\n",
       "      <td>Retrotransposon_LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20369537</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>357</td>\n",
       "      <td>1</td>\n",
       "      <td>Retrotransposon_LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17972336 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0      1  2             Code long\n",
       "3960575   pcontig_000  14602  2  DNA_transposon_noCat\n",
       "5030528   pcontig_000  14602  2  Retrotransposon_LARD\n",
       "3960576   pcontig_000  14603  2  DNA_transposon_noCat\n",
       "5030529   pcontig_000  14603  2  Retrotransposon_LARD\n",
       "3960577   pcontig_000  14604  2  DNA_transposon_noCat\n",
       "5030530   pcontig_000  14604  2  Retrotransposon_LARD\n",
       "3960578   pcontig_000  14605  2  DNA_transposon_noCat\n",
       "5030531   pcontig_000  14605  2  Retrotransposon_LARD\n",
       "3960579   pcontig_000  14606  2  DNA_transposon_noCat\n",
       "5030532   pcontig_000  14606  2  Retrotransposon_LARD\n",
       "3960580   pcontig_000  14607  2  DNA_transposon_noCat\n",
       "5030533   pcontig_000  14607  2  Retrotransposon_LARD\n",
       "3960581   pcontig_000  14608  2  DNA_transposon_noCat\n",
       "5030534   pcontig_000  14608  2  Retrotransposon_LARD\n",
       "3960582   pcontig_000  14609  2  DNA_transposon_noCat\n",
       "5030535   pcontig_000  14609  2  Retrotransposon_LARD\n",
       "3960583   pcontig_000  14610  2  DNA_transposon_noCat\n",
       "5030536   pcontig_000  14610  2  Retrotransposon_LARD\n",
       "3960584   pcontig_000  14611  2  DNA_transposon_noCat\n",
       "5030537   pcontig_000  14611  2  Retrotransposon_LARD\n",
       "3960585   pcontig_000  14612  2  DNA_transposon_noCat\n",
       "5030538   pcontig_000  14612  2  Retrotransposon_LARD\n",
       "3960586   pcontig_000  14613  2  DNA_transposon_noCat\n",
       "5030539   pcontig_000  14613  2  Retrotransposon_LARD\n",
       "3960587   pcontig_000  14614  2  DNA_transposon_noCat\n",
       "5030540   pcontig_000  14614  2  Retrotransposon_LARD\n",
       "3960588   pcontig_000  14615  2  DNA_transposon_noCat\n",
       "5030541   pcontig_000  14615  2  Retrotransposon_LARD\n",
       "3960589   pcontig_000  14616  2  DNA_transposon_noCat\n",
       "5030542   pcontig_000  14616  2  Retrotransposon_LARD\n",
       "...               ...    ... ..                   ...\n",
       "5428098   pcontig_248    343  2  Retrotransposon_LARD\n",
       "20369523  pcontig_248    343  1   Retrotransposon_LTR\n",
       "5428099   pcontig_248    344  2  Retrotransposon_LARD\n",
       "20369524  pcontig_248    344  1   Retrotransposon_LTR\n",
       "5428100   pcontig_248    345  2  Retrotransposon_LARD\n",
       "20369525  pcontig_248    345  1   Retrotransposon_LTR\n",
       "5428101   pcontig_248    346  2  Retrotransposon_LARD\n",
       "20369526  pcontig_248    346  1   Retrotransposon_LTR\n",
       "5428102   pcontig_248    347  2  Retrotransposon_LARD\n",
       "20369527  pcontig_248    347  1   Retrotransposon_LTR\n",
       "5428103   pcontig_248    348  2  Retrotransposon_LARD\n",
       "20369528  pcontig_248    348  1   Retrotransposon_LTR\n",
       "5428104   pcontig_248    349  2  Retrotransposon_LARD\n",
       "20369529  pcontig_248    349  1   Retrotransposon_LTR\n",
       "5428105   pcontig_248    350  2  Retrotransposon_LARD\n",
       "20369530  pcontig_248    350  1   Retrotransposon_LTR\n",
       "5428106   pcontig_248    351  2  Retrotransposon_LARD\n",
       "20369531  pcontig_248    351  1   Retrotransposon_LTR\n",
       "5428107   pcontig_248    352  2  Retrotransposon_LARD\n",
       "20369532  pcontig_248    352  1   Retrotransposon_LTR\n",
       "5428108   pcontig_248    353  2  Retrotransposon_LARD\n",
       "20369533  pcontig_248    353  1   Retrotransposon_LTR\n",
       "5428109   pcontig_248    354  2  Retrotransposon_LARD\n",
       "20369534  pcontig_248    354  1   Retrotransposon_LTR\n",
       "5428110   pcontig_248    355  2  Retrotransposon_LARD\n",
       "20369535  pcontig_248    355  1   Retrotransposon_LTR\n",
       "5428111   pcontig_248    356  2  Retrotransposon_LARD\n",
       "20369536  pcontig_248    356  1   Retrotransposon_LTR\n",
       "5428112   pcontig_248    357  2  Retrotransposon_LARD\n",
       "20369537  pcontig_248    357  1   Retrotransposon_LTR\n",
       "\n",
       "[17972336 rows x 4 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_REPET_noblast_classification[df_REPET_noblast_classification.duplicated([0,1], keep=False )]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_REPET_classification.drop_duplicates(subset=[0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_REPET_classification.sort_values([0,1], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_REPET_classification_dup = df_REPET_classification[df_REPET_classification.duplicated(subset=[0,1], keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered_2[(p_repet_gff_filtered_2[0] ==  'pcontig_000')&(p_repet_gff_filtered_2[3] < 14602)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>Classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>485096</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>4244</td>\n",
       "      <td>4312</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485095</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>4244</td>\n",
       "      <td>4312</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483288</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>10017</td>\n",
       "      <td>10551</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483287</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>10017</td>\n",
       "      <td>10551</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483305</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>10552</td>\n",
       "      <td>10574</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483306</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>10552</td>\n",
       "      <td>10574</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483951</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14601</td>\n",
       "      <td>14785</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483952</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14601</td>\n",
       "      <td>14785</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498268</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>14603</td>\n",
       "      <td>14784</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498269</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14603</td>\n",
       "      <td>14690</td>\n",
       "      <td>2.000000e-13</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498270</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14707</td>\n",
       "      <td>14784</td>\n",
       "      <td>3.000000e-12</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485309</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14786</td>\n",
       "      <td>14834</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485310</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14786</td>\n",
       "      <td>14834</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485268</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14838</td>\n",
       "      <td>14859</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485269</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14838</td>\n",
       "      <td>14859</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482924</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14860</td>\n",
       "      <td>15079</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482923</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>14860</td>\n",
       "      <td>15079</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483033</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>15091</td>\n",
       "      <td>16376</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483034</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15091</td>\n",
       "      <td>15140</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483086</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15141</td>\n",
       "      <td>15855</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483085</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>15141</td>\n",
       "      <td>15855</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498372</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15195</td>\n",
       "      <td>15362</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498371</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>15195</td>\n",
       "      <td>15499</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498373</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15416</td>\n",
       "      <td>15499</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498374</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>15748</td>\n",
       "      <td>15834</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498375</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15748</td>\n",
       "      <td>15834</td>\n",
       "      <td>6.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483035</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15856</td>\n",
       "      <td>16376</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485106</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485105</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485049</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485048</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>16484</td>\n",
       "      <td>16791</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498263</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>16541</td>\n",
       "      <td>16788</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498264</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16541</td>\n",
       "      <td>16788</td>\n",
       "      <td>3.000000e-44</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "      <td>DNA_transposon noCat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485004</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16792</td>\n",
       "      <td>16813</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485003</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>16792</td>\n",
       "      <td>16813</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484075</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>17086</td>\n",
       "      <td>17659</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon DIRS</td>\n",
       "      <td>Retrotransposon DIRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484074</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>17086</td>\n",
       "      <td>17659</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon DIRS</td>\n",
       "      <td>Retrotransposon DIRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485191</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>17687</td>\n",
       "      <td>17851</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485192</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>17687</td>\n",
       "      <td>17851</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483910</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>18338</td>\n",
       "      <td>18359</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483909</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>18338</td>\n",
       "      <td>18359</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "      <td>Retrotransposon LARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482933</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>18360</td>\n",
       "      <td>18560</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482934</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>18360</td>\n",
       "      <td>18560</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "      <td>DNA_transposon MITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485149</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>18914</td>\n",
       "      <td>19263</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485150</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>18914</td>\n",
       "      <td>19263</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "      <td>Retrotransposon LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485067</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>19267</td>\n",
       "      <td>19524</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485068</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>19267</td>\n",
       "      <td>19524</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482881</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match_part</td>\n",
       "      <td>19525</td>\n",
       "      <td>20138</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482880</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_TEs</td>\n",
       "      <td>match</td>\n",
       "      <td>19525</td>\n",
       "      <td>20138</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "      <td>DNA_transposon TIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                          1           2      3      4  \\\n",
       "485096  pcontig_000      Pst79p_anno_REPET_TEs  match_part   4244   4312   \n",
       "485095  pcontig_000      Pst79p_anno_REPET_TEs       match   4244   4312   \n",
       "483288  pcontig_000      Pst79p_anno_REPET_TEs  match_part  10017  10551   \n",
       "483287  pcontig_000      Pst79p_anno_REPET_TEs       match  10017  10551   \n",
       "483305  pcontig_000      Pst79p_anno_REPET_TEs       match  10552  10574   \n",
       "483306  pcontig_000      Pst79p_anno_REPET_TEs  match_part  10552  10574   \n",
       "483951  pcontig_000      Pst79p_anno_REPET_TEs       match  14601  14785   \n",
       "483952  pcontig_000      Pst79p_anno_REPET_TEs  match_part  14601  14785   \n",
       "498268  pcontig_000  Pst79p_anno_REPET_tblastx       match  14603  14784   \n",
       "498269  pcontig_000  Pst79p_anno_REPET_tblastx  match_part  14603  14690   \n",
       "498270  pcontig_000  Pst79p_anno_REPET_tblastx  match_part  14707  14784   \n",
       "485309  pcontig_000      Pst79p_anno_REPET_TEs       match  14786  14834   \n",
       "485310  pcontig_000      Pst79p_anno_REPET_TEs  match_part  14786  14834   \n",
       "485268  pcontig_000      Pst79p_anno_REPET_TEs       match  14838  14859   \n",
       "485269  pcontig_000      Pst79p_anno_REPET_TEs  match_part  14838  14859   \n",
       "482924  pcontig_000      Pst79p_anno_REPET_TEs  match_part  14860  15079   \n",
       "482923  pcontig_000      Pst79p_anno_REPET_TEs       match  14860  15079   \n",
       "483033  pcontig_000      Pst79p_anno_REPET_TEs       match  15091  16376   \n",
       "483034  pcontig_000      Pst79p_anno_REPET_TEs  match_part  15091  15140   \n",
       "483086  pcontig_000      Pst79p_anno_REPET_TEs  match_part  15141  15855   \n",
       "483085  pcontig_000      Pst79p_anno_REPET_TEs       match  15141  15855   \n",
       "498372  pcontig_000  Pst79p_anno_REPET_tblastx  match_part  15195  15362   \n",
       "498371  pcontig_000  Pst79p_anno_REPET_tblastx       match  15195  15499   \n",
       "498373  pcontig_000  Pst79p_anno_REPET_tblastx  match_part  15416  15499   \n",
       "498374  pcontig_000  Pst79p_anno_REPET_tblastx       match  15748  15834   \n",
       "498375  pcontig_000  Pst79p_anno_REPET_tblastx  match_part  15748  15834   \n",
       "483035  pcontig_000      Pst79p_anno_REPET_TEs  match_part  15856  16376   \n",
       "485106  pcontig_000      Pst79p_anno_REPET_TEs  match_part  16484  16791   \n",
       "485105  pcontig_000      Pst79p_anno_REPET_TEs       match  16484  16791   \n",
       "485049  pcontig_000      Pst79p_anno_REPET_TEs  match_part  16484  16791   \n",
       "485048  pcontig_000      Pst79p_anno_REPET_TEs       match  16484  16791   \n",
       "498263  pcontig_000  Pst79p_anno_REPET_tblastx       match  16541  16788   \n",
       "498264  pcontig_000  Pst79p_anno_REPET_tblastx  match_part  16541  16788   \n",
       "485004  pcontig_000      Pst79p_anno_REPET_TEs  match_part  16792  16813   \n",
       "485003  pcontig_000      Pst79p_anno_REPET_TEs       match  16792  16813   \n",
       "484075  pcontig_000      Pst79p_anno_REPET_TEs  match_part  17086  17659   \n",
       "484074  pcontig_000      Pst79p_anno_REPET_TEs       match  17086  17659   \n",
       "485191  pcontig_000      Pst79p_anno_REPET_TEs       match  17687  17851   \n",
       "485192  pcontig_000      Pst79p_anno_REPET_TEs  match_part  17687  17851   \n",
       "483910  pcontig_000      Pst79p_anno_REPET_TEs  match_part  18338  18359   \n",
       "483909  pcontig_000      Pst79p_anno_REPET_TEs       match  18338  18359   \n",
       "482933  pcontig_000      Pst79p_anno_REPET_TEs       match  18360  18560   \n",
       "482934  pcontig_000      Pst79p_anno_REPET_TEs  match_part  18360  18560   \n",
       "485149  pcontig_000      Pst79p_anno_REPET_TEs       match  18914  19263   \n",
       "485150  pcontig_000      Pst79p_anno_REPET_TEs  match_part  18914  19263   \n",
       "485067  pcontig_000      Pst79p_anno_REPET_TEs       match  19267  19524   \n",
       "485068  pcontig_000      Pst79p_anno_REPET_TEs  match_part  19267  19524   \n",
       "482881  pcontig_000      Pst79p_anno_REPET_TEs  match_part  19525  20138   \n",
       "482880  pcontig_000      Pst79p_anno_REPET_TEs       match  19525  20138   \n",
       "\n",
       "                   5  6  7                     8        Classification  \n",
       "485096  0.000000e+00  +  .   DNA_transposon MITE   DNA_transposon MITE  \n",
       "485095  0.000000e+00  +  .   DNA_transposon MITE   DNA_transposon MITE  \n",
       "483288  0.000000e+00  -  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483287  0.000000e+00  -  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483305  0.000000e+00  -  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483306  0.000000e+00  -  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483951  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "483952  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "498268  0.000000e+00  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "498269  2.000000e-13  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "498270  3.000000e-12  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "485309  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "485310  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "485268  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "485269  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "482924  0.000000e+00  -  .   DNA_transposon MITE   DNA_transposon MITE  \n",
       "482923  0.000000e+00  -  .   DNA_transposon MITE   DNA_transposon MITE  \n",
       "483033  0.000000e+00  +  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "483034  0.000000e+00  +  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "483086  0.000000e+00  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483085  0.000000e+00  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "498372  1.000000e-12  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "498371  0.000000e+00  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "498373  1.000000e-12  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "498374  0.000000e+00  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "498375  6.000000e-12  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483035  0.000000e+00  +  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "485106  0.000000e+00  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "485105  0.000000e+00  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "485049  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "485048  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "498263  0.000000e+00  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "498264  3.000000e-44  -  .  DNA_transposon noCat  DNA_transposon noCat  \n",
       "485004  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "485003  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "484075  0.000000e+00  -  .  Retrotransposon DIRS  Retrotransposon DIRS  \n",
       "484074  0.000000e+00  -  .  Retrotransposon DIRS  Retrotransposon DIRS  \n",
       "485191  0.000000e+00  -  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "485192  0.000000e+00  -  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "483910  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "483909  0.000000e+00  -  .  Retrotransposon LARD  Retrotransposon LARD  \n",
       "482933  0.000000e+00  +  .   DNA_transposon MITE   DNA_transposon MITE  \n",
       "482934  0.000000e+00  +  .   DNA_transposon MITE   DNA_transposon MITE  \n",
       "485149  0.000000e+00  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "485150  0.000000e+00  +  .   Retrotransposon LTR   Retrotransposon LTR  \n",
       "485067  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "485068  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "482881  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  \n",
       "482880  0.000000e+00  -  .    DNA_transposon TIR    DNA_transposon TIR  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_gff_filtered_2[(p_repet_gff_filtered_2[0] ==  'pcontig_000')&(p_repet_gff_filtered_2[3] < 20000)].sort_values(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_REPET_classification_dup.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "45267875/ genome_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!head 'Pst_104E_v12_p_ctg.REPET.ID_column.cov'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "RE_id_gff = pybedtools.BedTool(out_dir+'/'+genome+'.REPET.ID_column.gff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "os.chdir(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "repet_gff = 'Pst_104E_v12_p_ctg.REPET.gff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "repet_prefix = '.'.join(repet_gff.split('.')[0:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "g = RE_id_gff.remove_invalid().saveas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Next, we create a function to pass only features for a particular\n",
    "# featuretype.  This is similar to a \"grep\" operation when applied to every\n",
    "# feature in a BedTool\n",
    "def id_filter(feature, _id):\n",
    "    if feature[8] == _id:\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# subset the id and safe in specific folder\n",
    "# return the subsetted file as bedtool\n",
    "def subset_id(_id):\n",
    "    #ClassI are retrotransposon form blast\n",
    "    if 'ClassI:' in _id:\n",
    "        out_path = TE_path_dict['Retrotransposon']   \n",
    "    #ClassII are DNA_transponson\n",
    "    elif 'ClassII' in _id:\n",
    "        out_path = TE_path_dict['DNA_transposon'] \n",
    "    #The rest with '_' should be REPET_TEs\n",
    "    elif _id.split('_')[0] in list(code_dict.keys()):\n",
    "        key = code_dict[_id.split('_')[0]].split(' ')[0]\n",
    "        out_path = TE_path_dict[key]\n",
    "    #everything without '_' at the end should be SSR\n",
    "    elif '_' not in _id:\n",
    "        out_path = TE_path_dict['SSR']\n",
    "    out_fn = out_path+'/'+repet_prefix+'.'+_id+'.gff'\n",
    "    result = g.filter(id_filter, _id).saveas(out_fn)\n",
    "    cov_fn = out_fn.replace('gff','cov')\n",
    "    cov = result.genome_coverage(dz=True,g=p_genome_file)\n",
    "    cov.saveas(cov_fn)\n",
    "    #return pybedtools.BedTool(result.fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Write a loop that takes in the filtered id'ed gff file, filters the file according to the id field. The corresponding bedfile is saved in a specific folder for the type of TE (RNA, DNA, SSR, noCat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TE_types = ['Retrotransposon', 'DNA_transposon', 'noCat', 'SSR']\n",
    "TE_path = [os.path.join(out_dir, x) for x in TE_types]\n",
    "TE_path_dict = dict(zip(TE_types, TE_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for TE_type in TE_types:\n",
    "    new_path = os.path.join(out_dir, TE_type)\n",
    "    if not os.path.exists(new_path):\n",
    "        os.mkdir(new_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_id = frequency_of_REs.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this is REALLY slow for now. Would need to parallize this step. Look at the pool function of \n",
    "#multiprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "[subset_id(x) for x in _id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "full_path_list[0].replace('gff','cov')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_test_id = _id[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "g = RE_id_gff.remove_invalid().saveas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "[x for x in _test_id if '_' not in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_genome_file = 'Pst_104E_v12_p_ctg.genome_file'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_cov_RE = g.genome_coverage(dz=True,g=p_genome_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_cov_RE.saveas('Pst_104E_v12_p_ctg.REPET.ID_column.cov')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered[1].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "[x for x in p_repet_gff_filtered_2[(p_repet_gff_filtered_2[1] == 'Pst79p_anno_REPET_TE')  ][8].unique() if '_' not in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "[x for x in p_repet_gff_filtered_2[(p_repet_gff_filtered_2[1] == 'Pst79p_anno_REPET_tblastx') | (p_repet_gff_filtered_2[1] == 'Pst79p_anno_REPET_blastx') ][8].unique() if '_' not in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_repet_gff[p_repet_gff[8].str.contains('Copia6-VV_I_')][8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
