{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This juypter notebook is tartget towards defining alleles "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The idea is that alleles are on corresponding haplotig and a good blast hit.\n",
    "\n",
    "map all h contigs on their corresponding p alleles. Use the Assembletic '*oriented_coords.csv' file for this purpose.\n",
    "This needs to be converted to a bed file or a gff file in order to pull out the overlap between gene models and the h_contig using either bedtools or pandas. Check what overlaps can spit out on the other side. \n",
    ">Considerations here are what happens if nothing overlaps or what if muliptle overlaps. \n",
    "If nothing overlaps it would be good to look if this area is covered by 2x coverage when doing ph mapping.\n",
    "\n",
    "Do gene blast of p on h with certain cut off. Pull out the best hit that is located on the corresponding haplotig. If that is not there look for best hit on other haplotigs or primary contigs that is not itself.\n",
    "\n",
    ">Considerations are that p-h parings that are best hits and corresponding haplotig overule all others. There can only be a single hit. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from pybedtools import BedTool\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "import subprocess\n",
    "import shutil\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Define the PATH\n",
    "BASE_AA_PATH = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12'\n",
    "BASE_A_PATH = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly'\n",
    "ASSEMBLETICS_INPATH = os.path.join(BASE_AA_PATH, 'Assembletics')\n",
    "BLAST_DB = os.path.join(BASE_AA_PATH, 'blast_DB')\n",
    "OUT_PATH = os.path.join(BASE_AA_PATH, 'allele_analysis')\n",
    "OUT_tmp = os.path.join(OUT_PATH, 'tmp')\n",
    "if not os.path.isdir(OUT_PATH):\n",
    "    os.mkdir(OUT_PATH)\n",
    "if not os.path.isdir(OUT_tmp):\n",
    "    os.mkdir(OUT_tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Define your p and h genome and move it into the allele analysis folder\n",
    "p_genome = 'Pst_104E_v12_p_ctg'\n",
    "h_genome = 'Pst_104E_v12_h_ctg'\n",
    "genome_file_suffix = '.genome_file'\n",
    "for x in (x + '.fa' for x in [p_genome, h_genome]):\n",
    "    shutil.copy2(BASE_A_PATH+'/'+x, OUT_tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Define ENV parameters for blast hits and threads used in blast analysis\n",
    "n_threads = 16\n",
    "e_value = 1e-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def att_column_p(x, y, z, a, b):\n",
    "    '''Generate attribute column of h on p p_gff dataframe.\n",
    "    '''\n",
    "    string = 'h_contig=%s;query_start=%s;query_stop=%s;query_length=%s;query_aln_ln=%s' % (x, str(y), str(z), str(a), str(b))\n",
    "    return string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def att_column_h(x, y, z, a, b):\n",
    "    '''Generate attribute column of h on p h_gff dataframe.\n",
    "    '''\n",
    "    string = 'p_contig=%s;ref_start=%s;ref_stop=%s;ref_length=%s;ref_aln_ln=%s' % (x, str(y), str(z), str(a), str(b))\n",
    "    return string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def gene_contig_subset(df, contig, feature):\n",
    "    '''Input the gff and subset the feature columne by gene.\n",
    "        Return data frame subset by contig and gene'''\n",
    "    tmp_df = df[(df[0].str.contains(contig)) & (df[2]==feature)]\n",
    "    tmp_df.sort_values(3,inplace=True)\n",
    "    tmp_df.reset_index(drop=True, inplace=True)\n",
    "    return tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def same_contig_blast(x,y):\n",
    "    '''Function that checks if the blast hit in columne y is on the same contig as the the query sequence in\n",
    "    column y.\n",
    "    '''\n",
    "    q_contig = re.search(r'[p|h]([a-z]*_[0-9]*)_?', x).group(1)\n",
    "    if y != 'NaN':\n",
    "        hit_contig = re.search(r'[p|h]([a-z]*_[0-9]*)_?', y).group(1)\n",
    "    else:\n",
    "        hit_contig = 'NaN'\n",
    "    if q_contig == hit_contig:\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Considerations for blast analysis\n",
    "Do gene blast and protein blast. Initially go off protein blast results. They should be more imformative as it also includes the coding region and the frame.\n",
    "> Pull this in as a dataframe and filter it down as well to each contig. Do blasts both ways as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "makeblastdb -in Pst_104E_v12_h_ctg.anno.CDS.fa -dbtype nucl is done!\n",
      "makeblastdb -in Pst_104E_v12_h_ctg.anno.protein.fa -dbtype prot is done!\n",
      "All databases generated and ready to go!\n"
     ]
    }
   ],
   "source": [
    "#generate the blast databases if not already present\n",
    "os.chdir(BLAST_DB)\n",
    "blast_dir_content = os.listdir(BLAST_DB)\n",
    "for x in blast_dir_content:\n",
    "    if x.endswith('.fa') and ({os.path.isfile(x + e) for e in ['.psq', '.phr', '.pin'] } != {True}\\\n",
    "           and {os.path.isfile(x + e) for e in ['.nin', '.nhr', '.nsq'] } != {True} ):\n",
    "\n",
    "        make_DB_options = ['-in']\n",
    "        make_DB_options.append(x)\n",
    "        make_DB_options.append('-dbtype')\n",
    "        if 'protein' in x:\n",
    "            make_DB_options.append('prot')\n",
    "        else:\n",
    "            make_DB_options.append('nucl')\n",
    "        make_DB_command = 'makeblastdb %s' % ' '.join(make_DB_options)\n",
    "        make_DB_stderr = subprocess.check_output(make_DB_command, shell=True, stderr=subprocess.STDOUT)\n",
    "        print('%s is done!' % make_DB_command)\n",
    "print(\"All databases generated and ready to go!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastp -query Pst_104E_v12_p_ctg.anno.protein.fa -db Pst_104E_v12_h_ctg.anno.protein.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_p_ctg.Pst_104E_v12_h_ctg.0.001.blastp.outfmt6\n",
      "Previously done already!\n",
      "blastp -query Pst_104E_v12_h_ctg.anno.protein.fa -db Pst_104E_v12_p_ctg.anno.protein.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_h_ctg.Pst_104E_v12_p_ctg.0.001.blastp.outfmt6\n",
      "Previously done already!\n",
      "blastn -query Pst_104E_v12_p_ctg.anno.gene.fa -db Pst_104E_v12_h_ctg.anno.gene.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_p_ctg.Pst_104E_v12_h_ctg.0.001.blastn.outfmt6\n",
      "Previously done already!\n",
      "blastn -query Pst_104E_v12_h_ctg.anno.gene.fa -db Pst_104E_v12_p_ctg.anno.gene.fa -outfmt 6 -evalue 0.001 -num_threads 16 > /home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/allele_analysis/Pst_104E_v12_h_ctg.Pst_104E_v12_p_ctg.0.001.blastn.outfmt6\n",
      "Previously done already!\n"
     ]
    }
   ],
   "source": [
    "blast_dict = {}\n",
    "blast_dict['gene_fa'] = [x for x  in os.listdir(BLAST_DB) if x.endswith('gene.fa') and 'ph_ctg' not in x]\n",
    "blast_dict['protein_fa'] = [x for x  in os.listdir(BLAST_DB) if x.endswith('protein.fa') and 'ph_ctg' not in x]\n",
    "blast_stderr_dict = {}\n",
    "for key in blast_dict.keys():\n",
    "    for n, query in enumerate(blast_dict[key]):\n",
    "        tmp_list = blast_dict[key][:]\n",
    "        tmp_stderr_list = []\n",
    "        #this loops through all remaining files and does blast against all those\n",
    "        del tmp_list[n]\n",
    "        for db in tmp_list:\n",
    "            blast_options = ['-query']\n",
    "            blast_options.append(query)\n",
    "            blast_options.append('-db')\n",
    "\n",
    "\n",
    "            blast_options.append(db)\n",
    "            blast_options.append('-outfmt 6')\n",
    "            blast_options.append('-evalue')\n",
    "            blast_options.append(str(e_value))\n",
    "            blast_options.append('-num_threads')\n",
    "            blast_options.append(str(n_threads))\n",
    "            blast_options.append('>')\n",
    "            if 'gene' in query and 'gene' in db:\n",
    "                out_name_list = [ query.split('.')[0], db.split('.')[0], str(e_value), 'blastn.outfmt6']\n",
    "                out_name = os.path.join(OUT_PATH,'.'.join(out_name_list))\n",
    "                blast_options.append(out_name)\n",
    "                blast_command = 'blastn %s' % ' '.join(blast_options)\n",
    "            elif 'protein' in query and 'protein' in db:\n",
    "                out_name_list = [ query.split('.')[0], db.split('.')[0], str(e_value), 'blastp.outfmt6']\n",
    "                out_name = os.path.join(OUT_PATH,'.'.join(out_name_list))\n",
    "                blast_options.append(out_name)\n",
    "                blast_command = 'blastp %s' % ' '.join(blast_options)\n",
    "            print(blast_command)\n",
    "            if not os.path.exists(out_name):\n",
    "                blast_stderr_dict[blast_command] = subprocess.check_output(blast_command, shell=True, stderr=subprocess.STDOUT)\n",
    "                print(\"New blast run and done!\")\n",
    "            else:\n",
    "                blast_stderr_dict[blast_command] = 'Previously done already!'\n",
    "                print('Previously done already!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#make a sequnece length dict for all genes and proteins for which blast was performed\n",
    "seq_list = []\n",
    "length_list =[]\n",
    "for key in blast_dict.keys():\n",
    "    for file in blast_dict[key]:\n",
    "        for seq in SeqIO.parse(open(file), 'fasta'):\n",
    "            seq_list.append(seq.id)\n",
    "            length_list.append(len(seq.seq))\n",
    "length_dict = dict(zip(seq_list, length_list))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60498"
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(length_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#get assembletics folders\n",
    "ass_folders = [os.path.join(ASSEMBLETICS_INPATH, x) for x in os.listdir(ASSEMBLETICS_INPATH) if x.endswith('_php_8kbp')]\n",
    "ass_folders.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#read in the gff files\n",
    "p_gff = pd.read_csv(BASE_A_PATH+'/'+p_genome+'.anno.gff3', sep='\\t', header=None)\n",
    "h_gff = pd.read_csv(BASE_A_PATH+'/'+h_genome+'.anno.gff3', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#read in the blast df and add QLgth and QCov columns\n",
    "blast_out_dict = {}\n",
    "blast_header = ['Query', 'Target', 'PctID', 'AlnLgth', 'NumMis', 'NumGap', 'StartQuery', 'StopQuery', 'StartTarget',\\\n",
    "              'StopTarget', 'e-value','BitScore']\n",
    "for key in blast_stderr_dict.keys():\n",
    "    file = key.split('>')[-1][1:]\n",
    "    tmp_key = file.split('/')[-1]\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header=None, names=blast_header)\n",
    "    tmp_df[\"QLgth\"] = tmp_df[\"Query\"].apply(lambda x: length_dict[x])\n",
    "    tmp_df[\"QCov\"] = tmp_df['AlnLgth']/tmp_df['QLgth']*100\n",
    "    tmp_df.sort_values(by=['Query', 'e-value','BitScore', ],ascending=[True, True, False], inplace=True)\n",
    "    #now make sure to add proteins/genes without blast hit to the dataframes\n",
    "    #check if gene blast or protein blast and pull out respective query file\n",
    "    if file.split('.')[-2] == 'blastp':\n",
    "        tmp_blast_seq = [x for x in blast_dict['protein_fa'] if x.startswith(tmp_key.split('.')[0])][0]\n",
    "    elif file.split('.')[-2] == 'blastn':\n",
    "        tmp_blast_seq = [x for x in blast_dict['gene_fa'] if x.startswith(tmp_key.split('.')[0])][0]\n",
    "    tmp_all_blast_seq = []\n",
    "    #make list off all query sequences\n",
    "    for seq in SeqIO.parse(os.path.join(BLAST_DB, tmp_blast_seq), 'fasta'):\n",
    "        tmp_all_blast_seq.append(str(seq.id))\n",
    "    tmp_all_queries_w_hit = tmp_df[\"Query\"].unique()\n",
    "    tmp_queries_no_hit = set(tmp_all_blast_seq) - set(tmp_all_queries_w_hit)\n",
    "    no_hit_list = []\n",
    "    #loop over the quieres with no hit and make list of list out of them the first element being the query id\n",
    "    for x in tmp_queries_no_hit:\n",
    "        NA_list = ['NaN'] * len(tmp_df.columns)\n",
    "        NA_list[0] = x\n",
    "        no_hit_list.append(NA_list)\n",
    "    tmp_no_hit_df = pd.DataFrame(no_hit_list)\n",
    "    tmp_no_hit_df.columns = tmp_df.columns\n",
    "    tmp_no_hit_df['QLgth'] = tmp_no_hit_df.Query.apply(lambda x: length_dict[x])\n",
    "    tmp_df = tmp_df.append(tmp_no_hit_df)\n",
    "    tmp_df['q_contig == s_contig'] = tmp_df.apply(lambda row: same_contig_blast(row['Query'], row['Target']), axis =1)\n",
    "    tmp_df.reset_index(inplace=True, drop=True)\n",
    "    blast_out_dict[tmp_key] = tmp_df.iloc[:,:]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "for folder in ass_folders[:1]:\n",
    "    '''\n",
    "    From oriented_coords.csv files generate gff files with the following set up.\n",
    "    h_contig overlap\n",
    "    'query', 'ref', 'h_feature', 'query_start', 'query_end', 'query_aln_len', 'strand', 'frame', 'attribute_h'\n",
    "    p_contig overlap\n",
    "    'ref', 'query', 'p_feature','ref_start','ref_end', 'alignment_length', 'strand', 'frame', 'attribute_p'.\n",
    "    Save those file in a new folder for further downstream analysis. The same folder should include the contig and gene filtered\n",
    "    gffs. \n",
    "    '''\n",
    "    orient_coords_file = [os.path.join(folder, x) for x in os.listdir(folder) if x.endswith('oriented_coords.csv')][0]\n",
    "    #load in df and generate several additions columns\n",
    "    tmp_df = pd.read_csv(orient_coords_file, sep=',')\n",
    "    tmp_df['p_feature'] = \"haplotig\"\n",
    "    tmp_df['h_feature'] ='primary_contig'\n",
    "    tmp_df['strand'] = \"+\"\n",
    "    tmp_df['frame'] = 0\n",
    "    tmp_df['query_aln_len'] = abs(tmp_df['query_end']-tmp_df['query_start'])\n",
    "    tmp_df['alignment_length'] = abs(tmp_df['ref_end'] - tmp_df['ref_start'])\n",
    "    tmp_df.reset_index(drop=True, inplace=True)\n",
    "    tmp_df.sort_values('query', inplace=True)\n",
    "    tmp_df['attribute_p'] = tmp_df.apply(lambda row: att_column_p(row['query'],row['query_start'], row['query_end'],\\\n",
    "                                                         row['query_length'], row['query_aln_len']), axis=1)\n",
    "    \n",
    "    tmp_df['attribute_h'] = tmp_df.apply(lambda row: att_column_h(row['ref'], row['ref_start'], row['ref_end'],\\\n",
    "                                                                  row['ref_length'], row['alignment_length']), axis=1)\n",
    "    \n",
    "    #generate tmp gff dataframe\n",
    "    tmp_p_gff_df = tmp_df.loc[:, ['ref', 'query', 'p_feature','ref_start','ref_end', 'alignment_length', 'strand', 'frame', 'attribute_p']]\n",
    "    tmp_p_gff_df.sort_values('ref_start',inplace=True)\n",
    "    tmp_h_gff_df = tmp_df.loc[:, ['query', 'ref', 'h_feature', 'query_start', 'query_end', 'query_aln_len', 'strand', 'frame', 'attribute_h']]\n",
    "    tmp_h_gff_df.sort_values('query_start', inplace=True)\n",
    "    #make the outfolder and save gff files of overlap and filtered gene files\n",
    "    folder_suffix = folder.split('/')[-1]\n",
    "    #get the contig\n",
    "    contig = re.search(r'p([a-z]*_[0-9]*)_', folder_suffix).group(1)\n",
    "    out_folder = os.path.join(OUT_PATH, folder_suffix)\n",
    "    if not os.path.exists(out_folder):\n",
    "        os.mkdir(out_folder)\n",
    "    out_name_p = os.path.join(out_folder, folder_suffix )\n",
    "    tmp_p_gff_df.to_csv(out_name_p+'.p_by_h_cov.gff', sep='\\t' ,header=None, index =None)\n",
    "    out_name_h =  os.path.join(out_folder, folder_suffix)\n",
    "    tmp_h_gff_df.to_csv(out_name_h+'.h_by_p_cov.gff' , sep='\\t' ,header=None, index =None)\n",
    "    #write those out to new folder for php together with the p and h gff of genes\n",
    "    p_gene_gff = gene_contig_subset(p_gff, contig, 'gene')\n",
    "    p_gene_gff.to_csv(out_name_p+'.p_gene.gff', sep='\\t' ,header=None, index =None)\n",
    "    h_gene_gff = gene_contig_subset(h_gff, contig, 'gene')\n",
    "    h_gene_gff.to_csv(out_name_h+'.h_gene.gff', sep='\\t' ,header=None, index =None)\n",
    "    \n",
    "    #next would be to do overlap between the gene gff and the corresponding alingment gff. Write this out. Dict for each gene and its corresponding h_contig/p_contig\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#now loop over the outfolders\n",
    "out_folders = [os.path.join(OUT_PATH, x) for x in os.listdir(OUT_PATH) if x.endswith('_php_8kbp')]\n",
    "out_folders.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Pst_104E_v12_pcontig_000_php_8kbp.h_by_p_cov.gff', 'Pst_104E_v12_pcontig_000_php_8kbp.h_gene.gff', 'Pst_104E_v12_pcontig_000_php_8kbp.p_by_h_cov.gff', 'Pst_104E_v12_pcontig_000_php_8kbp.p_gene.gff']\n"
     ]
    }
   ],
   "source": [
    "for out_folder in out_folders:\n",
    "    tmp_folder_content = os.listdir(out_folder)\n",
    "    tmp_p_gene_gff = [os.path.join(out_folder, x) for x in tmp_folder_content if x.endswith('p_gene.gff') ][0]\n",
    "    tmp_p_gene_bed = pybedtools.BedTool(tmp_p_gene_gff).remove_invalid().saveas()\n",
    "    tmp_p_by_h_gff = [os.path.join(out_folder, x) for x in tmp_folder_content if x.endswith('p_by_h_cov.gff') ][0]\n",
    "    tmp_p_by_h_bed = pybedtools.BedTool(tmp_p_by_h_gff).remove_invalid().saveas()\n",
    "    #generate a bed intersect\n",
    "    p_gene_and_p_by_h = tmp_p_gene_bed.intersect(tmp_p_by_h_bed, wb=True)\n",
    "    #turn this into a dataframe \n",
    "    p_gene_and_p_by_h_df = p_gene_and_p_by_h.to_dataframe()\n",
    "    p_gene_and_p_by_h_df['p_gene'] = p_gene_and_p_by_h_df[8].str.extract(r'ID=([^;]*);')\n",
    "    p_gene_and_p_by_h_df['p_protein'] = p_gene_and_p_by_h_df['p_gene'].str.replace('TU', 'model')\n",
    "    #safe this in the same folder as dataframe\n",
    "    p_gene_and_p_by_h_df.to_csv(tmp_p_gene_gff.replace('p_gene.gff', 'gene_haplotig_intersect.df'), sep='\\t' )\n",
    "    print(os.listdir(out_folder))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h = tmp_p_gene_bed.intersect(tmp_p_by_h_bed, wb=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/pybedtools/bedtool.py:3199: UserWarning: Default names for filetype gff are:\n",
      "['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
      "but file has 18 fields; you can supply custom names with the `names` kwarg\n",
      "  % (self.file_type, _names, self.field_count()))\n"
     ]
    }
   ],
   "source": [
    "p_gene_and_p_by_h_df = p_gene_and_p_by_h.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "p_gene_and_p_by_h_df['p_gene'] = p_gene_and_p_by_h_df[8].str.extract(r'ID=([^;]*);')\n",
    "p_gene_and_p_by_h_df['p_protein'] = p_gene_and_p_by_h_df['p_gene'].str.replace('TU', 'model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>p_gene</th>\n",
       "      <th>p_protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>5109</td>\n",
       "      <td>5788</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=evm.TU.pcontig_000.2;Name=gene_model_pconti...</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>haplotig</td>\n",
       "      <td>4096</td>\n",
       "      <td>6881</td>\n",
       "      <td>2785</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>h_contig=hcontig_000_003;query_start=1;query_s...</td>\n",
       "      <td>evm.TU.pcontig_000.2</td>\n",
       "      <td>evm.model.pcontig_000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>12149</td>\n",
       "      <td>13224</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=evm.TU.pcontig_000.4;Name=gene_model_pconti...</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>haplotig</td>\n",
       "      <td>10766</td>\n",
       "      <td>14593</td>\n",
       "      <td>3827</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>h_contig=hcontig_000_003;query_start=6520;quer...</td>\n",
       "      <td>evm.TU.pcontig_000.4</td>\n",
       "      <td>evm.model.pcontig_000.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>22054</td>\n",
       "      <td>23784</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=evm.TU.pcontig_000.5;Name=gene_model_pconti...</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>haplotig</td>\n",
       "      <td>21297</td>\n",
       "      <td>24866</td>\n",
       "      <td>3569</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>h_contig=hcontig_000_031;query_start=1;query_s...</td>\n",
       "      <td>evm.TU.pcontig_000.5</td>\n",
       "      <td>evm.model.pcontig_000.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>22054</td>\n",
       "      <td>23784</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=evm.TU.pcontig_000.5;Name=gene_model_pconti...</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>haplotig</td>\n",
       "      <td>14589</td>\n",
       "      <td>65671</td>\n",
       "      <td>51082</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>h_contig=hcontig_000_003;query_start=19775;que...</td>\n",
       "      <td>evm.TU.pcontig_000.5</td>\n",
       "      <td>evm.model.pcontig_000.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>28352</td>\n",
       "      <td>29795</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=evm.TU.pcontig_000.6;Name=gene_model_pconti...</td>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>haplotig</td>\n",
       "      <td>14589</td>\n",
       "      <td>65671</td>\n",
       "      <td>51082</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>h_contig=hcontig_000_003;query_start=19775;que...</td>\n",
       "      <td>evm.TU.pcontig_000.6</td>\n",
       "      <td>evm.model.pcontig_000.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0    1     2      3      4  5  6  7  \\\n",
       "0  pcontig_000  EVM  gene   5109   5788  .  +  .   \n",
       "1  pcontig_000  EVM  gene  12149  13224  .  -  .   \n",
       "2  pcontig_000  EVM  gene  22054  23784  .  -  .   \n",
       "3  pcontig_000  EVM  gene  22054  23784  .  -  .   \n",
       "4  pcontig_000  EVM  gene  28352  29795  .  +  .   \n",
       "\n",
       "                                                   8            9  \\\n",
       "0  ID=evm.TU.pcontig_000.2;Name=gene_model_pconti...  pcontig_000   \n",
       "1  ID=evm.TU.pcontig_000.4;Name=gene_model_pconti...  pcontig_000   \n",
       "2  ID=evm.TU.pcontig_000.5;Name=gene_model_pconti...  pcontig_000   \n",
       "3  ID=evm.TU.pcontig_000.5;Name=gene_model_pconti...  pcontig_000   \n",
       "4  ID=evm.TU.pcontig_000.6;Name=gene_model_pconti...  pcontig_000   \n",
       "\n",
       "                10        11     12     13     14 15  16  \\\n",
       "0  hcontig_000_003  haplotig   4096   6881   2785  +   0   \n",
       "1  hcontig_000_003  haplotig  10766  14593   3827  +   0   \n",
       "2  hcontig_000_031  haplotig  21297  24866   3569  +   0   \n",
       "3  hcontig_000_003  haplotig  14589  65671  51082  +   0   \n",
       "4  hcontig_000_003  haplotig  14589  65671  51082  +   0   \n",
       "\n",
       "                                                  17                p_gene  \\\n",
       "0  h_contig=hcontig_000_003;query_start=1;query_s...  evm.TU.pcontig_000.2   \n",
       "1  h_contig=hcontig_000_003;query_start=6520;quer...  evm.TU.pcontig_000.4   \n",
       "2  h_contig=hcontig_000_031;query_start=1;query_s...  evm.TU.pcontig_000.5   \n",
       "3  h_contig=hcontig_000_003;query_start=19775;que...  evm.TU.pcontig_000.5   \n",
       "4  h_contig=hcontig_000_003;query_start=19775;que...  evm.TU.pcontig_000.6   \n",
       "\n",
       "                 p_protein  \n",
       "0  evm.model.pcontig_000.2  \n",
       "1  evm.model.pcontig_000.4  \n",
       "2  evm.model.pcontig_000.5  \n",
       "3  evm.model.pcontig_000.5  \n",
       "4  evm.model.pcontig_000.6  "
      ]
     },
     "execution_count": 464,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_gene_and_p_by_h_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_gene_and_p_by_h_df_2 = p_gene_and_p_by_h_df.loc[:, [10, 'p_protein']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10</th>\n",
       "      <th>p_protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>hcontig_000_031</td>\n",
       "      <td>evm.model.pcontig_000.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>hcontig_000_003</td>\n",
       "      <td>evm.model.pcontig_000.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>563</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>hcontig_000_054</td>\n",
       "      <td>evm.model.pcontig_000.684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>588 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  10                  p_protein\n",
       "0    hcontig_000_003    evm.model.pcontig_000.2\n",
       "1    hcontig_000_003    evm.model.pcontig_000.4\n",
       "2    hcontig_000_031    evm.model.pcontig_000.5\n",
       "3    hcontig_000_003    evm.model.pcontig_000.5\n",
       "4    hcontig_000_003    evm.model.pcontig_000.6\n",
       "5    hcontig_000_031    evm.model.pcontig_000.6\n",
       "6    hcontig_000_003    evm.model.pcontig_000.7\n",
       "7    hcontig_000_031    evm.model.pcontig_000.7\n",
       "8    hcontig_000_003    evm.model.pcontig_000.8\n",
       "9    hcontig_000_031    evm.model.pcontig_000.8\n",
       "10   hcontig_000_003    evm.model.pcontig_000.9\n",
       "11   hcontig_000_031    evm.model.pcontig_000.9\n",
       "12   hcontig_000_003   evm.model.pcontig_000.10\n",
       "13   hcontig_000_031   evm.model.pcontig_000.10\n",
       "14   hcontig_000_003   evm.model.pcontig_000.11\n",
       "15   hcontig_000_031   evm.model.pcontig_000.11\n",
       "16   hcontig_000_003   evm.model.pcontig_000.12\n",
       "17   hcontig_000_031   evm.model.pcontig_000.12\n",
       "18   hcontig_000_003   evm.model.pcontig_000.13\n",
       "19   hcontig_000_003   evm.model.pcontig_000.14\n",
       "20   hcontig_000_003   evm.model.pcontig_000.15\n",
       "21   hcontig_000_003   evm.model.pcontig_000.16\n",
       "22   hcontig_000_003   evm.model.pcontig_000.17\n",
       "23   hcontig_000_003   evm.model.pcontig_000.18\n",
       "24   hcontig_000_003   evm.model.pcontig_000.20\n",
       "25   hcontig_000_003   evm.model.pcontig_000.21\n",
       "26   hcontig_000_003   evm.model.pcontig_000.22\n",
       "27   hcontig_000_003   evm.model.pcontig_000.23\n",
       "28   hcontig_000_003   evm.model.pcontig_000.24\n",
       "29   hcontig_000_003   evm.model.pcontig_000.25\n",
       "..               ...                        ...\n",
       "558  hcontig_000_054  evm.model.pcontig_000.652\n",
       "559  hcontig_000_054  evm.model.pcontig_000.653\n",
       "560  hcontig_000_054  evm.model.pcontig_000.654\n",
       "561  hcontig_000_054  evm.model.pcontig_000.655\n",
       "562  hcontig_000_054  evm.model.pcontig_000.656\n",
       "563  hcontig_000_054  evm.model.pcontig_000.657\n",
       "564  hcontig_000_054  evm.model.pcontig_000.658\n",
       "565  hcontig_000_054  evm.model.pcontig_000.659\n",
       "566  hcontig_000_054  evm.model.pcontig_000.662\n",
       "567  hcontig_000_054  evm.model.pcontig_000.663\n",
       "568  hcontig_000_054  evm.model.pcontig_000.664\n",
       "569  hcontig_000_054  evm.model.pcontig_000.665\n",
       "570  hcontig_000_054  evm.model.pcontig_000.666\n",
       "571  hcontig_000_054  evm.model.pcontig_000.667\n",
       "572  hcontig_000_054  evm.model.pcontig_000.668\n",
       "573  hcontig_000_054  evm.model.pcontig_000.670\n",
       "574  hcontig_000_054  evm.model.pcontig_000.671\n",
       "575  hcontig_000_054  evm.model.pcontig_000.672\n",
       "576  hcontig_000_054  evm.model.pcontig_000.673\n",
       "577  hcontig_000_054  evm.model.pcontig_000.674\n",
       "578  hcontig_000_054  evm.model.pcontig_000.675\n",
       "579  hcontig_000_054  evm.model.pcontig_000.676\n",
       "580  hcontig_000_054  evm.model.pcontig_000.677\n",
       "581  hcontig_000_054  evm.model.pcontig_000.678\n",
       "582  hcontig_000_054  evm.model.pcontig_000.679\n",
       "583  hcontig_000_054  evm.model.pcontig_000.680\n",
       "584  hcontig_000_054  evm.model.pcontig_000.681\n",
       "585  hcontig_000_054  evm.model.pcontig_000.682\n",
       "586  hcontig_000_054  evm.model.pcontig_000.683\n",
       "587  hcontig_000_054  evm.model.pcontig_000.684\n",
       "\n",
       "[588 rows x 2 columns]"
      ]
     },
     "execution_count": 466,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_gene_and_p_by_h_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "contig_000_df['On p_by_h'] = contig_000_df.apply(lambda row: haplotig_checker(row['Query'], row['Target']), axis =1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def haplotig_checker(Query, Target):\n",
    "    '''Check if protein blast hit lies on the corresponding h_contig'''\n",
    "    _tmp_df = p_gene_and_p_by_h_df_2[p_gene_and_p_by_h_df_2['p_protein'] == Query ]\n",
    "    if Target == 'NaN':\n",
    "        return False\n",
    "    else:\n",
    "        h_contig = re.search(r'(hcontig_[^.]*).', Target).group(1)\n",
    "        if sum(_tmp_df[10].str.contains(h_contig)) > 0:\n",
    "               return True\n",
    "        else:\n",
    "               return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>Target</th>\n",
       "      <th>PctID</th>\n",
       "      <th>AlnLgth</th>\n",
       "      <th>NumMis</th>\n",
       "      <th>NumGap</th>\n",
       "      <th>StartQuery</th>\n",
       "      <th>StopQuery</th>\n",
       "      <th>StartTarget</th>\n",
       "      <th>StopTarget</th>\n",
       "      <th>e-value</th>\n",
       "      <th>BitScore</th>\n",
       "      <th>QLgth</th>\n",
       "      <th>QCov</th>\n",
       "      <th>q_contig == s_contig</th>\n",
       "      <th>On p_by_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_031.5</td>\n",
       "      <td>70.19</td>\n",
       "      <td>161</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>1</td>\n",
       "      <td>161</td>\n",
       "      <td>4e-69</td>\n",
       "      <td>207</td>\n",
       "      <td>114</td>\n",
       "      <td>141.228</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_003.11</td>\n",
       "      <td>59.68</td>\n",
       "      <td>124</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1e-38</td>\n",
       "      <td>130</td>\n",
       "      <td>114</td>\n",
       "      <td>108.772</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.107</td>\n",
       "      <td>99.21</td>\n",
       "      <td>1141</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>0</td>\n",
       "      <td>2330</td>\n",
       "      <td>1144</td>\n",
       "      <td>99.7378</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.112</td>\n",
       "      <td>72.12</td>\n",
       "      <td>104</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>473</td>\n",
       "      <td>576</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>1e-38</td>\n",
       "      <td>142</td>\n",
       "      <td>1144</td>\n",
       "      <td>9.09091</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>evm.model.pcontig_000.101</td>\n",
       "      <td>evm.model.hcontig_000_003.108</td>\n",
       "      <td>89.03</td>\n",
       "      <td>237</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>231</td>\n",
       "      <td>1</td>\n",
       "      <td>237</td>\n",
       "      <td>2e-150</td>\n",
       "      <td>421</td>\n",
       "      <td>231</td>\n",
       "      <td>102.597</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Query                         Target  PctID AlnLgth  \\\n",
       "0   evm.model.pcontig_000.10    evm.model.hcontig_000_031.5  70.19     161   \n",
       "1   evm.model.pcontig_000.10   evm.model.hcontig_000_003.11  59.68     124   \n",
       "2  evm.model.pcontig_000.100  evm.model.hcontig_000_003.107  99.21    1141   \n",
       "3  evm.model.pcontig_000.100  evm.model.hcontig_000_003.112  72.12     104   \n",
       "4  evm.model.pcontig_000.101  evm.model.hcontig_000_003.108  89.03     237   \n",
       "\n",
       "  NumMis NumGap StartQuery StopQuery StartTarget StopTarget e-value BitScore  \\\n",
       "0      1      1          1       114           1        161   4e-69      207   \n",
       "1      3      1          1        77           1        124   1e-38      130   \n",
       "2      9      0          1      1141           1       1141       0     2330   \n",
       "3     13      1        473       576           4         91   1e-38      142   \n",
       "4     20      1          1       231           1        237  2e-150      421   \n",
       "\n",
       "   QLgth     QCov q_contig == s_contig On p_by_h  \n",
       "0    114  141.228                 True      True  \n",
       "1    114  108.772                 True      True  \n",
       "2   1144  99.7378                 True      True  \n",
       "3   1144  9.09091                 True      True  \n",
       "4    231  102.597                 True      True  "
      ]
     },
     "execution_count": 497,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_000_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'hcontig_000_031'"
      ]
     },
     "execution_count": 488,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.search(r'(hcontig[^.]*).','evm.model.hcontig_000_031.5').group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "grouby = [(p_gene_and_p_by_h_df.groupby('p_gene').count()  > 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[                            0      1      2      3      4      5      6  \\\n",
       " p_gene                                                                    \n",
       " evm.TU.pcontig_000.10    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.100  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.101  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.102  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.103  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.104  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.105  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.108  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.109  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.11    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.110  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.111  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.112  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.113  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.114  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.115  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.116  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.117  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.12    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.13   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.132  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.133  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.134  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.135  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.136  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.137  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.138  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.139  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.14   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.140  False  False  False  False  False  False  False   \n",
       " ...                       ...    ...    ...    ...    ...    ...    ...   \n",
       " evm.TU.pcontig_000.69   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.7     True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.70   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.72   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.73   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.74   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.75   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.77   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.78   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.8     True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.81   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.82   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.83   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.84   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.85   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.86   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.87   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.88   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.89   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.9     True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.90   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.91   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.92   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.93   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.94   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.95   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.96    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.97   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.98   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.99   False  False  False  False  False  False  False   \n",
       " \n",
       "                             7      8      9     10     11     12     13  \\\n",
       " p_gene                                                                    \n",
       " evm.TU.pcontig_000.10    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.100  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.101  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.102  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.103  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.104  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.105  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.108  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.109  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.11    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.110  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.111  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.112  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.113  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.114  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.115  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.116  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.117  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.12    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.13   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.132  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.133  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.134  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.135  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.136  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.137  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.138  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.139  False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.14   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.140  False  False  False  False  False  False  False   \n",
       " ...                       ...    ...    ...    ...    ...    ...    ...   \n",
       " evm.TU.pcontig_000.69   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.7     True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.70   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.72   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.73   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.74   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.75   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.77   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.78   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.8     True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.81   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.82   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.83   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.84   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.85   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.86   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.87   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.88   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.89   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.9     True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.90   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.91   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.92   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.93   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.94   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.95   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.96    True   True   True   True   True   True   True   \n",
       " evm.TU.pcontig_000.97   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.98   False  False  False  False  False  False  False   \n",
       " evm.TU.pcontig_000.99   False  False  False  False  False  False  False   \n",
       " \n",
       "                            14     15     16     17 p_protein  \n",
       " p_gene                                                        \n",
       " evm.TU.pcontig_000.10    True   True   True   True      True  \n",
       " evm.TU.pcontig_000.100  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.101  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.102  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.103  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.104  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.105  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.108  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.109  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.11    True   True   True   True      True  \n",
       " evm.TU.pcontig_000.110  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.111  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.112  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.113  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.114  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.115  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.116  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.117  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.12    True   True   True   True      True  \n",
       " evm.TU.pcontig_000.13   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.132  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.133  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.134  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.135  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.136  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.137  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.138  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.139  False  False  False  False     False  \n",
       " evm.TU.pcontig_000.14   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.140  False  False  False  False     False  \n",
       " ...                       ...    ...    ...    ...       ...  \n",
       " evm.TU.pcontig_000.69   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.7     True   True   True   True      True  \n",
       " evm.TU.pcontig_000.70   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.72   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.73   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.74   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.75   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.77   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.78   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.8     True   True   True   True      True  \n",
       " evm.TU.pcontig_000.81   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.82   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.83   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.84   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.85   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.86   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.87   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.88   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.89   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.9     True   True   True   True      True  \n",
       " evm.TU.pcontig_000.90   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.91   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.92   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.93   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.94   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.95   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.96    True   True   True   True      True  \n",
       " evm.TU.pcontig_000.97   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.98   False  False  False  False     False  \n",
       " evm.TU.pcontig_000.99   False  False  False  False     False  \n",
       " \n",
       " [522 rows x 19 columns]]"
      ]
     },
     "execution_count": 462,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouby #[grouby[0] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "522"
      ]
     },
     "execution_count": 435,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(p_gene_and_p_by_h_df['p_gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_gene_to_haplotig_dict = p_gene_and_p_by_h_df['p']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df = blast_out_dict['Pst_104E_v12_p_ctg.Pst_104E_v12_h_ctg.0.001.blastp.outfmt6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contig_000_df = contig_000_df[contig_000_df['Query'].str.contains('pcontig_000')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>Target</th>\n",
       "      <th>PctID</th>\n",
       "      <th>AlnLgth</th>\n",
       "      <th>NumMis</th>\n",
       "      <th>NumGap</th>\n",
       "      <th>StartQuery</th>\n",
       "      <th>StopQuery</th>\n",
       "      <th>StartTarget</th>\n",
       "      <th>StopTarget</th>\n",
       "      <th>e-value</th>\n",
       "      <th>BitScore</th>\n",
       "      <th>QLgth</th>\n",
       "      <th>QCov</th>\n",
       "      <th>q_contig == s_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_031.5</td>\n",
       "      <td>70.19</td>\n",
       "      <td>161</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>1</td>\n",
       "      <td>161</td>\n",
       "      <td>4e-69</td>\n",
       "      <td>207</td>\n",
       "      <td>114</td>\n",
       "      <td>141.228</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_003.11</td>\n",
       "      <td>59.68</td>\n",
       "      <td>124</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1e-38</td>\n",
       "      <td>130</td>\n",
       "      <td>114</td>\n",
       "      <td>108.772</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.107</td>\n",
       "      <td>99.21</td>\n",
       "      <td>1141</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>0</td>\n",
       "      <td>2330</td>\n",
       "      <td>1144</td>\n",
       "      <td>99.7378</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.112</td>\n",
       "      <td>72.12</td>\n",
       "      <td>104</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>473</td>\n",
       "      <td>576</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>1e-38</td>\n",
       "      <td>142</td>\n",
       "      <td>1144</td>\n",
       "      <td>9.09091</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>evm.model.pcontig_000.101</td>\n",
       "      <td>evm.model.hcontig_000_003.108</td>\n",
       "      <td>89.03</td>\n",
       "      <td>237</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>231</td>\n",
       "      <td>1</td>\n",
       "      <td>237</td>\n",
       "      <td>2e-150</td>\n",
       "      <td>421</td>\n",
       "      <td>231</td>\n",
       "      <td>102.597</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Query                         Target  PctID AlnLgth  \\\n",
       "0   evm.model.pcontig_000.10    evm.model.hcontig_000_031.5  70.19     161   \n",
       "1   evm.model.pcontig_000.10   evm.model.hcontig_000_003.11  59.68     124   \n",
       "2  evm.model.pcontig_000.100  evm.model.hcontig_000_003.107  99.21    1141   \n",
       "3  evm.model.pcontig_000.100  evm.model.hcontig_000_003.112  72.12     104   \n",
       "4  evm.model.pcontig_000.101  evm.model.hcontig_000_003.108  89.03     237   \n",
       "\n",
       "  NumMis NumGap StartQuery StopQuery StartTarget StopTarget e-value BitScore  \\\n",
       "0      1      1          1       114           1        161   4e-69      207   \n",
       "1      3      1          1        77           1        124   1e-38      130   \n",
       "2      9      0          1      1141           1       1141       0     2330   \n",
       "3     13      1        473       576           4         91   1e-38      142   \n",
       "4     20      1          1       231           1        237  2e-150      421   \n",
       "\n",
       "   QLgth     QCov q_contig == s_contig  \n",
       "0    114  141.228                 True  \n",
       "1    114  108.772                 True  \n",
       "2   1144  99.7378                 True  \n",
       "3   1144  9.09091                 True  \n",
       "4    231  102.597                 True  "
      ]
     },
     "execution_count": 463,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_000_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>Target</th>\n",
       "      <th>PctID</th>\n",
       "      <th>AlnLgth</th>\n",
       "      <th>NumMis</th>\n",
       "      <th>NumGap</th>\n",
       "      <th>StartQuery</th>\n",
       "      <th>StopQuery</th>\n",
       "      <th>StartTarget</th>\n",
       "      <th>StopTarget</th>\n",
       "      <th>e-value</th>\n",
       "      <th>BitScore</th>\n",
       "      <th>QLgth</th>\n",
       "      <th>QCov</th>\n",
       "      <th>q_contig == s_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_031.5</td>\n",
       "      <td>70.19</td>\n",
       "      <td>161</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>1</td>\n",
       "      <td>161</td>\n",
       "      <td>4e-69</td>\n",
       "      <td>207</td>\n",
       "      <td>114</td>\n",
       "      <td>141.228</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>evm.model.pcontig_000.10</td>\n",
       "      <td>evm.model.hcontig_000_003.11</td>\n",
       "      <td>59.68</td>\n",
       "      <td>124</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1e-38</td>\n",
       "      <td>130</td>\n",
       "      <td>114</td>\n",
       "      <td>108.772</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.107</td>\n",
       "      <td>99.21</td>\n",
       "      <td>1141</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>0</td>\n",
       "      <td>2330</td>\n",
       "      <td>1144</td>\n",
       "      <td>99.7378</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>evm.model.pcontig_000.100</td>\n",
       "      <td>evm.model.hcontig_000_003.112</td>\n",
       "      <td>72.12</td>\n",
       "      <td>104</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>473</td>\n",
       "      <td>576</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>1e-38</td>\n",
       "      <td>142</td>\n",
       "      <td>1144</td>\n",
       "      <td>9.09091</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>evm.model.pcontig_000.101</td>\n",
       "      <td>evm.model.hcontig_000_003.108</td>\n",
       "      <td>89.03</td>\n",
       "      <td>237</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>231</td>\n",
       "      <td>1</td>\n",
       "      <td>237</td>\n",
       "      <td>2e-150</td>\n",
       "      <td>421</td>\n",
       "      <td>231</td>\n",
       "      <td>102.597</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>evm.model.pcontig_000.101</td>\n",
       "      <td>evm.model.hcontig_000_003.146</td>\n",
       "      <td>69.2</td>\n",
       "      <td>237</td>\n",
       "      <td>51</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>230</td>\n",
       "      <td>28</td>\n",
       "      <td>258</td>\n",
       "      <td>5e-97</td>\n",
       "      <td>305</td>\n",
       "      <td>231</td>\n",
       "      <td>102.597</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>evm.model.pcontig_000.102</td>\n",
       "      <td>evm.model.hcontig_000_003.109</td>\n",
       "      <td>100</td>\n",
       "      <td>107</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "      <td>7e-73</td>\n",
       "      <td>214</td>\n",
       "      <td>107</td>\n",
       "      <td>100</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>evm.model.pcontig_000.102</td>\n",
       "      <td>evm.model.hcontig_000_003.146</td>\n",
       "      <td>81.9</td>\n",
       "      <td>105</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>348</td>\n",
       "      <td>452</td>\n",
       "      <td>4e-49</td>\n",
       "      <td>167</td>\n",
       "      <td>107</td>\n",
       "      <td>98.1308</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>evm.model.pcontig_000.103</td>\n",
       "      <td>evm.model.hcontig_000_003.110</td>\n",
       "      <td>97.73</td>\n",
       "      <td>88</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>2e-57</td>\n",
       "      <td>174</td>\n",
       "      <td>95</td>\n",
       "      <td>92.6316</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>evm.model.pcontig_000.103</td>\n",
       "      <td>evm.model.hcontig_000_003.162</td>\n",
       "      <td>83.82</td>\n",
       "      <td>68</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>73</td>\n",
       "      <td>85</td>\n",
       "      <td>152</td>\n",
       "      <td>5e-36</td>\n",
       "      <td>127</td>\n",
       "      <td>95</td>\n",
       "      <td>71.5789</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>evm.model.pcontig_000.104</td>\n",
       "      <td>evm.model.hcontig_000_003.111</td>\n",
       "      <td>98.53</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>3e-43</td>\n",
       "      <td>136</td>\n",
       "      <td>68</td>\n",
       "      <td>100</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>evm.model.pcontig_000.105</td>\n",
       "      <td>evm.model.hcontig_000_003.112</td>\n",
       "      <td>100</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1e-85</td>\n",
       "      <td>248</td>\n",
       "      <td>122</td>\n",
       "      <td>100</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>evm.model.pcontig_000.105</td>\n",
       "      <td>evm.model.hcontig_000_003.107</td>\n",
       "      <td>73.08</td>\n",
       "      <td>104</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>473</td>\n",
       "      <td>576</td>\n",
       "      <td>1e-40</td>\n",
       "      <td>144</td>\n",
       "      <td>122</td>\n",
       "      <td>85.2459</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>evm.model.pcontig_000.108</td>\n",
       "      <td>evm.model.hcontig_000_003.117</td>\n",
       "      <td>99.27</td>\n",
       "      <td>682</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>0</td>\n",
       "      <td>1392</td>\n",
       "      <td>682</td>\n",
       "      <td>100</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>evm.model.pcontig_000.108</td>\n",
       "      <td>evm.model.hcontig_000_003.133</td>\n",
       "      <td>26.05</td>\n",
       "      <td>119</td>\n",
       "      <td>80</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>128</td>\n",
       "      <td>60</td>\n",
       "      <td>170</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>47.4</td>\n",
       "      <td>682</td>\n",
       "      <td>17.4487</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>evm.model.pcontig_000.109</td>\n",
       "      <td>evm.model.hcontig_000_003.117</td>\n",
       "      <td>99.27</td>\n",
       "      <td>682</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>1</td>\n",
       "      <td>682</td>\n",
       "      <td>0</td>\n",
       "      <td>1392</td>\n",
       "      <td>682</td>\n",
       "      <td>100</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>evm.model.pcontig_000.109</td>\n",
       "      <td>evm.model.hcontig_000_003.133</td>\n",
       "      <td>26.05</td>\n",
       "      <td>119</td>\n",
       "      <td>80</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>128</td>\n",
       "      <td>60</td>\n",
       "      <td>170</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>47.4</td>\n",
       "      <td>682</td>\n",
       "      <td>17.4487</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Query                         Target  PctID AlnLgth  \\\n",
       "0    evm.model.pcontig_000.10    evm.model.hcontig_000_031.5  70.19     161   \n",
       "1    evm.model.pcontig_000.10   evm.model.hcontig_000_003.11  59.68     124   \n",
       "2   evm.model.pcontig_000.100  evm.model.hcontig_000_003.107  99.21    1141   \n",
       "3   evm.model.pcontig_000.100  evm.model.hcontig_000_003.112  72.12     104   \n",
       "4   evm.model.pcontig_000.101  evm.model.hcontig_000_003.108  89.03     237   \n",
       "5   evm.model.pcontig_000.101  evm.model.hcontig_000_003.146   69.2     237   \n",
       "6   evm.model.pcontig_000.102  evm.model.hcontig_000_003.109    100     107   \n",
       "7   evm.model.pcontig_000.102  evm.model.hcontig_000_003.146   81.9     105   \n",
       "8   evm.model.pcontig_000.103  evm.model.hcontig_000_003.110  97.73      88   \n",
       "9   evm.model.pcontig_000.103  evm.model.hcontig_000_003.162  83.82      68   \n",
       "10  evm.model.pcontig_000.104  evm.model.hcontig_000_003.111  98.53      68   \n",
       "11  evm.model.pcontig_000.105  evm.model.hcontig_000_003.112    100     122   \n",
       "12  evm.model.pcontig_000.105  evm.model.hcontig_000_003.107  73.08     104   \n",
       "13  evm.model.pcontig_000.108  evm.model.hcontig_000_003.117  99.27     682   \n",
       "50  evm.model.pcontig_000.108  evm.model.hcontig_000_003.133  26.05     119   \n",
       "55  evm.model.pcontig_000.109  evm.model.hcontig_000_003.117  99.27     682   \n",
       "92  evm.model.pcontig_000.109  evm.model.hcontig_000_003.133  26.05     119   \n",
       "\n",
       "   NumMis NumGap StartQuery StopQuery StartTarget StopTarget e-value BitScore  \\\n",
       "0       1      1          1       114           1        161   4e-69      207   \n",
       "1       3      1          1        77           1        124   1e-38      130   \n",
       "2       9      0          1      1141           1       1141       0     2330   \n",
       "3      13      1        473       576           4         91   1e-38      142   \n",
       "4      20      1          1       231           1        237  2e-150      421   \n",
       "5      51      6         10       230          28        258   5e-97      305   \n",
       "6       0      0          1       107           1        107   7e-73      214   \n",
       "7      19      0          1       105         348        452   4e-49      167   \n",
       "8       2      0          8        95           1         88   2e-57      174   \n",
       "9      11      0          6        73          85        152   5e-36      127   \n",
       "10      1      0          1        68           1         68   3e-43      136   \n",
       "11      0      0          1       122           1        122   1e-85      248   \n",
       "12     12      1          4        91         473        576   1e-40      144   \n",
       "13      5      0          1       682           1        682       0     1392   \n",
       "50     80      2         10       128          60        170   1e-05     47.4   \n",
       "55      5      0          1       682           1        682       0     1392   \n",
       "92     80      2         10       128          60        170   1e-05     47.4   \n",
       "\n",
       "    QLgth     QCov q_contig == s_contig  \n",
       "0     114  141.228                 True  \n",
       "1     114  108.772                 True  \n",
       "2    1144  99.7378                 True  \n",
       "3    1144  9.09091                 True  \n",
       "4     231  102.597                 True  \n",
       "5     231  102.597                 True  \n",
       "6     107      100                 True  \n",
       "7     107  98.1308                 True  \n",
       "8      95  92.6316                 True  \n",
       "9      95  71.5789                 True  \n",
       "10     68      100                 True  \n",
       "11    122      100                 True  \n",
       "12    122  85.2459                 True  \n",
       "13    682      100                 True  \n",
       "50    682  17.4487                 True  \n",
       "55    682      100                 True  \n",
       "92    682  17.4487                 True  "
      ]
     },
     "execution_count": 455,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_000_df[(contig_000_df['Query'].str.contains('pcontig_000.10')) & (contig_000_df['Target'].str.contains('hcontig_000'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def feature_specific_haplotig(x,y df):\n",
    "    '''Transfers the specific haplotig that overlaps with the gene sequence'''\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
