{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/benjamin/anaconda3/lib/python3.5/site-packages/Bio/SearchIO/__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.\n",
      "  BiopythonExperimentalWarning)\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SearchIO\n",
    "from pybedtools import BedTool\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import multiprocessing\n",
    "import re\n",
    "import time\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is a restart of the TE_filtering_and_summary_p_contigs notebook. It is aimed at adding two columns to the TE summary dataframe.\n",
    "* age of TE\n",
    "* classification\n",
    "\n",
    "The age should be plotted out by superfamily and the new dataframe provided for the paper as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ID_filter_gff(_feature, _id):\n",
    "    \"\"\"\n",
    "    This filter parses out the top level id form the 9th gff column form a REPET gff file.\n",
    "    It has a specific search pattern for each feature type in column 2.\n",
    "    _type is defined by the feature '_'.join(feature.split(\"_\")[-2:])\n",
    "    This function expects that the variable genome either ends with p_ctg or h_ctg and adapts the\n",
    "    search pattern accordingly.\n",
    "    \"\"\"\n",
    "    _type = '_'.join(_feature.split(\"_\")[-2:])\n",
    "    if _type == 'REPET_TEs':\n",
    "        if genome.endswith('p_ctg'):\n",
    "            TE_pattern = r'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([^;| ]*)'\n",
    "        elif genome.endswith('h_ctg'):\n",
    "            TE_pattern = r'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_[0-9]*_([^;| ]*)'\n",
    "        TE_prog = re.compile(TE_pattern)\n",
    "        TE_match = TE_prog.search(_id)\n",
    "\n",
    "        try:\n",
    "            return TE_match.group(1)\n",
    "        except AttributeError:\n",
    "            print(_id)\n",
    "\n",
    "    if _type == 'REPET_SSRs':\n",
    "        if genome.endswith('p_ctg'):\n",
    "            SSR_pattern = 'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([A-Z,a-z,0-9,-]*)'\n",
    "        elif genome.endswith('h_ctg'):\n",
    "            SSR_pattern = 'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_[0-9]*_([A-Z,a-z,0-9,-]*)'\n",
    "        SSR_prog = re.compile(SSR_pattern)\n",
    "        SSR_match = SSR_prog.search(_id)\n",
    "        return SSR_match.group(1)\n",
    "    if _type == 'REPET_tblastx' or _type == 'REPET_blastx':\n",
    "        if genome.endswith('p_ctg'):\n",
    "            blast_prog = re.compile(r'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_([^;| ]*)')\n",
    "        elif genome.endswith('h_ctg'):\n",
    "             blast_prog = re.compile(r'ID=[A-Z,a-z,0-9,-]*_[A-Z,a-z,0-9]*_[0-9]*_[0-9]*_([^;| ]*)')\n",
    "        #blast_prog = re.compile(blast_pattern)\n",
    "        blast_match = blast_prog.search(_id)\n",
    "        return blast_match.group(1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def blast_hit_gff(_feature, _row8, _id):\n",
    "    \"\"\"\n",
    "    This filter parses the blast hit for REPET_TEs from the new 'ID' column. If no blast hit available returns Pastec ids.\n",
    "    If the result is blast already the value is simple parse the blast hit.\n",
    "    SSRs also get SSR\n",
    "    !!!Requires the three_letter_dict to be defined previously.!!!\n",
    "    _type is defined by the feature '_'.join(feature.split(\"_\")[-2:])\n",
    "    \"\"\"\n",
    "    _type = '_'.join(_feature.split(\"_\")[-2:])\n",
    "    if _type == 'REPET_TEs':\n",
    "        #split the pastec_cat into the first three letter code\n",
    "        #the spliting of the 'ID' column needs to be done differently depending on the h or p contigs.\n",
    "        #h contigs contain one additional '_' in the contig id\n",
    "\n",
    "        pastec_cat = _id.split('_')[0]\n",
    "        if 'TE_BLR' in _row8:\n",
    "            #hit_list = [x.split(';')[3] for x in _row8]\n",
    "            blast_hit_pattern = r'TE_BLR\\w*: (\\S*)[ |;]'\n",
    "            blast_hit_prog = re.compile(blast_hit_pattern)\n",
    "            TE_match = blast_hit_prog.findall(_row8)\n",
    "            first_sub_class = ':'.join(TE_match[0][:-1].split(':')[1:])\n",
    "            if len([x for x in TE_match if first_sub_class in x]) == len(TE_match):\n",
    "                if ';' in first_sub_class:\n",
    "                    return first_sub_class.split(';')[0]\n",
    "                else:\n",
    "                    return first_sub_class\n",
    "#fix this here to include the there letter code of the first bit of the ID similar to the blast hits\n",
    "#e.g. ClassI:?:? and so on. a dict might be the easiest here.\n",
    "            \n",
    "            else:\n",
    "                return three_letter_dict[pastec_cat]\n",
    "        else:\n",
    "            return three_letter_dict[pastec_cat]\n",
    "    if _type == 'REPET_SSRs':\n",
    "        return 'SSR'\n",
    "        \n",
    "\n",
    "        return SSR_match.group(1)\n",
    "    if _type == 'REPET_tblastx' or _type == 'REPET_blastx':\n",
    "        return ':'.join(_id.split(':')[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def TE_classification_filter(_id, level = 0):\n",
    "    \"\"\"\n",
    "    This function pulls out the class == level1, Order == level2, Superfamily == leve3.\n",
    "    If SSR or noCat return these values.\n",
    "    \n",
    "    \"\"\"\n",
    "    if len(_id.split(':')) == 1:\n",
    "        return _id\n",
    "    if level == 0:\n",
    "        _class = _id.split(':')[0]\n",
    "        if _class == 'ClassI':\n",
    "            return 'Retrotransposon'\n",
    "        if _class == 'ClassII':\n",
    "            return 'DNA_transposon'\n",
    "    elif level == 1:\n",
    "        _order = _id.split(':')[1]\n",
    "        if _order == '?':\n",
    "            return 'noCat'\n",
    "        else:\n",
    "            return _order\n",
    "    elif level == 2:\n",
    "        _superfamily = _id.split(':')[2]\n",
    "        if _superfamily == '?':\n",
    "            return 'noCat'\n",
    "        else:\n",
    "            return _superfamily\n",
    "    else:\n",
    "        print('Something wrong! Check if level is 0, 1 or 2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This needs to be updated here according to genome (either p_ctg or h_ctg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "source_dir = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly'\n",
    "genome = 'Pst_104E_v12_p_ctg'\n",
    "out_dir = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/Pst_104E_v12/revision/TE_dating'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#remove all commenting lines from the initial repet file\n",
    "!grep -v \"^#\" {source_dir}/{genome}.REPET.gff > {out_dir}/{genome}.REPET.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff = pd.read_csv(out_dir+'/'+genome+'.REPET.gff', sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This needs to be updated here according to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#This needs to be updated here according to genome\n",
    "TE_post_analysis_p = '/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/REPET/Pst79_p/Pst79_p_full_annotate/postanalysis/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TE_post_analysis_p_header = 'TE      length  covg    frags   fullLgthFrags   copies  fullLgthCopies  meanId  sdId    minId   q25Id   medId   q75Id   maxId   meanLgth        sdLgth  minLgth q25Lgth medLgth q75Lgth maxLgth meanLgthPerc    sdLgthPerc      minLgthPerc  q25LgthPerc     medLgthPerc     q75LgthPerc     maxLgthPerc'.split(' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TE_post_analysis_p_header = [x for x in TE_post_analysis_p_header if x != '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastclust.log\r\n",
      "error.log\r\n",
      "Pst79p_anno_chr_allTEs_nr_noSSR_join_path.annotStatsPerTE.tab\r\n",
      "Pst79p_anno_chr_allTEs_nr_noSSR_join_path.globalAnnotStatsPerTE.txt\r\n",
      "Pst79p_anno_chr_allTEs_path.annotStatsPerCluster.tab\r\n",
      "Pst79p_anno_chr_allTEs_path.annotStatsPerTE.tab\r\n",
      "Pst79p_anno_chr_allTEs_path.globalAnnotStatsPerTE.txt\r\n",
      "Pst79p_anno_refTEs.fa\r\n",
      "Pst79p_anno_refTEs.fa_blastclust.classifStatsPerCluster.tab\r\n",
      "Pst79p_anno_refTEs.fa_blastclust.globalStatsPerCluster.txt\r\n",
      "Pst79p_anno_refTEs.fa_blastclust.statsPerCluster.tab\r\n",
      "Pst79p_anno_refTEs.fa_blastclust.tab\r\n"
     ]
    }
   ],
   "source": [
    "!ls {TE_post_analysis_p}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This needs to be updated here according to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Check the code dict anyway\n",
      "{'DTX-comp-chim': 'DNA_transposon TIR', 'DMX-incomp': 'DNA_transposon Maverick', 'RIX-comp': 'Retrotransposon LINE', 'RYX-comp-chim': 'Retrotransposon DIRS', 'DXX-MITE': 'DNA_transposon MITE', 'RXX-TRIM': 'Retrotransposon TRIM', 'DTX-incomp': 'DNA_transposon TIR', 'RYX-comp': 'Retrotransposon DIRS', 'noCat': 'noCat', 'DTX-incomp-chim': 'DNA_transposon TIR', 'RLX-comp': 'Retrotransposon LTR', 'RSX-incomp-chim': 'Retrotransposon SINE', 'RXX-TRIM-chim': 'Retrotransposon TRIM', 'RIX-incomp-chim': 'Retrotransposon LINE', 'RLX-incomp-chim': 'Retrotransposon LTR', 'DHX-incomp': 'DNA_transposon Helitron', 'RXX': 'Retrotransposon noCat', 'RYX-incomp': 'Retrotransposon DIRS', 'DHX-incomp-chim': 'DNA_transposon Helitron', 'DXX-MITE-chim': 'DNA_transposon MITE', 'RXX-chim': 'Retrotransposon noCat', 'RIX-incomp': 'Retrotransposon LINE', 'DXX': 'DNA_transposon noCat', 'RSX-incomp': 'Retrotransposon SINE', 'RXX-LARD-chim': 'Retrotransposon LARD', 'RLX-incomp': 'Retrotransposon LTR', 'RLX-comp-chim': 'Retrotransposon LTR', 'XXX-chim': 'noCat', 'RPX-incomp-chim': 'Retrotransposon PLE', 'DTX-comp': 'DNA_transposon TIR', 'RXX-LARD': 'Retrotransposon LARD', 'RYX-incomp-chim': 'Retrotransposon DIRS', 'DHX-comp': 'DNA_transposon Helitron', 'PotentialHostGene': 'Potential Host Gene'}\n"
     ]
    }
   ],
   "source": [
    "#this needs to be fixed up to pick the proper summary table\n",
    "p_repet_summary_df = pd.read_csv(TE_post_analysis_p+'/'+'Pst79p_anno_chr_allTEs_nr_noSSR_join_path.annotStatsPerTE.tab' ,\\\n",
    "                                names = TE_post_analysis_p_header, header=None, sep='\\t', skiprows=1 )\n",
    "\n",
    "#check if I can filter the tab files for removing all TEs that are on the 2000 plus contigs\n",
    "#remove tRNAs TEs with infernal\n",
    "\n",
    "p_repet_summary_df['Code'] = p_repet_summary_df['TE'].apply(lambda x: x.split('_')[0])\n",
    "\n",
    "code_keys = p_repet_summary_df['Code'].unique()\n",
    "\n",
    "code_keys.sort()\n",
    "\n",
    "code_long = ['DNA_transposon Helitron', 'DNA_transposon Helitron', 'DNA_transposon Helitron', 'DNA_transposon Maverick',\\\n",
    "            'DNA_transposon TIR', 'DNA_transposon TIR', 'DNA_transposon TIR', 'DNA_transposon TIR', 'DNA_transposon noCat',\\\n",
    "             'DNA_transposon MITE','DNA_transposon MITE', 'Potential Host Gene', 'Retrotransposon LINE', 'Retrotransposon LINE',\\\n",
    "             'Retrotransposon LINE','Retrotransposon LTR','Retrotransposon LTR', 'Retrotransposon LTR', 'Retrotransposon LTR', 'Retrotransposon PLE', \\\n",
    "             'Retrotransposon SINE',  'Retrotransposon SINE', 'Retrotransposon noCat', 'Retrotransposon LARD',\\\n",
    "             'Retrotransposon LARD', 'Retrotransposon TRIM', 'Retrotransposon TRIM', 'Retrotransposon noCat',  \\\n",
    "             'Retrotransposon DIRS','Retrotransposon DIRS','Retrotransposon DIRS','Retrotransposon DIRS',\\\n",
    "             'noCat', 'noCat']\n",
    "if len(code_keys) != len(code_long):\n",
    "    print('Check the code_long list, because different length of keys and values!\\n\\n')\n",
    "else:\n",
    "    print('Check the code dict anyway')\n",
    "code_dict = dict(zip(code_keys, code_long))\n",
    "print(code_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>copies</th>\n",
       "      <th>covg</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Code long</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA_transposon Helitron</th>\n",
       "      <td>1075</td>\n",
       "      <td>817566</td>\n",
       "      <td>2989.764706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon MITE</th>\n",
       "      <td>3789</td>\n",
       "      <td>886304</td>\n",
       "      <td>490.240741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon Maverick</th>\n",
       "      <td>268</td>\n",
       "      <td>345406</td>\n",
       "      <td>8562.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon TIR</th>\n",
       "      <td>19166</td>\n",
       "      <td>12711595</td>\n",
       "      <td>4020.474359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_transposon noCat</th>\n",
       "      <td>5286</td>\n",
       "      <td>2376456</td>\n",
       "      <td>3034.734513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Potential Host Gene</th>\n",
       "      <td>1372</td>\n",
       "      <td>1375304</td>\n",
       "      <td>6120.490566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon DIRS</th>\n",
       "      <td>1337</td>\n",
       "      <td>1049299</td>\n",
       "      <td>6874.526316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon LARD</th>\n",
       "      <td>10752</td>\n",
       "      <td>4947564</td>\n",
       "      <td>5407.736111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon LINE</th>\n",
       "      <td>323</td>\n",
       "      <td>237992</td>\n",
       "      <td>4446.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon LTR</th>\n",
       "      <td>18893</td>\n",
       "      <td>16276421</td>\n",
       "      <td>6384.627841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon PLE</th>\n",
       "      <td>75</td>\n",
       "      <td>41843</td>\n",
       "      <td>8954.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon SINE</th>\n",
       "      <td>163</td>\n",
       "      <td>31138</td>\n",
       "      <td>317.923077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon TRIM</th>\n",
       "      <td>507</td>\n",
       "      <td>238695</td>\n",
       "      <td>1583.434783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Retrotransposon noCat</th>\n",
       "      <td>112</td>\n",
       "      <td>101098</td>\n",
       "      <td>4365.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>noCat</th>\n",
       "      <td>1218</td>\n",
       "      <td>850828</td>\n",
       "      <td>1679.829268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         copies      covg       length\n",
       "Code long                                             \n",
       "DNA_transposon Helitron    1075    817566  2989.764706\n",
       "DNA_transposon MITE        3789    886304   490.240741\n",
       "DNA_transposon Maverick     268    345406  8562.666667\n",
       "DNA_transposon TIR        19166  12711595  4020.474359\n",
       "DNA_transposon noCat       5286   2376456  3034.734513\n",
       "Potential Host Gene        1372   1375304  6120.490566\n",
       "Retrotransposon DIRS       1337   1049299  6874.526316\n",
       "Retrotransposon LARD      10752   4947564  5407.736111\n",
       "Retrotransposon LINE        323    237992  4446.625000\n",
       "Retrotransposon LTR       18893  16276421  6384.627841\n",
       "Retrotransposon PLE          75     41843  8954.000000\n",
       "Retrotransposon SINE        163     31138   317.923077\n",
       "Retrotransposon TRIM        507    238695  1583.434783\n",
       "Retrotransposon noCat       112    101098  4365.857143\n",
       "noCat                      1218    850828  1679.829268"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_summary_df['Code long'] = p_repet_summary_df['Code'].apply(lambda x: code_dict[x])\n",
    "p_repet_summary_sum_df = pd.pivot_table(p_repet_summary_df, values=['covg', 'copies'], index='Code long', aggfunc=np.sum)\n",
    "p_repet_summary_mean_df = pd.pivot_table(p_repet_summary_df, values='length', index='Code long', aggfunc=np.mean)\n",
    "pd.concat([p_repet_summary_sum_df,p_repet_summary_mean_df], axis=1 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This needs to be updated here according to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#now filter the gff dataframe to delete all the high coverage contigs\n",
    "#This might would have to be fixed as well. If we don't delete it as files should be already filtered\n",
    "contigs_smaller_2000 = pd.read_csv('/home/benjamin/genome_assembly/PST79/FALCON/p_assemblies/v9_1/032017_assembly/pcontig_smaller_2000.txt',\\\n",
    "                                  header=None)[0].tolist()\n",
    "\n",
    "p_repet_gff = pd.read_csv(out_dir+'/'+genome+'.REPET.gff', sep='\\t', header = None)\n",
    "\n",
    "p_repet_gff_filtered = p_repet_gff[p_repet_gff[0].isin(contigs_smaller_2000)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#filter out potential host genes\n",
    "p_repet_gff_filtered = p_repet_gff_filtered[~p_repet_gff_filtered[8].str.contains(\"Potential\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered['ID'] = p_repet_gff_filtered.apply(lambda row: ID_filter_gff(row[1], row[8]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DXX-MITE_MCL772_Pst79_REPET-B-G2372-Map5',\n",
       "       'RLX-incomp_MCL1006_Pst79_REPET-B-R3086-Map4',\n",
       "       'RXX-LARD_MCL8_Pst79_REPET-L-B574-Map1_reversed', ...,\n",
       "       'LOA_Ele7:ClassI:LINE:I', 'I-3B_CQ:ClassI:LINE:I', 'AACATCAAT3'], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_gff_filtered['ID'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#re-generate the code dict using the gff ID as Code keys\n",
    "\n",
    "\n",
    "code_keys_gff = p_repet_gff_filtered[p_repet_gff_filtered[1].str.contains('REPET_TE')]['ID'].unique()\n",
    "\n",
    "code_keys_gff = list({x.split('_')[0] for x in code_keys_gff})\n",
    "\n",
    "code_keys_gff.sort()\n",
    "\n",
    "#remove Potential host genes from long code list as those were filtered out previously.\n",
    "code_long.remove('Potential Host Gene')\n",
    "\n",
    "if len(code_keys_gff) != len(code_long):\n",
    "    print(\"Go and check something is wrong at the code key stage!\")\n",
    "\n",
    "code_dict = dict(zip(code_keys_gff, code_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aas\n"
     ]
    }
   ],
   "source": [
    "three_letter_code = list({x for x in code_keys_gff})\n",
    "\n",
    "three_letter_code.sort()\n",
    "\n",
    "three_letter_values = []\n",
    "for x in three_letter_code:\n",
    "    if 'MITE' in x:\n",
    "        _value = \"ClassII:MITE:?\"\n",
    "        three_letter_values.append(_value)\n",
    "        continue\n",
    "    if 'LARD' in x:\n",
    "        _value = 'ClassI:LARD:?'\n",
    "        three_letter_values.append(_value)\n",
    "        continue\n",
    "    if 'TRIM' in x:\n",
    "        _value = 'ClassI:TRIM:?'\n",
    "        three_letter_values.append(_value)\n",
    "        continue\n",
    "    _value =''\n",
    "    if x[0] == 'D':\n",
    "        _value = _value + 'ClassII:'\n",
    "    if x[0] == 'R':\n",
    "        _value = _value + 'ClassI:'\n",
    "    if x[0] != 'D' and x[0] != 'R':\n",
    "        _value = 'noCat'\n",
    "        three_letter_values.append(_value)\n",
    "        continue\n",
    "    if x[1] == 'T':\n",
    "        _value = _value + 'TIR:?'\n",
    "    if x[1] == 'H':\n",
    "        _value = _value + 'Helitron:?'\n",
    "    if x[1] == 'M':\n",
    "        _value = _value + 'Maverick:?'\n",
    "    if x[0:2] == 'DY':\n",
    "        _value = _value + ':Crypton:?'\n",
    "    if x[1] == 'X':\n",
    "        _value = _value + '?:?'\n",
    "    if x[1] == 'I':\n",
    "        _value = _value + 'LINE:?'\n",
    "    if x[1] == 'L':\n",
    "        _value = _value + 'LTR:?'\n",
    "    if x[1] == 'P':\n",
    "        _value = _value + 'Penelope:?'\n",
    "    if x[1] == 'S':\n",
    "        _value = _value + 'SINE:?'\n",
    "    if x[0:2] == 'RY':\n",
    "        _value = _value + 'DIRS:?'    \n",
    "    three_letter_values.append(_value)\n",
    "\n",
    "if len(three_letter_code) == len(three_letter_values):\n",
    "    print(\"Aas\")\n",
    "    three_letter_dict = dict(zip(three_letter_code, three_letter_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(three_letter_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered['Class:Order:Superfamily'] = p_repet_gff_filtered.apply(lambda row: blast_hit_gff(row[1], row[8], row['ID']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ClassII:?:Ginger2/TDD\n",
      "ClassII:?:Ginger2_TDD\n"
     ]
    }
   ],
   "source": [
    "#generate a dict that can be used to rename the Class:Order:Superfamily column considering that partial matches ([2] == match_part) might contain different\n",
    "#IDs even though they are the same TE only partial.\n",
    "_tmp_subset = p_repet_gff_filtered[~p_repet_gff_filtered[1].str.contains('SSR')].loc[:, 'ID':].sort_values(by=['ID','Class:Order:Superfamily'])\\\n",
    ".drop_duplicates(subset='ID', keep ='last')\n",
    "\n",
    "TE_COS_dict = dict(zip(_tmp_subset.loc[:, 'ID'], _tmp_subset.loc[:, 'Class:Order:Superfamily' ]))\n",
    "\n",
    "_tmp_subset = p_repet_gff_filtered[p_repet_gff_filtered[1].str.contains('SSR')].loc[:, 'ID':].sort_values(by=['ID','Class:Order:Superfamily'])\\\n",
    ".drop_duplicates(subset='ID', keep ='last')\n",
    "\n",
    "_tmp_dict = dict(zip(_tmp_subset.loc[:, 'ID'], _tmp_subset.loc[:, 'Class:Order:Superfamily' ]))\n",
    "\n",
    "TE_COS_dict.update(_tmp_dict)\n",
    "#remove all backslashes from the values as this will conflict with the output later on\n",
    "for x in TE_COS_dict.keys():\n",
    "    if '/' in TE_COS_dict[x]:\n",
    "        value = TE_COS_dict[x]\n",
    "        print(value)\n",
    "        TE_COS_dict[x] = value.replace('/','_')\n",
    "        print(TE_COS_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#drop zeros from p_repet_summary_df\n",
    "p_repet_summary_df = p_repet_summary_df[p_repet_summary_df.copies > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#remove Potential host genes\n",
    "p_repet_summary_df = p_repet_summary_df[~p_repet_summary_df.TE.str.startswith(\"Potential\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_summary_df[\"COS\"] = p_repet_summary_df[\"TE\"].apply(lambda x: TE_COS_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "COS\n",
       "ClassI:?:?                      77923\n",
       "ClassI:DIRS:?                    5230\n",
       "ClassI:DIRS:DIRS               411921\n",
       "ClassI:LARD:?                 2580686\n",
       "ClassI:LINE:?                    4187\n",
       "ClassI:LINE:I                  207662\n",
       "ClassI:LINE:Jockey               9239\n",
       "ClassI:LTR:?                  1660264\n",
       "ClassI:LTR:Copia              4399303\n",
       "ClassI:LTR:Gypsy             10817418\n",
       "ClassI:LTR:Retrovirus           18586\n",
       "ClassI:Penelope:?               41843\n",
       "ClassI:SINE:?                   31138\n",
       "ClassI:TRIM:?                  183085\n",
       "ClassII:?:?                   4053567\n",
       "ClassII:?:Academ               267214\n",
       "ClassII:Helitron:?             563604\n",
       "ClassII:Helitron:Helitron      253962\n",
       "ClassII:MITE:?                 886304\n",
       "ClassII:Maverick:?             324099\n",
       "ClassII:Maverick:Maverick       21307\n",
       "ClassII:TIR:?                 3786988\n",
       "ClassII:TIR:CACTA             1092698\n",
       "ClassII:TIR:MuDR              1506387\n",
       "ClassII:TIR:P                   53627\n",
       "ClassII:TIR:PIF-Harbinger     1816676\n",
       "ClassII:TIR:Tc1-Mariner       1281300\n",
       "ClassII:TIR:hAT               3705159\n",
       "noCat                          850828\n",
       "Name: covg, dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_summary_df.groupby(\"COS\")[\"covg\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ClassI:LINE:L1',\n",
       " 'ClassI:LINE:R2',\n",
       " 'ClassI:LINE:RTE',\n",
       " 'ClassI:LTR:Bel-Pao',\n",
       " 'ClassI:LTR:ERV',\n",
       " 'ClassI:PLE:Penelope',\n",
       " 'ClassII:?:Ginger1',\n",
       " 'ClassII:?:Ginger2_TDD',\n",
       " 'ClassII:?:Kolobok',\n",
       " 'ClassII:?:Sola',\n",
       " 'ClassII:Crypton:Crypton',\n",
       " 'ClassII:TIR:PiggyBac',\n",
       " 'ClassII:TIR:Transib',\n",
       " 'SSR'}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(p_repet_gff_filtered['Class:Order:Superfamily'].unique()) \\\n",
    "    - set(p_repet_summary_df[\"COS\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>ID</th>\n",
       "      <th>Class:Order:Superfamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>14603.0</td>\n",
       "      <td>14784.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86142_pcontig_000_DNA3-5_PSt:ClassII:?:?;...</td>\n",
       "      <td>DNA3-5_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14603.0</td>\n",
       "      <td>14690.0</td>\n",
       "      <td>2.000000e-13</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86142-1_pcontig_000_DNA3-5_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-5_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>14707.0</td>\n",
       "      <td>14784.0</td>\n",
       "      <td>3.000000e-12</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86142-2_pcontig_000_DNA3-5_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-5_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>15195.0</td>\n",
       "      <td>15499.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86190_pcontig_000_Gypsy-2_PGr-I:ClassI:LT...</td>\n",
       "      <td>Gypsy-2_PGr-I:ClassI:LTR:Gypsy</td>\n",
       "      <td>ClassI:LTR:Gypsy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15195.0</td>\n",
       "      <td>15362.0</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86190-1_pcontig_000_Gypsy-2_PGr-I:ClassI:...</td>\n",
       "      <td>Gypsy-2_PGr-I:ClassI:LTR:Gypsy</td>\n",
       "      <td>ClassI:LTR:Gypsy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15416.0</td>\n",
       "      <td>15499.0</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86190-2_pcontig_000_Gypsy-2_PGr-I:ClassI:...</td>\n",
       "      <td>Gypsy-2_PGr-I:ClassI:LTR:Gypsy</td>\n",
       "      <td>ClassI:LTR:Gypsy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>15748.0</td>\n",
       "      <td>15834.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86191_pcontig_000_Gypsy-2_PGr-I:ClassI:LT...</td>\n",
       "      <td>Gypsy-2_PGr-I:ClassI:LTR:Gypsy</td>\n",
       "      <td>ClassI:LTR:Gypsy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>15748.0</td>\n",
       "      <td>15834.0</td>\n",
       "      <td>6.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86191-1_pcontig_000_Gypsy-2_PGr-I:ClassI:...</td>\n",
       "      <td>Gypsy-2_PGr-I:ClassI:LTR:Gypsy</td>\n",
       "      <td>ClassI:LTR:Gypsy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>16541.0</td>\n",
       "      <td>16788.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86140_pcontig_000_DNA3-4_PSt:ClassII:?:?;...</td>\n",
       "      <td>DNA3-4_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>16541.0</td>\n",
       "      <td>16788.0</td>\n",
       "      <td>3.000000e-44</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86140-1_pcontig_000_DNA3-4_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-4_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>27997.0</td>\n",
       "      <td>28147.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86255_pcontig_000_DNA3-16_PSt:ClassII:?:?...</td>\n",
       "      <td>DNA3-16_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>27997.0</td>\n",
       "      <td>28147.0</td>\n",
       "      <td>6.000000e-18</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86255-1_pcontig_000_DNA3-16_PSt:ClassII:?...</td>\n",
       "      <td>DNA3-16_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>33911.0</td>\n",
       "      <td>34183.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86139_pcontig_000_DNA3-11_PSt:ClassII:?:?...</td>\n",
       "      <td>DNA3-11_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>33911.0</td>\n",
       "      <td>33983.0</td>\n",
       "      <td>6.000000e-20</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86139-1_pcontig_000_DNA3-11_PSt:ClassII:?...</td>\n",
       "      <td>DNA3-11_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>34059.0</td>\n",
       "      <td>34183.0</td>\n",
       "      <td>6.000000e-20</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86139-2_pcontig_000_DNA3-11_PSt:ClassII:?...</td>\n",
       "      <td>DNA3-11_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>36386.0</td>\n",
       "      <td>36472.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86280_pcontig_000_DNA3-7_PSt:ClassII:?:?;...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>36386.0</td>\n",
       "      <td>36445.0</td>\n",
       "      <td>9.000000e-21</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86280-1_pcontig_000_DNA3-7_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>36443.0</td>\n",
       "      <td>36472.0</td>\n",
       "      <td>9.000000e-21</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86280-2_pcontig_000_DNA3-7_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>36473.0</td>\n",
       "      <td>36821.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86281_pcontig_000_DNA3-7_PSt:ClassII:?:?;...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>36473.0</td>\n",
       "      <td>36541.0</td>\n",
       "      <td>3.000000e-27</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86281-1_pcontig_000_DNA3-7_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>36589.0</td>\n",
       "      <td>36705.0</td>\n",
       "      <td>1.000000e-27</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86281-2_pcontig_000_DNA3-7_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>36694.0</td>\n",
       "      <td>36821.0</td>\n",
       "      <td>1.000000e-27</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86281-3_pcontig_000_DNA3-7_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-7_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>45525.0</td>\n",
       "      <td>46032.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86146_pcontig_000_DNA9-1_PSt:ClassII:?:?;...</td>\n",
       "      <td>DNA9-1_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>45525.0</td>\n",
       "      <td>45623.0</td>\n",
       "      <td>1.000000e-20</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86146-1_pcontig_000_DNA9-1_PSt:ClassII:?:...</td>\n",
       "      <td>DNA9-1_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>45631.0</td>\n",
       "      <td>45723.0</td>\n",
       "      <td>2.000000e-38</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86146-2_pcontig_000_DNA9-1_PSt:ClassII:?:...</td>\n",
       "      <td>DNA9-1_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>45700.0</td>\n",
       "      <td>45786.0</td>\n",
       "      <td>2.000000e-38</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86146-3_pcontig_000_DNA9-1_PSt:ClassII:?:...</td>\n",
       "      <td>DNA9-1_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>45761.0</td>\n",
       "      <td>46032.0</td>\n",
       "      <td>2.000000e-38</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86146-4_pcontig_000_DNA9-1_PSt:ClassII:?:...</td>\n",
       "      <td>DNA9-1_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>52273.0</td>\n",
       "      <td>52475.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms86145_pcontig_000_DNA3-9_PSt:ClassII:?:?;...</td>\n",
       "      <td>DNA3-9_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>52273.0</td>\n",
       "      <td>52386.0</td>\n",
       "      <td>6.000000e-23</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86145-1_pcontig_000_DNA3-9_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-9_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>pcontig_000</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>52389.0</td>\n",
       "      <td>52475.0</td>\n",
       "      <td>9.000000e-15</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp86145-2_pcontig_000_DNA3-9_PSt:ClassII:?:...</td>\n",
       "      <td>DNA3-9_PSt:ClassII:?:?</td>\n",
       "      <td>ClassII:?:?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539639</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>17841.0</td>\n",
       "      <td>17912.0</td>\n",
       "      <td>5.000000e-81</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14172-1_pcontig_235_ATCOPIA43I:ClassI:LTR...</td>\n",
       "      <td>ATCOPIA43I:ClassI:LTR:Copia</td>\n",
       "      <td>ClassI:LTR:Copia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539640</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>19654.0</td>\n",
       "      <td>19734.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14239_pcontig_235_RTAg4:ClassI:LINE:I;Tar...</td>\n",
       "      <td>RTAg4:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539641</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>19654.0</td>\n",
       "      <td>19734.0</td>\n",
       "      <td>1.000000e-13</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14239-1_pcontig_235_RTAg4:ClassI:LINE:I;P...</td>\n",
       "      <td>RTAg4:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539642</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>20265.0</td>\n",
       "      <td>20408.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14236_pcontig_235_R1-2_DEu:ClassI:LINE:I;...</td>\n",
       "      <td>R1-2_DEu:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539643</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>20265.0</td>\n",
       "      <td>20408.0</td>\n",
       "      <td>1.000000e-17</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14236-1_pcontig_235_R1-2_DEu:ClassI:LINE:...</td>\n",
       "      <td>R1-2_DEu:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539644</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>20636.0</td>\n",
       "      <td>20663.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14230_pcontig_235_Loa-1_DEl:ClassI:LINE:I...</td>\n",
       "      <td>Loa-1_DEl:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539645</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>20636.0</td>\n",
       "      <td>20663.0</td>\n",
       "      <td>1.000000e-11</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14230-1_pcontig_235_Loa-1_DEl:ClassI:LINE...</td>\n",
       "      <td>Loa-1_DEl:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539646</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>20664.0</td>\n",
       "      <td>20810.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14235_pcontig_235_R1-1_SIn:ClassI:LINE:I;...</td>\n",
       "      <td>R1-1_SIn:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539647</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>20664.0</td>\n",
       "      <td>20810.0</td>\n",
       "      <td>5.000000e-12</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14235-1_pcontig_235_R1-1_SIn:ClassI:LINE:...</td>\n",
       "      <td>R1-1_SIn:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539648</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>21173.0</td>\n",
       "      <td>21259.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14228_pcontig_235_Jockey-8_LMi:ClassI:LIN...</td>\n",
       "      <td>Jockey-8_LMi:ClassI:LINE:Jockey</td>\n",
       "      <td>ClassI:LINE:Jockey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539649</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>21173.0</td>\n",
       "      <td>21259.0</td>\n",
       "      <td>2.000000e-17</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14228-1_pcontig_235_Jockey-8_LMi:ClassI:L...</td>\n",
       "      <td>Jockey-8_LMi:ClassI:LINE:Jockey</td>\n",
       "      <td>ClassI:LINE:Jockey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539650</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>21565.0</td>\n",
       "      <td>21645.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14227_pcontig_235_Jockey-5_DK:ClassI:LINE...</td>\n",
       "      <td>Jockey-5_DK:ClassI:LINE:Jockey</td>\n",
       "      <td>ClassI:LINE:Jockey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539651</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>21565.0</td>\n",
       "      <td>21645.0</td>\n",
       "      <td>6.000000e-25</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14227-1_pcontig_235_Jockey-5_DK:ClassI:LI...</td>\n",
       "      <td>Jockey-5_DK:ClassI:LINE:Jockey</td>\n",
       "      <td>ClassI:LINE:Jockey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539652</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>21787.0</td>\n",
       "      <td>21861.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14174_pcontig_235_Copia-10_SB-I:ClassI:LT...</td>\n",
       "      <td>Copia-10_SB-I:ClassI:LTR:Copia</td>\n",
       "      <td>ClassI:LTR:Copia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539653</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>21787.0</td>\n",
       "      <td>21861.0</td>\n",
       "      <td>2.000000e-57</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14174-1_pcontig_235_Copia-10_SB-I:ClassI:...</td>\n",
       "      <td>Copia-10_SB-I:ClassI:LTR:Copia</td>\n",
       "      <td>ClassI:LTR:Copia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539654</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>22068.0</td>\n",
       "      <td>22154.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14231_pcontig_235_LOA_Ele7:ClassI:LINE:I;...</td>\n",
       "      <td>LOA_Ele7:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539655</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>22068.0</td>\n",
       "      <td>22154.0</td>\n",
       "      <td>2.000000e-18</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14231-1_pcontig_235_LOA_Ele7:ClassI:LINE:...</td>\n",
       "      <td>LOA_Ele7:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539656</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>22657.0</td>\n",
       "      <td>22782.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14240_pcontig_235_SART-5_MOc:ClassI:LINE:...</td>\n",
       "      <td>SART-5_MOc:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539657</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>22657.0</td>\n",
       "      <td>22782.0</td>\n",
       "      <td>5.000000e-19</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14240-1_pcontig_235_SART-5_MOc:ClassI:LIN...</td>\n",
       "      <td>SART-5_MOc:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539658</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>23879.0</td>\n",
       "      <td>24010.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14225_pcontig_235_I-3B_CQ:ClassI:LINE:I;T...</td>\n",
       "      <td>I-3B_CQ:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539659</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>23879.0</td>\n",
       "      <td>24010.0</td>\n",
       "      <td>4.000000e-15</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14225-1_pcontig_235_I-3B_CQ:ClassI:LINE:I...</td>\n",
       "      <td>I-3B_CQ:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539662</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>25684.0</td>\n",
       "      <td>25884.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms14238_pcontig_235_R1A_NVi:ClassI:LINE:I;T...</td>\n",
       "      <td>R1A_NVi:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539663</th>\n",
       "      <td>pcontig_235</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>25684.0</td>\n",
       "      <td>25884.0</td>\n",
       "      <td>2.000000e-17</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp14238-1_pcontig_235_R1A_NVi:ClassI:LINE:I...</td>\n",
       "      <td>R1A_NVi:ClassI:LINE:I</td>\n",
       "      <td>ClassI:LINE:I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539682</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>11706.0</td>\n",
       "      <td>11942.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms1389_pcontig_248_Helitron-3_ZM:ClassII:He...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539683</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>11706.0</td>\n",
       "      <td>11822.0</td>\n",
       "      <td>6.000000e-25</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp1389-1_pcontig_248_Helitron-3_ZM:ClassII:...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539684</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>11811.0</td>\n",
       "      <td>11942.0</td>\n",
       "      <td>6.000000e-25</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp1389-2_pcontig_248_Helitron-3_ZM:ClassII:...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539685</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match</td>\n",
       "      <td>12495.0</td>\n",
       "      <td>12776.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ms1390_pcontig_248_Helitron-3_ZM:ClassII:He...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539686</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>12495.0</td>\n",
       "      <td>12608.0</td>\n",
       "      <td>5.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp1390-1_pcontig_248_Helitron-3_ZM:ClassII:...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539687</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>12603.0</td>\n",
       "      <td>12665.0</td>\n",
       "      <td>5.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp1390-2_pcontig_248_Helitron-3_ZM:ClassII:...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539688</th>\n",
       "      <td>pcontig_248</td>\n",
       "      <td>Pst79p_anno_REPET_tblastx</td>\n",
       "      <td>match_part</td>\n",
       "      <td>12717.0</td>\n",
       "      <td>12776.0</td>\n",
       "      <td>5.000000e-12</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=mp1390-3_pcontig_248_Helitron-3_ZM:ClassII:...</td>\n",
       "      <td>Helitron-3_ZM:ClassII:Helitron:Helitron</td>\n",
       "      <td>ClassII:Helitron:Helitron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>257944 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                          1           2        3        4  \\\n",
       "8       pcontig_000  Pst79p_anno_REPET_tblastx       match  14603.0  14784.0   \n",
       "9       pcontig_000  Pst79p_anno_REPET_tblastx  match_part  14603.0  14690.0   \n",
       "10      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  14707.0  14784.0   \n",
       "21      pcontig_000  Pst79p_anno_REPET_tblastx       match  15195.0  15499.0   \n",
       "22      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  15195.0  15362.0   \n",
       "23      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  15416.0  15499.0   \n",
       "24      pcontig_000  Pst79p_anno_REPET_tblastx       match  15748.0  15834.0   \n",
       "25      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  15748.0  15834.0   \n",
       "31      pcontig_000  Pst79p_anno_REPET_tblastx       match  16541.0  16788.0   \n",
       "32      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  16541.0  16788.0   \n",
       "55      pcontig_000  Pst79p_anno_REPET_tblastx       match  27997.0  28147.0   \n",
       "56      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  27997.0  28147.0   \n",
       "61      pcontig_000  Pst79p_anno_REPET_tblastx       match  33911.0  34183.0   \n",
       "62      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  33911.0  33983.0   \n",
       "63      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  34059.0  34183.0   \n",
       "72      pcontig_000  Pst79p_anno_REPET_tblastx       match  36386.0  36472.0   \n",
       "73      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  36386.0  36445.0   \n",
       "74      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  36443.0  36472.0   \n",
       "75      pcontig_000  Pst79p_anno_REPET_tblastx       match  36473.0  36821.0   \n",
       "76      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  36473.0  36541.0   \n",
       "77      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  36589.0  36705.0   \n",
       "78      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  36694.0  36821.0   \n",
       "85      pcontig_000  Pst79p_anno_REPET_tblastx       match  45525.0  46032.0   \n",
       "86      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  45525.0  45623.0   \n",
       "87      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  45631.0  45723.0   \n",
       "88      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  45700.0  45786.0   \n",
       "89      pcontig_000  Pst79p_anno_REPET_tblastx  match_part  45761.0  46032.0   \n",
       "107     pcontig_000  Pst79p_anno_REPET_tblastx       match  52273.0  52475.0   \n",
       "108     pcontig_000  Pst79p_anno_REPET_tblastx  match_part  52273.0  52386.0   \n",
       "109     pcontig_000  Pst79p_anno_REPET_tblastx  match_part  52389.0  52475.0   \n",
       "...             ...                        ...         ...      ...      ...   \n",
       "539639  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  17841.0  17912.0   \n",
       "539640  pcontig_235  Pst79p_anno_REPET_tblastx       match  19654.0  19734.0   \n",
       "539641  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  19654.0  19734.0   \n",
       "539642  pcontig_235  Pst79p_anno_REPET_tblastx       match  20265.0  20408.0   \n",
       "539643  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  20265.0  20408.0   \n",
       "539644  pcontig_235  Pst79p_anno_REPET_tblastx       match  20636.0  20663.0   \n",
       "539645  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  20636.0  20663.0   \n",
       "539646  pcontig_235  Pst79p_anno_REPET_tblastx       match  20664.0  20810.0   \n",
       "539647  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  20664.0  20810.0   \n",
       "539648  pcontig_235  Pst79p_anno_REPET_tblastx       match  21173.0  21259.0   \n",
       "539649  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  21173.0  21259.0   \n",
       "539650  pcontig_235  Pst79p_anno_REPET_tblastx       match  21565.0  21645.0   \n",
       "539651  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  21565.0  21645.0   \n",
       "539652  pcontig_235  Pst79p_anno_REPET_tblastx       match  21787.0  21861.0   \n",
       "539653  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  21787.0  21861.0   \n",
       "539654  pcontig_235  Pst79p_anno_REPET_tblastx       match  22068.0  22154.0   \n",
       "539655  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  22068.0  22154.0   \n",
       "539656  pcontig_235  Pst79p_anno_REPET_tblastx       match  22657.0  22782.0   \n",
       "539657  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  22657.0  22782.0   \n",
       "539658  pcontig_235  Pst79p_anno_REPET_tblastx       match  23879.0  24010.0   \n",
       "539659  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  23879.0  24010.0   \n",
       "539662  pcontig_235  Pst79p_anno_REPET_tblastx       match  25684.0  25884.0   \n",
       "539663  pcontig_235  Pst79p_anno_REPET_tblastx  match_part  25684.0  25884.0   \n",
       "539682  pcontig_248  Pst79p_anno_REPET_tblastx       match  11706.0  11942.0   \n",
       "539683  pcontig_248  Pst79p_anno_REPET_tblastx  match_part  11706.0  11822.0   \n",
       "539684  pcontig_248  Pst79p_anno_REPET_tblastx  match_part  11811.0  11942.0   \n",
       "539685  pcontig_248  Pst79p_anno_REPET_tblastx       match  12495.0  12776.0   \n",
       "539686  pcontig_248  Pst79p_anno_REPET_tblastx  match_part  12495.0  12608.0   \n",
       "539687  pcontig_248  Pst79p_anno_REPET_tblastx  match_part  12603.0  12665.0   \n",
       "539688  pcontig_248  Pst79p_anno_REPET_tblastx  match_part  12717.0  12776.0   \n",
       "\n",
       "                   5  6  7                                                  8  \\\n",
       "8       0.000000e+00  -  .  ID=ms86142_pcontig_000_DNA3-5_PSt:ClassII:?:?;...   \n",
       "9       2.000000e-13  -  .  ID=mp86142-1_pcontig_000_DNA3-5_PSt:ClassII:?:...   \n",
       "10      3.000000e-12  -  .  ID=mp86142-2_pcontig_000_DNA3-5_PSt:ClassII:?:...   \n",
       "21      0.000000e+00  +  .  ID=ms86190_pcontig_000_Gypsy-2_PGr-I:ClassI:LT...   \n",
       "22      1.000000e-12  +  .  ID=mp86190-1_pcontig_000_Gypsy-2_PGr-I:ClassI:...   \n",
       "23      1.000000e-12  +  .  ID=mp86190-2_pcontig_000_Gypsy-2_PGr-I:ClassI:...   \n",
       "24      0.000000e+00  +  .  ID=ms86191_pcontig_000_Gypsy-2_PGr-I:ClassI:LT...   \n",
       "25      6.000000e-12  +  .  ID=mp86191-1_pcontig_000_Gypsy-2_PGr-I:ClassI:...   \n",
       "31      0.000000e+00  -  .  ID=ms86140_pcontig_000_DNA3-4_PSt:ClassII:?:?;...   \n",
       "32      3.000000e-44  -  .  ID=mp86140-1_pcontig_000_DNA3-4_PSt:ClassII:?:...   \n",
       "55      0.000000e+00  -  .  ID=ms86255_pcontig_000_DNA3-16_PSt:ClassII:?:?...   \n",
       "56      6.000000e-18  -  .  ID=mp86255-1_pcontig_000_DNA3-16_PSt:ClassII:?...   \n",
       "61      0.000000e+00  -  .  ID=ms86139_pcontig_000_DNA3-11_PSt:ClassII:?:?...   \n",
       "62      6.000000e-20  -  .  ID=mp86139-1_pcontig_000_DNA3-11_PSt:ClassII:?...   \n",
       "63      6.000000e-20  -  .  ID=mp86139-2_pcontig_000_DNA3-11_PSt:ClassII:?...   \n",
       "72      0.000000e+00  +  .  ID=ms86280_pcontig_000_DNA3-7_PSt:ClassII:?:?;...   \n",
       "73      9.000000e-21  +  .  ID=mp86280-1_pcontig_000_DNA3-7_PSt:ClassII:?:...   \n",
       "74      9.000000e-21  +  .  ID=mp86280-2_pcontig_000_DNA3-7_PSt:ClassII:?:...   \n",
       "75      0.000000e+00  -  .  ID=ms86281_pcontig_000_DNA3-7_PSt:ClassII:?:?;...   \n",
       "76      3.000000e-27  -  .  ID=mp86281-1_pcontig_000_DNA3-7_PSt:ClassII:?:...   \n",
       "77      1.000000e-27  -  .  ID=mp86281-2_pcontig_000_DNA3-7_PSt:ClassII:?:...   \n",
       "78      1.000000e-27  -  .  ID=mp86281-3_pcontig_000_DNA3-7_PSt:ClassII:?:...   \n",
       "85      0.000000e+00  +  .  ID=ms86146_pcontig_000_DNA9-1_PSt:ClassII:?:?;...   \n",
       "86      1.000000e-20  +  .  ID=mp86146-1_pcontig_000_DNA9-1_PSt:ClassII:?:...   \n",
       "87      2.000000e-38  +  .  ID=mp86146-2_pcontig_000_DNA9-1_PSt:ClassII:?:...   \n",
       "88      2.000000e-38  +  .  ID=mp86146-3_pcontig_000_DNA9-1_PSt:ClassII:?:...   \n",
       "89      2.000000e-38  +  .  ID=mp86146-4_pcontig_000_DNA9-1_PSt:ClassII:?:...   \n",
       "107     0.000000e+00  -  .  ID=ms86145_pcontig_000_DNA3-9_PSt:ClassII:?:?;...   \n",
       "108     6.000000e-23  -  .  ID=mp86145-1_pcontig_000_DNA3-9_PSt:ClassII:?:...   \n",
       "109     9.000000e-15  -  .  ID=mp86145-2_pcontig_000_DNA3-9_PSt:ClassII:?:...   \n",
       "...              ... .. ..                                                ...   \n",
       "539639  5.000000e-81  +  .  ID=mp14172-1_pcontig_235_ATCOPIA43I:ClassI:LTR...   \n",
       "539640  0.000000e+00  -  .  ID=ms14239_pcontig_235_RTAg4:ClassI:LINE:I;Tar...   \n",
       "539641  1.000000e-13  -  .  ID=mp14239-1_pcontig_235_RTAg4:ClassI:LINE:I;P...   \n",
       "539642  0.000000e+00  -  .  ID=ms14236_pcontig_235_R1-2_DEu:ClassI:LINE:I;...   \n",
       "539643  1.000000e-17  -  .  ID=mp14236-1_pcontig_235_R1-2_DEu:ClassI:LINE:...   \n",
       "539644  0.000000e+00  -  .  ID=ms14230_pcontig_235_Loa-1_DEl:ClassI:LINE:I...   \n",
       "539645  1.000000e-11  -  .  ID=mp14230-1_pcontig_235_Loa-1_DEl:ClassI:LINE...   \n",
       "539646  0.000000e+00  -  .  ID=ms14235_pcontig_235_R1-1_SIn:ClassI:LINE:I;...   \n",
       "539647  5.000000e-12  -  .  ID=mp14235-1_pcontig_235_R1-1_SIn:ClassI:LINE:...   \n",
       "539648  0.000000e+00  -  .  ID=ms14228_pcontig_235_Jockey-8_LMi:ClassI:LIN...   \n",
       "539649  2.000000e-17  -  .  ID=mp14228-1_pcontig_235_Jockey-8_LMi:ClassI:L...   \n",
       "539650  0.000000e+00  -  .  ID=ms14227_pcontig_235_Jockey-5_DK:ClassI:LINE...   \n",
       "539651  6.000000e-25  -  .  ID=mp14227-1_pcontig_235_Jockey-5_DK:ClassI:LI...   \n",
       "539652  0.000000e+00  +  .  ID=ms14174_pcontig_235_Copia-10_SB-I:ClassI:LT...   \n",
       "539653  2.000000e-57  +  .  ID=mp14174-1_pcontig_235_Copia-10_SB-I:ClassI:...   \n",
       "539654  0.000000e+00  -  .  ID=ms14231_pcontig_235_LOA_Ele7:ClassI:LINE:I;...   \n",
       "539655  2.000000e-18  -  .  ID=mp14231-1_pcontig_235_LOA_Ele7:ClassI:LINE:...   \n",
       "539656  0.000000e+00  -  .  ID=ms14240_pcontig_235_SART-5_MOc:ClassI:LINE:...   \n",
       "539657  5.000000e-19  -  .  ID=mp14240-1_pcontig_235_SART-5_MOc:ClassI:LIN...   \n",
       "539658  0.000000e+00  -  .  ID=ms14225_pcontig_235_I-3B_CQ:ClassI:LINE:I;T...   \n",
       "539659  4.000000e-15  -  .  ID=mp14225-1_pcontig_235_I-3B_CQ:ClassI:LINE:I...   \n",
       "539662  0.000000e+00  -  .  ID=ms14238_pcontig_235_R1A_NVi:ClassI:LINE:I;T...   \n",
       "539663  2.000000e-17  -  .  ID=mp14238-1_pcontig_235_R1A_NVi:ClassI:LINE:I...   \n",
       "539682  0.000000e+00  -  .  ID=ms1389_pcontig_248_Helitron-3_ZM:ClassII:He...   \n",
       "539683  6.000000e-25  -  .  ID=mp1389-1_pcontig_248_Helitron-3_ZM:ClassII:...   \n",
       "539684  6.000000e-25  -  .  ID=mp1389-2_pcontig_248_Helitron-3_ZM:ClassII:...   \n",
       "539685  0.000000e+00  +  .  ID=ms1390_pcontig_248_Helitron-3_ZM:ClassII:He...   \n",
       "539686  5.000000e-12  +  .  ID=mp1390-1_pcontig_248_Helitron-3_ZM:ClassII:...   \n",
       "539687  5.000000e-12  +  .  ID=mp1390-2_pcontig_248_Helitron-3_ZM:ClassII:...   \n",
       "539688  5.000000e-12  +  .  ID=mp1390-3_pcontig_248_Helitron-3_ZM:ClassII:...   \n",
       "\n",
       "                                             ID    Class:Order:Superfamily  \n",
       "8                        DNA3-5_PSt:ClassII:?:?                ClassII:?:?  \n",
       "9                        DNA3-5_PSt:ClassII:?:?                ClassII:?:?  \n",
       "10                       DNA3-5_PSt:ClassII:?:?                ClassII:?:?  \n",
       "21               Gypsy-2_PGr-I:ClassI:LTR:Gypsy           ClassI:LTR:Gypsy  \n",
       "22               Gypsy-2_PGr-I:ClassI:LTR:Gypsy           ClassI:LTR:Gypsy  \n",
       "23               Gypsy-2_PGr-I:ClassI:LTR:Gypsy           ClassI:LTR:Gypsy  \n",
       "24               Gypsy-2_PGr-I:ClassI:LTR:Gypsy           ClassI:LTR:Gypsy  \n",
       "25               Gypsy-2_PGr-I:ClassI:LTR:Gypsy           ClassI:LTR:Gypsy  \n",
       "31                       DNA3-4_PSt:ClassII:?:?                ClassII:?:?  \n",
       "32                       DNA3-4_PSt:ClassII:?:?                ClassII:?:?  \n",
       "55                      DNA3-16_PSt:ClassII:?:?                ClassII:?:?  \n",
       "56                      DNA3-16_PSt:ClassII:?:?                ClassII:?:?  \n",
       "61                      DNA3-11_PSt:ClassII:?:?                ClassII:?:?  \n",
       "62                      DNA3-11_PSt:ClassII:?:?                ClassII:?:?  \n",
       "63                      DNA3-11_PSt:ClassII:?:?                ClassII:?:?  \n",
       "72                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "73                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "74                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "75                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "76                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "77                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "78                       DNA3-7_PSt:ClassII:?:?                ClassII:?:?  \n",
       "85                       DNA9-1_PSt:ClassII:?:?                ClassII:?:?  \n",
       "86                       DNA9-1_PSt:ClassII:?:?                ClassII:?:?  \n",
       "87                       DNA9-1_PSt:ClassII:?:?                ClassII:?:?  \n",
       "88                       DNA9-1_PSt:ClassII:?:?                ClassII:?:?  \n",
       "89                       DNA9-1_PSt:ClassII:?:?                ClassII:?:?  \n",
       "107                      DNA3-9_PSt:ClassII:?:?                ClassII:?:?  \n",
       "108                      DNA3-9_PSt:ClassII:?:?                ClassII:?:?  \n",
       "109                      DNA3-9_PSt:ClassII:?:?                ClassII:?:?  \n",
       "...                                         ...                        ...  \n",
       "539639              ATCOPIA43I:ClassI:LTR:Copia           ClassI:LTR:Copia  \n",
       "539640                      RTAg4:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539641                      RTAg4:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539642                   R1-2_DEu:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539643                   R1-2_DEu:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539644                  Loa-1_DEl:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539645                  Loa-1_DEl:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539646                   R1-1_SIn:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539647                   R1-1_SIn:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539648          Jockey-8_LMi:ClassI:LINE:Jockey         ClassI:LINE:Jockey  \n",
       "539649          Jockey-8_LMi:ClassI:LINE:Jockey         ClassI:LINE:Jockey  \n",
       "539650           Jockey-5_DK:ClassI:LINE:Jockey         ClassI:LINE:Jockey  \n",
       "539651           Jockey-5_DK:ClassI:LINE:Jockey         ClassI:LINE:Jockey  \n",
       "539652           Copia-10_SB-I:ClassI:LTR:Copia           ClassI:LTR:Copia  \n",
       "539653           Copia-10_SB-I:ClassI:LTR:Copia           ClassI:LTR:Copia  \n",
       "539654                   LOA_Ele7:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539655                   LOA_Ele7:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539656                 SART-5_MOc:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539657                 SART-5_MOc:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539658                    I-3B_CQ:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539659                    I-3B_CQ:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539662                    R1A_NVi:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539663                    R1A_NVi:ClassI:LINE:I              ClassI:LINE:I  \n",
       "539682  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "539683  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "539684  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "539685  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "539686  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "539687  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "539688  Helitron-3_ZM:ClassII:Helitron:Helitron  ClassII:Helitron:Helitron  \n",
       "\n",
       "[257944 rows x 11 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_repet_gff_filtered[p_repet_gff_filtered[1].str.contains('blast')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered.to_csv(out_dir+'/'+genome+'.REPET.long.df', sep='\\t', header = None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered['Class:Order:Superfamily'] = p_repet_gff_filtered['ID'].apply(lambda x: TE_COS_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These are the unique Class:Order:Superfamily classifiers of this dataframe:\n",
      "['ClassII:MITE:?' 'ClassI:LTR:?' 'ClassII:?:?' 'ClassI:LTR:Gypsy'\n",
      " 'ClassII:TIR:?' 'ClassII:TIR:PIF-Harbinger' 'ClassI:LTR:Copia'\n",
      " 'ClassII:TIR:hAT' 'ClassI:LARD:?' 'ClassII:TIR:CACTA' 'noCat'\n",
      " 'ClassII:TIR:Tc1-Mariner' 'SSR' 'ClassI:DIRS:DIRS' 'ClassII:TIR:MuDR'\n",
      " 'ClassII:Maverick:?' 'ClassI:LTR:ERV' 'ClassII:Helitron:?' 'ClassI:TRIM:?'\n",
      " 'ClassII:TIR:P' 'ClassII:?:Academ' 'ClassI:SINE:?' 'ClassI:?:?'\n",
      " 'ClassII:Helitron:Helitron' 'ClassI:LINE:I' 'ClassI:Penelope:?'\n",
      " 'ClassI:LINE:Jockey' 'ClassII:Maverick:Maverick' 'ClassI:DIRS:?'\n",
      " 'ClassII:TIR:Transib' 'ClassII:?:Kolobok' 'ClassII:?:Sola'\n",
      " 'ClassII:?:Ginger1' 'ClassI:LTR:Retrovirus' 'ClassI:LINE:L1'\n",
      " 'ClassI:LTR:Bel-Pao' 'ClassI:LINE:?' 'ClassI:LINE:RTE' 'ClassI:LINE:R2'\n",
      " 'ClassII:Crypton:Crypton' 'ClassII:TIR:PiggyBac' 'ClassI:PLE:Penelope'\n",
      " 'ClassII:?:Ginger2_TDD']\n"
     ]
    }
   ],
   "source": [
    "print('These are the unique Class:Order:Superfamily classifiers of this dataframe:')\n",
    "print(p_repet_gff_filtered['Class:Order:Superfamily'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#have a rough summary of the coverage not considering overlaps.\n",
    "p_repet_gff_filtered.drop_duplicates(subset=[3,4,'ID'], inplace =True)\n",
    "p_repet_gff_filtered['Length'] = p_repet_gff_filtered[4] - p_repet_gff_filtered[3]\n",
    "p_repet_gff_filtered['Class'] = p_repet_gff_filtered.apply(lambda row: TE_classification_filter(row['Class:Order:Superfamily'], 0), axis=1)\n",
    "p_repet_gff_filtered['Order'] = p_repet_gff_filtered.apply(lambda row: TE_classification_filter(row['Class:Order:Superfamily'], 1), axis=1)\n",
    "p_repet_gff_filtered['Superfamily'] = p_repet_gff_filtered.apply(lambda row: TE_classification_filter(row['Class:Order:Superfamily'], 2), axis=1)\n",
    "p_repet_gff_len_COS = p_repet_gff_filtered.groupby(by=['Class','Order','Superfamily'])['Length'].sum()\n",
    "p_repet_gff_len_S = p_repet_gff_filtered.groupby(by=['Class:Order:Superfamily'])['Length'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(\"This is the summary of overlapping coverage according to Class, Order, Superfamily\")\n",
    "print(p_repet_gff_len_COS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "print(\"This is the summary of overlapping coverage according to Superfamily\")\n",
    "print(p_repet_gff_len_S)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "num_unique_TEs = len(p_repet_gff_filtered[~p_repet_gff_filtered[1].str.contains('SSR')]['ID'].unique())\n",
    "num_unique_TE_super = len(p_repet_gff_filtered[~p_repet_gff_filtered[1].str.contains('SSR')]['Class:Order:Superfamily'].unique())\n",
    "\n",
    "print('This is the number of unique TEs: %i\\nThis is the number of unique TE superfamilies: %i' % (num_unique_TEs, num_unique_TE_super))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered.groupby(by=['Class','Order','Superfamily'])['Length'].count() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered.groupby(by=['Class:Order:Superfamily'])['Length'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "p_repet_gff_filtered.to_csv(out_dir+'/'+genome+'.REPET.long_v2.df', sep='\\t', header = None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#make new gff files where the ID column is the superfamily level\n",
    "p_repet_gff_superfamily = p_repet_gff_filtered.iloc[:,:]\n",
    "p_repet_gff_superfamily[8] = p_repet_gff_superfamily['Class:Order:Superfamily']\n",
    "p_repet_gff_superfamily.iloc[:,0:9].to_csv(out_dir+'/'+genome+'.REPET.superfamily.gff', sep='\\t', header = None, index=None,columns=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#make new gff file where the ID column is the TE level\n",
    "p_repet_gff_TE = p_repet_gff_filtered.iloc[:,:]\n",
    "p_repet_gff_TE[8] = p_repet_gff_TE['ID']\n",
    "p_repet_gff_TE.iloc[:,0:9].to_csv(out_dir+'/'+genome+'.REPET.TE.gff', sep='\\t', header = None, index=None,columns=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#generate the directory structure to safe specific coverage files\n",
    "os.chdir(out_dir)\n",
    "TE_types = ['Retrotransposon', 'DNA_transposon', 'noCat', 'SSR']\n",
    "TE_path = [os.path.join(out_dir, x) for x in TE_types]\n",
    "TE_path_dict = dict(zip(TE_types, TE_path))\n",
    "for TE_type in TE_types:\n",
    "    new_path = os.path.join(out_dir, TE_type)\n",
    "    if not os.path.exists(new_path):\n",
    "        os.mkdir(new_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# subset the id and safe in specific folder\n",
    "# return the subsetted file as bedtool\n",
    "def subset_id(_id, bed_object, repet_prefix):\n",
    "    #ClassI are retrotransposon form blast\n",
    "    if 'ClassI:' in _id:\n",
    "        out_path = TE_path_dict['Retrotransposon']   \n",
    "    #ClassII are DNA_transponson\n",
    "    elif 'ClassII' in _id:\n",
    "        out_path = TE_path_dict['DNA_transposon'] \n",
    "    #The rest with '_' should be REPET_TEs\n",
    "    elif _id == 'noCat':\n",
    "        out_path = TE_path_dict['noCat']\n",
    "    #everything without '_' at the end should be SSR\n",
    "    elif _id == 'SSR':\n",
    "        out_path = TE_path_dict['SSR']\n",
    "    out_fn = out_path+'/'+repet_prefix+'.'+_id+'.gff'\n",
    "    result = bed_object.filter(id_filter, _id).saveas(out_fn)\n",
    "    cov_fn = out_fn.replace('gff','cov')\n",
    "    cov = result.genome_coverage(dz=True,g=p_genome_file)\n",
    "    cov.saveas(cov_fn)\n",
    "    #_len = len(pd.read_csv(cov_fn, header=None, sep='\\t'))\n",
    "    #_dict[_id] = _len\n",
    "    #return pybedtools.BedTool(result.fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Next, we create a function to pass only features for a particular\n",
    "# featuretype.  This is similar to a \"grep\" operation when applied to every\n",
    "# feature in a BedTool\n",
    "def id_filter(feature, _id):\n",
    "    if feature[8] == _id:\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "repet_prefix_TE = genome+'.REPET.TE'\n",
    "repet_prefix_S = genome+'.REPET.superfamily'\n",
    "p_genome_file = genome+'.genome_file'\n",
    "genome_df = pd.read_csv(p_genome_file, sep='\\t', header=None,names=['contig', 'length'])\n",
    "\n",
    "genome_size = genome_df['length'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#pull in the classification gff, make classification array, loop over array to save all the cov_dataframes\n",
    "RE_TE_gff = pybedtools.BedTool(out_dir+'/'+genome+'.REPET.TE.gff')\n",
    "g_TE = RE_TE_gff.remove_invalid().saveas(out_dir+'/'+genome+'.REPET.TE.bedobject')\n",
    "#use the blast filtered dataframe as well\n",
    "RE_S_gff = pybedtools.BedTool(out_dir+'/'+genome+'.REPET.superfamily.gff')\n",
    "g_S = RE_S_gff.remove_invalid().saveas(out_dir+'/'+genome+'.REPET.superfamily.bedobject')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#use simple loop to loop over the bedcov genome coverage per classification. Keep track if everything is already done.\n",
    "jobs = []\n",
    "bed_file = g_S\n",
    "superfamilies = p_repet_gff_superfamily['Class:Order:Superfamily'].unique()\n",
    "for superfamily in superfamilies:\n",
    "    subset_id(superfamily, bed_file, repet_prefix_S)\n",
    "    print('Doing %s' % superfamily)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cur_dir = os.path.abspath(os.path.curdir)\n",
    "#this caputures all REPET classifications add the superfamily level\n",
    "class_cov_files = []\n",
    "for dirpath, dirname, filenames in os.walk(cur_dir, topdown=True):\n",
    "    if dirpath == cur_dir:\n",
    "        continue\n",
    "    cov_files = [dirpath +'/'+x for x in os.listdir(dirpath) if x.endswith('.cov') and repet_prefix_S in x]\n",
    "    for file in cov_files:\n",
    "        class_cov_files.append(file)\n",
    "\n",
    "#make a large summary dataframe from all the cov files where the last \n",
    "df_list =[]\n",
    "class_cov_files.sort()\n",
    "for file in class_cov_files:\n",
    "    print(file)\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header = None)\n",
    "    tmp_df[\"Class:Order:Superfamily\"] = file.split('.')[-2]\n",
    "    tmp_df.drop_duplicates(inplace=True) #drop all the duplicates meaning same position in the genome and same superfamily\n",
    "    df_list.append(tmp_df)\n",
    "    print(file.split('.')[-2])\n",
    "\n",
    "df_REPET_classification = pd.concat(df_list)\n",
    "df_REPET_classification.to_csv(out_dir+'/'+ repet_prefix_S +'.cov', sep='\\t', header =None, index=None)\n",
    "\n",
    "cov_per_superfamily = df_REPET_classification.pivot_table(values=1, columns= \"Class:Order:Superfamily\", aggfunc='count')\n",
    "cov_per_contig_per_superfamily = df_REPET_classification.groupby([0, \"Class:Order:Superfamily\"])[1].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cov_all_TEs = df_REPET_classification.drop_duplicates([0,1]) #this gets ride of the overlap between different TE families and classes\n",
    "cov_all_TEs = len(cov_all_TEs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#make superfamily df and add columns for Class, order and superfamily\n",
    "cov_per_superfamily_df = cov_per_superfamily.append(pd.DataFrame.from_dict({'cov_all_TEs': cov_all_TEs}, orient='index'))\n",
    "cov_per_superfamily_df.rename(columns={0: 'bp'}, inplace=True)\n",
    "cov_per_superfamily_df['%'] = cov_per_superfamily_df['bp']/genome_size*100\n",
    "cov_per_superfamily_df['Class:Order:Superfamily'] = cov_per_superfamily_df.index\n",
    "\n",
    "cov_per_superfamily_df['Class'] = cov_per_superfamily_df.apply(lambda row: TE_classification_filter(row['Class:Order:Superfamily'], 0), axis=1)\n",
    "cov_per_superfamily_df['Order'] = cov_per_superfamily_df.apply(lambda row: TE_classification_filter(row['Class:Order:Superfamily'], 1), axis=1)\n",
    "cov_per_superfamily_df['Superfamily'] = cov_per_superfamily_df.apply(lambda row: TE_classification_filter(row['Class:Order:Superfamily'], 2), axis=1)\n",
    "cov_per_superfamily_df.to_csv(out_dir+'/'+genome+'.REPET.summary.tab', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#consider combining these cov data frames into classes and orders as well and simply using those as id column, drop duplicats \n",
    "#use those as real coverage analysis at those level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cur_dir = os.path.abspath(os.path.curdir)\n",
    "#this caputures all REPET classifications add the superfamily level\n",
    "class_cov_files = []\n",
    "for dirpath, dirname, filenames in os.walk(cur_dir, topdown=True):\n",
    "    if dirpath == cur_dir:\n",
    "        continue\n",
    "    cov_files = [dirpath +'/'+x for x in os.listdir(dirpath) if x.endswith('.cov') and repet_prefix_S in x]\n",
    "    for file in cov_files:\n",
    "        class_cov_files.append(file)\n",
    "\n",
    "#make a large summary dataframe from all the cov files where the last \n",
    "df_list =[]\n",
    "class_cov_files.sort()\n",
    "for file in class_cov_files:\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header = None)\n",
    "    tmp_df[\"Class\"] = file.split('.')[-2].split(':')[0] #parse out the Class from the file name\n",
    "    tmp_df.drop_duplicates(inplace=True) #drop all the duplicates meaning same position in the genome and same Class\n",
    "    df_list.append(tmp_df)\n",
    "    print(file.split('.')[-2].split(':')[0])\n",
    "\n",
    "df_REPET_classification_class = pd.concat(df_list)\n",
    "df_REPET_classification_class.drop_duplicates(inplace=True)\n",
    "df_REPET_classification_class.to_csv(out_dir+'/'+ repet_prefix_S.replace('superfamily', 'Class') +'.cov', sep='\\t', header =None, index=None)\n",
    "\n",
    "cov_per_class = df_REPET_classification_class.pivot_table(values=1, columns= \"Class\", aggfunc='count')\n",
    "cov_per_contig_per_class = df_REPET_classification_class.groupby([0, \"Class\"])[1].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this parsing of of the order is neccessary to eliminate any overlap at the order level. See drop duplicates line 23.\n",
    "cur_dir = os.path.abspath(os.path.curdir)\n",
    "#this caputures all REPET classifications add the superfamily level\n",
    "class_cov_files = []\n",
    "for dirpath, dirname, filenames in os.walk(cur_dir, topdown=True):\n",
    "    if dirpath == cur_dir:\n",
    "        continue\n",
    "    cov_files = [dirpath +'/'+x for x in os.listdir(dirpath) if x.endswith('.cov') and repet_prefix_S in x]\n",
    "    for file in cov_files:\n",
    "        class_cov_files.append(file)\n",
    "\n",
    "#make a large summary dataframe from all the cov files where the last \n",
    "df_list =[]\n",
    "class_cov_files.sort()\n",
    "for file in class_cov_files:\n",
    "    tmp_df = pd.read_csv(file, sep='\\t', header = None)\n",
    "    if ':' in file:\n",
    "        tmp_df[\"Order\"] = ':'.join(file.split('.')[-2].split(':')[0:2]) #parse out the order from the file name\n",
    "        print(':'.join(file.split('.')[-2].split(':')[0:2]))\n",
    "    else:\n",
    "        tmp_df[\"Order\"] = file.split('.')[-2].split(':')[0]\n",
    "        print(file.split('.')[-2].split(':')[0])\n",
    "    tmp_df.drop_duplicates(inplace=True) #drop all the duplicates meaning same position in the genome and same Order\n",
    "    df_list.append(tmp_df)\n",
    "\n",
    "df_REPET_orderification_order = pd.concat(df_list)\n",
    "df_REPET_orderification_order.drop_duplicates(inplace=True)\n",
    "df_REPET_orderification_order.to_csv(out_dir+'/'+ repet_prefix_S.replace('superfamily', 'Order') +'.cov', sep='\\t', header =None, index=None)\n",
    "\n",
    "cov_per_order = df_REPET_orderification_order.pivot_table(values=1, columns= \"Order\", aggfunc='count')\n",
    "cov_per_contig_per_order = df_REPET_orderification_order.groupby([0, \"Order\"])[1].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cov_per_order_df = cov_per_order.append(pd.DataFrame.from_dict({'cov_all_TEs': cov_all_TEs}, orient='index'))\n",
    "\n",
    "cov_per_order_df.rename(columns={0: 'bp'}, inplace=True)\n",
    "\n",
    "cov_per_order_df['%'] = round(cov_per_order_df['bp']/genome_size*100, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cov_per_class_df = cov_per_class.append(pd.DataFrame.from_dict({'Total RE coverage': cov_all_TEs}, orient='index'))\n",
    "\n",
    "cov_per_class_df.rename(columns={0: 'bp'}, inplace=True)\n",
    "\n",
    "cov_per_class_df['%'] = round(cov_per_class_df['bp']/genome_size*100, 3)\n",
    "\n",
    "cov_per_class_df.sort_values('%', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.style.available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.style.use('seaborn-talk')\n",
    "fig, (ax0, ax1, ax2) = plt.subplots(nrows=3, ncols=1, figsize=(12,15))\n",
    "fig.suptitle(\"Repetitive elements in P. striformis f. sp. tritici %s\" % genome, fontsize=14, fontweight = 'bold')\n",
    "\n",
    "#color cycle from color blind people \n",
    "CB_color_cycle = ['#377eb8', '#ff7f00', '#4daf4a',\n",
    "                  '#f781bf', '#a65628', '#984ea3',\n",
    "'#999999', '#e41a1c', '#dede00']\n",
    "\n",
    "#plot the overall genome coverage by repetitive element category\n",
    "cov_per_class_df.plot(kind='barh', y='%', ax=ax0,  color='r')\n",
    "ax0.set_xlim([-5,60])\n",
    "ax0.legend().set_visible(False)\n",
    "ax0.set_yticklabels(list(cov_per_class_df.index),fontsize=10, fontweight='bold')\n",
    "ax0.set_ylabel(ylabel='RE categories', fontsize=14, fontweight='bold')\n",
    "\n",
    "#plot class I \n",
    "classI_df = classI_df = cov_per_superfamily_df[cov_per_superfamily_df['Class'] == 'Retrotransposon'].sort_values('%')\n",
    "\n",
    "#pick out the colors to do color matching on the order level\n",
    "tmp_cn = len(classI_df['Order'].unique())\n",
    "tmp_colors = CB_color_cycle[0:tmp_cn]\n",
    "tmp_col_dict = dict(zip(classI_df['Order'].unique(), tmp_colors))\n",
    "classI_df['Color'] = classI_df['Order'].apply(lambda x: tmp_col_dict[x])\n",
    "classI_df.plot(kind='barh', y = '%', ax=ax1, color=classI_df['Color'])\n",
    "ax1.set_xlim([-2,25])\n",
    "ax1.legend().set_visible(False)\n",
    "ax1.set_yticklabels(list(classI_df.index),fontsize=10, fontweight='bold')\n",
    "ax1.set_ylabel(ylabel='Class:Order:Superfamily', fontsize=14, fontweight='bold')\n",
    "ax1.set_title('ClassI: Retrotransposons', fontsize=14, fontweight='bold')\n",
    "\n",
    "#add tick lables\n",
    "\n",
    "for p, value in zip(ax1.patches, classI_df['%']):\n",
    "    ax1.annotate('{0:.3f}'.format(value), (18,p.get_y() * 1.005),fontsize=10, fontweight='bold' )\n",
    "\n",
    "#plot class II\n",
    "classII_df = classII_df = cov_per_superfamily_df[cov_per_superfamily_df['Class'] == 'DNA_transposon'].sort_values('%')\n",
    "\n",
    "#pick out the colors to do color matching on the order level\n",
    "tmp_cn = len(classII_df['Order'].unique())\n",
    "tmp_colors = CB_color_cycle[0:tmp_cn]\n",
    "tmp_col_dict = dict(zip(classII_df['Order'].unique(), tmp_colors))\n",
    "classII_df['Color'] = classII_df['Order'].apply(lambda x: tmp_col_dict[x])\n",
    "\n",
    "#plot the class II out\n",
    "classII_df.plot(kind='barh', y = '%', ax=ax2, color=classII_df['Color'])\n",
    "ax2.set_xlim([-2,25])\n",
    "ax2.legend().set_visible(False)\n",
    "ax2.set_yticklabels(list(classII_df.index),fontsize=10, fontweight='bold')\n",
    "ax2.set_ylabel(ylabel='Class:Order:Superfamily', fontsize=14, fontweight='bold')\n",
    "ax2.set_title('ClassII: DNA transposons', fontsize=14, fontweight='bold')\n",
    "ax2.set_xlabel('% genome coverage', fontsize=14, fontweight='bold')\n",
    "\n",
    "#add tick lables\n",
    "\n",
    "for p, value in zip(ax2.patches, classII_df['%']):\n",
    "    ax2.annotate('{0:.3f}'.format(value), (18 ,p.get_y() * 1.005),fontsize=10, fontweight='bold' )\n",
    "    \n",
    "fig.savefig(genome+'.REPET_summary.seaborn-talk.png', dpi=600, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Thats how far we got ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
